---
title: "Clean Slota workflow"
output: html_document
date: "2023-07-18"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

BiocManager::install("DropletUtils", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")
BiocManager::install("Seurat", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

remotes::install_github('chris-mcginnis-ucsf/DoubletFinder', lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

if (!require("SoupX", quietly = TRUE))
    install.packages("SoupX", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

if (!require("scCustomize", quietly = TRUE))
    install.packages("scCustomize", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

BiocManager::install("glmGamPoi", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

if (!require("DESeq2", quietly = TRUE))
    install.packages("DESeq2", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

install.packages("Polychrome", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")
BiocManager::install("org.Mm.eg.db", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")
BiocManager::install("clusterProfiler", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

```

```{r}
library(DropletUtils)
library(Seurat)
library(tidyverse)
library(dplyr)
library(stringr)
library(knitr)
library(DoubletFinder)
library(SoupX)
library(scCustomize)
library(DESeq2)
library(Polychrome)
library(org.Mm.eg.db)
library(clusterProfiler)
```

Make folder for plots

```{r}
#Specify the plotting folder to be made in the current working directory EDIT THIS ON EACH RUN TO AVOID OVERWRITING PLOTS
dir_name <- "Terminal_sample_investigation"

# create folder:
system(paste0("mkdir ", dir_name))

#save folder path for plot saving
plotting_path<-paste0("./", dir_name,"/")

rm(dir_name)
```

#Slota workflow

This workflow loads samples for the control and prion-treated data from
Slota group's hippocampal and cortex prion work. The samples are QC
assessed, filtered and plotted with clustering by gene expression
profiles.

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

# Load the data and format files for processing

```{bash}
# #CHATGPT assisted
# 
# #!/bin/bash
# 
# # Define the root directory
# root_dir="/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/SCP1962/other"
# 
# # Assuming the sample names are in a file called "Slota_samples.txt"
# sample_names_file="/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/Slota_samples.txt"
# 
# # Loop through each sample name in the file
# while IFS= read -r sample
# do
#     # Create a directory for the sample if it doesn't exist
#     mkdir -p "${root_dir}/${sample}"
# 
#     # Move the relevant files to the directory
#     mv "${root_dir}/${sample}"* "${root_dir}/${sample}"
# 
#     # Rename the files
#     mv "${root_dir}/${sample}/${sample}_barcodes.tsv" "${root_dir}/${sample}/barcodes.tsv"
#     mv "${root_dir}/${sample}/${sample}_features.tsv" "${root_dir}/${sample}/features.tsv"
#     mv "${root_dir}/${sample}/${sample}_matrix.mtx" "${root_dir}/${sample}/matrix.mtx"
# done < "$sample_names_file"
# 
# cd $root_dir
# gzip */{barcodes,features,matrix}*
```

```{r}
# Create list of folder names
sample_names<- readLines("/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/Slota_samples.txt")
#rename them for Slota folders


# Loop through each folder to process it
for (folder in sample_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/SCP1962/other/",folder)
  
  
  current_data<-Read10X(data.dir =folder_path)
  current_obj<-CreateSeuratObject(counts=current_data, project=folder)
  assign(paste0(folder), current_obj)
  
}

Slota_raw_obj_clean<-merge(PBS25HP, y=c(PBS48CX,PBS48HP,PBS60CX,PBS61CX,PBS73CX,RML122CX,RML122HP,RML132CX,RML132HP,RML133CX,RML133HP,RML134CX,RML134HP,RML138CX,RML138HP,RML140CX,RML140HP,RML142CX,RML142HP,RML145CX,RML145HP))

rm(current_data,current_obj, folder, sample_names)
```

# Tailor metadata

## Treatment annotation

adapted from <https://youtu.be/p49seH2_i8Y?t=312>

```{r}
Slota_raw_obj_clean$sample<-Slota_raw_obj_clean$orig.ident
#Dealing with new merged object 
Slota_raw_obj_clean@active.ident<-as.factor(Slota_raw_obj_clean$sample)

#Add metadata on treatment, adapted from Chat GPT

#Slota
sampleLabs_Slota<- (Slota_raw_obj_clean@active.ident)

treat_detect_Slota <- case_when(
  str_detect(sampleLabs_Slota, "RML.*HP*") ~ "RML_HP",
  str_detect(sampleLabs_Slota, "PBS.*HP*") ~ "PBS_HP",
  str_detect(sampleLabs_Slota, "RML.*CX*") ~ "RML_CX",
  str_detect(sampleLabs_Slota, "PBS.*CX*") ~ "PBS_CX",
  TRUE ~ NA_character_
)

Slota_raw_obj_clean@meta.data$treatment<-treat_detect_Slota
rm(treat_detect_Slota, sampleLabs_Slota)

```

## Identify mitochondrial and ribosomal populations

```{r}
#identify mitochondrial and ribosomal protein gene portions, annotations syntax tailored to mouse
Slota_raw_obj_clean[["percent.mt"]]<- PercentageFeatureSet(Slota_raw_obj_clean, pattern = "^Mt")
Slota_raw_obj_clean[["percent.rp"]]<- PercentageFeatureSet(Slota_raw_obj_clean, pattern = "^Rp")
Slota_raw_obj_clean[["percent.hb"]]<-PercentageFeatureSet(Slota_raw_obj_clean, "^Hb[^(p)]")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Slota_raw_obj_clean[["log10GenesPerUMI"]] <- log10(Slota_raw_obj_clean$nFeature_RNA) / log10(Slota_raw_obj_clean$nCount_RNA)

Slota_raw_obj_clean[["mitoRatio"]] <- Slota_raw_obj_clean@meta.data$percent.mt / 100
```

##Initial data assessment

```{r}
p1<-VlnPlot(Slota_raw_obj_clean, features=c("nFeature_RNA","nCount_RNA","percent.mt", "percent.rp","percent.hb"), ncol=5, pt.size = 0.0)#+ggtitle("Preliminary Data assessment", subtitle= "Slota, unfiltered")

# Plot each of these respectively, showing trends in cell health

#show mitochondrial proportion in reads
p2<-FeatureScatter(Slota_raw_obj_clean, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0)+ggtitle("Slota_UMIs per cell vs.mitochondrial genes detected", subtitle="UMIs per cell vs. % mitochondrial genes detected")


#show ribosomal proportion in reads
p3<-FeatureScatter(Slota_raw_obj_clean, feature1 = "nCount_RNA", feature2 = "percent.rp")+ggplot2::ggtitle("Slota_UMIs_per_cell_vsribosomal_protein_genes_detected", subtitle = "UMIs per cell vs. % ribosomal protein  genes detected")


#Ribosomal vs mitochondrial 
p4<-FeatureScatter(Slota_raw_obj_clean, feature1 = "percent.rp", feature2 = "percent.mt")+ggplot2::ggtitle("Slota_ribosomal_protein_genes_detected_vsmitochondrial_genes_detected",subtitle="% ribosomal protein  genes detected vs. % mitochondrial genes detected")

#Show overall reads and gene counts for unfiltered data.
p5<-FeatureScatter(Slota_raw_obj_clean, feature1 = "nCount_RNA", feature2="nFeature_RNA",raster=F, pt.size = 0.5)+ geom_smooth(method="lm")+ ggtitle("Slota_UMIs_per_cell_vs_number_of_genes",subtitle="UMIs per cell vs. number of genes ")

p1;p2;p3;p4;p5

#Save plots
png(paste0(plotting_path,"Terminal_EDA.png"), width = 3000,height = 1000)
p1
dev.off()
png(paste0(plotting_path,"Terminal_UMI_MT.png"), width = 1750,height = 1500)
p2
dev.off()
png(paste0(plotting_path,"Terminal_UMI_RP.png"),width = 1750,height = 1500)
p3
dev.off()
png(paste0(plotting_path,"Terminal_RP_MT.png"),  width = 1750,height = 1500)
p4
dev.off()
png(paste0(plotting_path,"Terminal_UMI_nGene.png"), width = 1750,height = 1500)
p5
dev.off()

```

Low mitochondrial counts, good overall quality indicating prefilters?

### Save the whole data and tidy up

```{r}
#Save the data object
save(Slota_raw_obj_clean, file="./Rdata/Slota_raw_obj_clean.RData")

rm(PBS25HP,PBS48CX,PBS48HP,PBS60CX,PBS61CX,PBS73CX,RML122CX,RML122HP,RML132CX,RML132HP,RML133CX,RML133HP,RML134CX,RML134HP,RML138CX,RML138HP,RML140CX,RML140HP,RML142CX,RML142HP,RML145CX,RML145HP)

gc()
```

#QC

##Exploration

### Create metadata variable to edit it without affecting core data

```{r}
# Rename columns from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md

#Create metadata variable to edit it without affecting core data
metadata_test<-Slota_raw_obj_clean@meta.data

#Saving cell barcodes as metainfo
metadata_test$cells<-rownames(metadata_test)

#Setting minimum feature count to filter cells with no genes UMIs vs. genes detected, as these will be discarded anyway
metadata_test <- subset(Slota_raw_obj_clean, subset= nFeature_RNA > 20)@meta.data

metadata_test <- metadata_test %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)
```

### Plots

```{r}
# Visualize the number of cell counts per sample with no filtering at all
p6<-metadata_test %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota_Cell_counts_per_sample", subtitle="Filtered to minimum 20 genes")


# Visualize the number UMIs/transcripts per cell
p7<-metadata_test %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Slota/UMI_transcripts_per_cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via histogram
p8<-metadata_test %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via boxplot
p9<-metadata_test %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="Filtered to minimum 20 genes")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p10<-metadata_test %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Slota UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle = "Filtered to minimum 20 genes")

# Visualize the distribution of mitochondrial gene expression detected per cell
p11<-metadata_test %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Slota distribution of mitochondrial gene expression per cell", subtitle = "Filtered to minimum 20 genes")

p6;p7;p8;p9;p10;p11

#Save plots
png(paste0(plotting_path,"Terminal_PreF_Counts_sample.png"), width = 2000,height = 1500)
p6
dev.off()
png(paste0(plotting_path,"Terminal_PreF_UMI_cell.png"),width = 2000,height = 1500)
p7
dev.off()
png(paste0(plotting_path,"Terminal_PreF_nGene_hist_sample.png"),width = 2000,height = 1500)
p8
dev.off()
png(paste0(plotting_path,"Terminal_PreF_nGene_hist_sample.png"),width = 2000,height = 1500)
p9
dev.off()
png(paste0(plotting_path,"Terminal_PreF_nGene_UMI_sample.png"),width = 2000,height = 1500)
p10
dev.off()
png(paste0(plotting_path,"Terminal_PreF_MT_cell.png"),width = 2000,height = 1500)
p11
dev.off()
```

## Filtering metadata

```{r}

#running author's filters

metadata_filtered<- metadata_test %>%
    filter(nGene > 1000 & percent.mt < 20 & percent.rp > 1 & percent.hb < 20 & nUMI < 35000) #FIXME check upper limit on nUMI
```

##Repeat QC plotting

```{r}
# Visualize the number of cell counts per sample with no filtering at all
p12<-metadata_filtered %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota_Cell_counts_per_sample", subtitle="nGene>150, nUMI>0, % mito< 20")


# Visualize the number UMIs/transcripts per cell
p13<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Slota_UMI_transcripts_per_cell", subtitle="nGene>150, nUMI>0, % mito< 20")


# Visualize the distribution of genes detected per cell via histogram
p14<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="nGene>150, nUMI>0, % mito< 20")


# Visualize the distribution of genes detected per cell via boxplot
p15<-metadata_filtered %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="nGene>150, nUMI>0, % mito< 20")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p16<-metadata_filtered %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Slota UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle="nGene>150, nUMI>0, % mito< 20")

# Visualize the distribution of mitochondrial gene expression detected per cell
p17<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Slota distribution of mitochondrial gene expression per cell", subtitle="nGene>150, nUMI>0, % mito< 20")

p12;p13;p14;p15;p16;p17

#Save plots
png(paste0(plotting_path,"Terminal_PostF_Counts_sample.png"), width = 2000,height = 1500)
p12
dev.off()
png(paste0(plotting_path,"Terminal_PostF_UMI_cell.png"),width = 2000,height = 1500)
p13
dev.off()
png(paste0(plotting_path,"Terminal_PostF_nGene_hist_sample.png"),width = 2000,height = 1500)
p14
dev.off()
png(paste0(plotting_path,"Terminal_PostF_nGene_hist_sample.png"),width = 2000,height = 1500)
p15
dev.off()
png(paste0(plotting_path,"Terminal_PostF_nGene_UMI_sample.png"),width = 2000,height = 1500)
p16
dev.off()
png(paste0(plotting_path,"Terminal_PostF_MT_cell.png"),width = 2000,height = 1500)
p17
dev.off()
```

## Assign filters to data

```{r}
#Assign the unfiltered metadata to Slota, updating the column names
Slota_raw_obj_clean@meta.data<-metadata_test 

Slota_filtered_clean<- subset(Slota_raw_obj_clean,subset= 
                              nGene > 1000 & 
                              percent.mt < 20 &
                              percent.rp > 1 &
                              percent.hb < 20 &
                              nUMI < 35000
                            )


# Gene-level filtering
# Keeping genes which appear in more than 3 cells for statistical power.

# Output a logical vector for every gene on whether the more than zero counts per cell
# Extract counts
counts <- GetAssayData(object = Slota_filtered_clean, slot = "counts")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 3 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 3

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
Slota_filtered_clean <- CreateSeuratObject(filtered_counts, meta.data = Slota_filtered_clean@meta.data)

rm(metadata_test, metadata_filtered,filtered_counts, nonzero,counts, keep_genes,Slota_raw_obj_clean)
gc()
```

### Saving object

```{r}
# Create .RData object to load at any time
save(Slota_filtered_clean, file="./Rdata/Slota_filtered_clean.RData")
```


# Splitting the data by tissue

```{r}
Slota_HP<-subset(Slota_filtered_clean, subset=sample=='PBS_HP' | treatment=='RML_HP')
Slota_CX<-subset(Slota_filtered_clean, subset=sample=='PBS_CX' | treatment=='RML_CX')

Slota_tissues<-list(Slota_HP,Slota_CX)

rm(Slota_HP,Slota_CX)
```




# Doublet finder for all samples

```{r}

#load in the filtered Slota object
#load("./Rdata/Slota_filtered_clean.RData")

#adapted from https://youtu.be/p49seH2_i8Y , https://github.com/kpatel427/YouTubeTutorials/blob/5b3c795cecafb22f0e992da069eae2efd6b5cb95/singleCell_doublets.R , https://rpubs.com/kenneditodd/doublet_finder_example 

#DoubletFinder, keep singlets
Slota_tissues <- lapply(Slota_tissues, function(tissue_group) {
  #Normalise+Scale with SCTransform?
  tissue_group <- SCTransform(tissue_group)
  tissue_group <- RunPCA(tissue_group)
  
  # Find significant PCs
  stdv <- tissue_group[["pca"]]@stdev
  sum.stdv <- sum(tissue_group[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] -
                       percent.stdv[2:length(percent.stdv)]) > 0.1),
              decreasing = TRUE)[1] + 1
  min.pc <- min(co1, co2)
  
  # Finish pre-processing
  tissue_group <- RunUMAP(tissue_group, dims = 1:min.pc)
  tissue_group <- FindNeighbors(object = tissue_group, dims = 1:min.pc)              
  tissue_group <- FindClusters(object = tissue_group, resolution = 0.1)
  
  # pK identification (no ground-truth)
  sweep.list <- paramSweep_v3(tissue_group, PCs = 1:min.pc, sct = TRUE)
  sweep.stats <- summarizeSweep(sweep.list)
  bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficient (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- as.numeric(bcmvn.max$pK)

  # Homotypic doublet proportion estimate
  annotations <- tissue_group@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.poi <- round(0.087 * nrow(tissue_group@meta.data)) #expected doublets based on values from 10X resources
  nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # Run DoubletFinder
  tissue_group <- doubletFinder_v3(seu = tissue_group, 
                             PCs = 1:min.pc, 
                             pK = optimal.pk,
                             nExp = nExp.poi.adj, 
                             sct = TRUE, 
                             pN=0.25)
  return(tissue_group)
  })
```

##Make singlet list

```{r}
#Rename the columns
Slota_list_clean <- lapply(Slota_list_clean, function(sample) {
  doublet_col_index <- which(grepl("DF.classifications_", colnames(sample@meta.data)))
  sample$doublet_finder <- sample@meta.data[doublet_col_index]
  sample@meta.data[doublet_col_index]<-NULL
  return(sample)
})


```

###Merge and save the object

```{r}
Slo_merged_clean <- Merge_Seurat_List(list_seurat = Slota_list_clean)

DefaultAssay(Slo_merged_clean)<-'RNA'

#FIXME show how many doublets there were
p18<-VlnPlot(Slo_merged_clean, features = "nFeature_RNA", group.by = 'doublet_finder', pt.size = 0.0)
table(Slo_merged_clean$doublet_finder)

save(Slo_merged_clean, file="Rdata/Slo_merged_clean.RData")

#subset data by singlet status and save list of singlets to continue with
Slo_singlets_merged_clean <- subset(Slo_merged_clean, subset= doublet_finder=='Singlet')

gc()

p18

#Save plots
png(paste0(plotting_path,"Terminal_Doublets.png"), width = 2000,height = 1500)
p18
dev.off()

```

Quite low nGenes for Doublts, try plotting the doublet info on the UMAP
of the whole dataset

```{r}
rm(Slota_list_clean,Slota_raw_obj_clean,Slota_filtered_clean)
```



# Dimensionality Reduction

Is integration of our data justified on the basis of treatment? Showing
where our data lies in the featurespace of genes to assess overlap of
treatments

```{r}
#https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_SC_SCT_normalization.md
# Normalize the counts
Slo_singlets_merged_clean <- NormalizeData(Slo_singlets_merged_clean)

# Identify the most variable genes with defaults
Slo_singlets_merged_clean <- FindVariableFeatures(Slo_singlets_merged_clean, verbose = T)
		     
# Scale the counts
Slo_singlets_merged_clean <- ScaleData(Slo_singlets_merged_clean)

# Perform PCA
Slo_singlets_merged_clean <- RunPCA(Slo_singlets_merged_clean)

#Inspect the variability accounted for by the PCS
#It appears that 23 PCs are responsible, so using 30 from now on
p19<-ElbowPlot(Slo_singlets_merged_clean, ndims=30)+ggtitle("Elbowplot of Slota data", subtitle = "NBH and ME7")
DimHeatmap(Slo_singlets_merged_clean, dims=1:9, cells=500, balanced=T)


#Finish dimensinoality reduction steps
Slo_singlets_merged_clean <- RunUMAP(Slo_singlets_merged_clean, reduction = "pca", dims = 1:30)
Slo_singlets_merged_clean <- FindNeighbors(Slo_singlets_merged_clean, reduction = "pca", dims = 1:30)
Slo_singlets_merged_clean <- FindClusters(Slo_singlets_merged_clean, resolution = c(0.2,0.3,0.5,0.7,1))

p19

#Save plots
png(paste0(plotting_path,"Terminal__.png"), width = 2000,height = 1500)
p19
dev.off()


#Save the object
save(Slo_singlets_merged_clean,file="Rdata/Slo_singlets_merged_clean.Rdata")

```

###Visualise the clustering resolutions for cluster number

```{r}
DimPlot(Slo_singlets_merged_clean, reduction = "umap", group.by = "treatment", raster=FALSE)+ ggtitle("Slota data, labelled by treatment", subtitle="Clustering resolution 1, 30 PCs used")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slo_singlets_merged_clean, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Slota data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#adapted from https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/08_SC_clustering_quality_control.md#segregation-of-clusters-by-various-sources-of-uninteresting-variation
# Determine metrics to plot present in Slo_singlets_merged_clean@meta.data
metrics <-  c("nUMI", "nGene", "mitoRatio")

#Explore any clustering mishaps by lookin for localisation of any of these traits in a given cluster
FeaturePlot(Slo_singlets_merged_clean, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            #min.cutoff = 'q10', #This was a default in the online script, I don't see the beenfit of colouring less of the range of values
            label = TRUE)
```

QC metrics do not skew by cluster, which is good

```{r}
#set the idents to 0.3
Slo_singlets_merged_clean@active.ident<-Slo_singlets_merged_clean@meta.data$RNA_snn_res.0.3
levels(Slo_singlets_merged_clean)

#Overwrite the saved the object
save(Slo_singlets_merged_clean,file="Rdata/Slo_singlets_merged_clean.Rdata")
```

#Can go from here to repeat astrocyte and DEG

```{r}
#load("Rdata/Slo_singlets_merged_clean.Rdata")

DefaultAssay(Slo_singlets_merged_clean)<-"RNA"

#Check for a gene of interest in the data's geneset
Gene_Names<-Slo_singlets_merged_clean@assays[["RNA"]]@counts@Dimnames[[1]]
filtered_gene_names<-Gene_Names[!grepl("^\\d", Gene_Names)]
filtered_gene_names[grep("^Slc1a2", filtered_gene_names)]

#try to identify astrocytes  with genes
astrocyte_genes<-c("Aldh1l1","Gfap","Slc1a2", "Slc1a3", "Glul", "Cd44")

FeaturePlot(Slo_singlets_merged_clean,features=astrocyte_genes, raster=F)#+ggtitle("Slota data, labelled with astrocyte gene markers from ABCAM", subtitle="Clustering resolution 0.3, 20 PCs used")

#As a reminder...
DimPlot(Slo_singlets_merged_clean, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Slota data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#No compare to the plots of astrocyte genes
VlnPlot(Slo_singlets_merged_clean, features = astrocyte_genes, group.by = 'RNA_snn_res.0.3', pt.size = 0.0)

#Glul is maybe a good indicator?
```

Treatment overlaps alot, so intergation is not required. The astrocyte
markers show that clusters 6, 14, 20 and 24 expressed the astrocyte
markers the strongest in our clusters or most definitively anyway

\##(SingleR annotation)

Could go here, could be followed by bubbleplot
<https://rdrr.io/github/milescsmith/seuratBubblePlot/man/bubbleplot.html>

Alternatively, load information from the spreadsheet at
Analysis/ABCAM_markers.xlsx then bubble plot

#Investigate astrocytes, ensure identification

```{r}
# Convert cluster names to character
Idents(Slo_singlets_merged_clean) <- Slo_singlets_merged_clean$RNA_snn_res.0.3

#Subset out astrocytes
Slo_astrocytes<-subset(Slo_singlets_merged_clean, idents=c('6', '24','14','20'))
```

#DEG work

Prepare the biomart info

```{r}
#Adapted from http://girke.bioinformatics.ucr.edu/CSHL_RNAseq/mydoc/mydoc_systemPipeRNAseq_07/
#Get annotation info
library("biomaRt")
m <- useMart("ensembl", dataset="mmusculus_gene_ensembl")
```

##HP

```{r}
#Named raw before normalisation
Slo_astrocytes_HP_raw<-subset(Slo_astrocytes, subset=treatment=='PBS_HP' | treatment=='RML_HP')


#subset, then cluster? Or add Slota first then do this analysis?
DefaultAssay(Slo_astrocytes_HP_raw)<-"RNA"

# Run the standard workflow for visualization and clustering
Slo_astrocytes_HP <- ScaleData(Slo_astrocytes_HP_raw, verbose = FALSE)
Slo_astrocytes_HP <- RunPCA(Slo_astrocytes_HP, npcs = 30, verbose = FALSE)
ElbowPlot(Slo_astrocytes_HP, ndims=30)+ggtitle("Elbowplot of subset Slota astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Slo_astrocytes_HP <- RunUMAP(Slo_astrocytes_HP, reduction = "pca", dims = 1:20)
Slo_astrocytes_HP <- FindNeighbors(Slo_astrocytes_HP, reduction = "pca", dims = 1:20)
Slo_astrocytes_HP <- FindClusters(Slo_astrocytes_HP, resolution = c(0.2,0.3,0.5,0.7,1))

#save for reloading
save(Slo_astrocytes_HP, file="Rdata/Slo_astrocytes_HP.Rdata")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slo_astrocytes_HP, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Slota data's HP astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Slo_astrocytes_HP, group.by = 'RNA_snn_res.0.5', reduction = 'umap', repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Slo_astrocytes_HP, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Slo_astrocytes_HP,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.5
Slo_astrocytes_HP@active.ident<-(Slo_astrocytes_HP@meta.data$RNA_snn_res.0.5)
levels(Slo_astrocytes_HP)

DefaultAssay(Slo_astrocytes_HP)<-'RNA'

#by treatment, show the genes DE in Prion condition. LogFC set to a *1 increase in fold change
Slo_astrocytes_HP@active.ident<-as.factor(Slo_astrocytes_HP$treatment)
Slota_astrocyte_prion_markers_HP<-FindMarkers(Slo_astrocytes_HP, ident.1 = 'RML_HP', ident.2 = 'PBS_HP',min.pct = 0.10, verbose = TRUE, test.use = "MAST")

#Sort the list in descending log2FC and ascending pvalues
Slota_astrocyte_prion_markers_clean_HP<-Slota_astrocyte_prion_markers_HP[order(-Slota_astrocyte_prion_markers_HP$avg_log2FC), ]


#Use biomart info
Slota_astrocyte_prion_markers_clean_HP$gene<-rownames(Slota_astrocyte_prion_markers_clean_HP)

gene_ids <- unique(Slota_astrocyte_prion_markers_clean_HP$gene)
  
#get relevant info from the mart
gene_info <- getBM(attributes = c("external_gene_name", "description","chromosome_name"), filters = "external_gene_name", values = gene_ids, mart = m)
  
#apply the info to the gene list
  Slota_astrocyte_prion_markers_clean_HP <- merge(Slota_astrocyte_prion_markers_clean_HP, gene_info, by.x = "gene", by.y = "external_gene_name", all.x = TRUE)
  

#Save the file
write.csv(Slota_astrocyte_prion_markers_clean_HP, file = "Slota_astrocyte_prion_markers_clean_HP.csv")

save(Slota_astrocyte_prion_markers_clean_HP, file="Rdata/Slota_astrocyte_prion_markers_clean_HP.RData")

table(Slo_astrocytes_HP$treatment)
```

## CX

```{r}
#HP
Slo_astrocytes_CX_raw<-subset(Slo_astrocytes, subset=treatment=='PBS_CX' | treatment=='RML_CX')


#subset, then cluster? Or add Slota first then do this analysis?
DefaultAssay(Slo_astrocytes_CX_raw)<-"RNA"

# Run the standard workflow for visualization and clustering
Slo_astrocytes_CX <- ScaleData(Slo_astrocytes_CX_raw, verbose = FALSE)
Slo_astrocytes_CX <- RunPCA(Slo_astrocytes_CX, npcs = 30, verbose = FALSE)
ElbowPlot(Slo_astrocytes_CX, ndims=30)+ggtitle("Elbowplot of subset Slota astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Slo_astrocytes_CX <- RunUMAP(Slo_astrocytes_CX, reduction = "pca", dims = 1:20)
Slo_astrocytes_CX <- FindNeighbors(Slo_astrocytes_CX, reduction = "pca", dims = 1:20)
Slo_astrocytes_CX <- FindClusters(Slo_astrocytes_CX, resolution = c(0.2,0.3,0.5,0.7,1))

#save for reloading
save(Slo_astrocytes_CX, file="Rdata/Slo_astrocytes_CX.Rdata")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slo_astrocytes_CX, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Slota data's CX astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Slo_astrocytes_CX, group.by = 'RNA_snn_res.0.5', reduction = 'umap', repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Slo_astrocytes_CX, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Slo_astrocytes_CX,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.5
Slo_astrocytes_CX@active.ident<-(Slo_astrocytes_CX@meta.data$RNA_snn_res.0.5)
levels(Slo_astrocytes_CX)

DefaultAssay(Slo_astrocytes_CX)<-'RNA'

#by treatment, show the genes DE in Prion condition. LogFC set to a *1 increase in fold change
Slo_astrocytes_CX@active.ident<-as.factor(Slo_astrocytes_CX$treatment)
Slota_astrocyte_prion_markers_CX<-FindMarkers(Slo_astrocytes_CX, ident.1 = 'RML_CX', ident.2 = 'PBS_CX',min.pct = 0.10, verbose = TRUE, test.use = "MAST")

#Sort the list in descending log2FC and ascending pvalues
Slota_astrocyte_prion_markers_clean_CX<-Slota_astrocyte_prion_markers_CX[order(-Slota_astrocyte_prion_markers_CX$avg_log2FC), ]

#Use biomart info
Slota_astrocyte_prion_markers_clean_CX$gene<-rownames(Slota_astrocyte_prion_markers_clean_CX)

gene_ids <- unique(Slota_astrocyte_prion_markers_clean_CX$gene)
  
#get relevant info from the mart
gene_info <- getBM(attributes = c("external_gene_name", "description","chromosome_name"), filters = "external_gene_name", values = gene_ids, mart = m)
  
#apply the info to the gene list
  Slota_astrocyte_prion_markers_clean_CX <- merge(Slota_astrocyte_prion_markers_clean_CX, gene_info, by.x = "gene", by.y = "external_gene_name", all.x = TRUE)


#Save the file
write.csv(Slota_astrocyte_prion_markers_clean_CX, file = "Slota_astrocyte_prion_markers_clean_CX.csv")

save(Slota_astrocyte_prion_markers_clean_CX, file="Rdata/Slota_astrocyte_prion_markers_clean_CX.RData")

table(Slo_astrocytes_CX_raw$treatment)
```

#Functional annotation

```{r}

# Upregulated genes
Slota_astrocyte_prion_markers_clean_UP<-subset(Slota_astrocyte_prion_markers_clean, subset=avg_log2FC>0.25)

#Create a list of just the genes: 
Slota_DEGs = rownames(Slota_astrocyte_prion_markers_clean_UP)
#Convert this gene SYMBOL list into Entrez ID:
Slota_DEGs_entrez = bitr(Slota_DEGs, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

pathway_results_Slota_DEGs = enrichGO(gene = Slota_DEGs_entrez$ENTREZID, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)

ggp = barplot(pathway_results_Slota_DEGs , showCategory=20)
ggp2 = dotplot(pathway_results_Slota_DEGs, showCategory=10)
ggp3 = goplot(pathway_results_Slota_DEGs, showCategory = 10)
ggp4 = cnetplot(pathway_results_Slota_DEGs, categorySize=p.adjust)

ggp
ggp2
ggp3
ggp4

# ggsave("SlotaUpbarplot.jpeg", plot = ggp, dpi = 300, width = 10, height = 10)
# ggsave("SlotaUpdotplot.jpeg", plot = ggp2, dpi = 300, width = 10, height = 10)
# ggsave("SlotaUpgoplot.jpeg", plot = ggp3, dpi = 300, width = 10, height = 10)
# ggsave("SlotaUpcnetplot.jpeg", plot = ggp4, dpi = 300, width = 10, height = 10)

```

# Downregulated DEGS doesn't seem to work yet, too few DEGs?

```{r}

# Donwregulated genes
Slota_astrocyte_prion_markers_clean_DOWN<-subset(Slota_astrocyte_prion_markers_clean, subset=avg_log2FC<(-0.25))

#Create a list of just the genes: 
Slota_DEGs_DOWN = rownames(Slota_astrocyte_prion_markers_clean_DOWN)
#Convert this gene SYMBOL list into Entrez ID:
Slota_DEGs_entrez_DOWN = bitr(Slota_DEGs_DOWN, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

pathway_results_Slota_DEGs_DOWN = enrichGO(gene = Slota_DEGs_entrez_DOWN$ENTREZID, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)

ggp5 = barplot(pathway_results_Slota_DEGs_DOWN , showCategory=20)
ggp6 = dotplot(pathway_results_Slota_DEGs_DOWN, showCategory=10)
ggp7 = goplot(pathway_results_Slota_DEGs_DOWN, showCategory = 10)
ggp8 = cnetplot(pathway_results_Slota_DEGs_DOWN, categorySize=p.adjust)

ggp5
ggp6
ggp7
ggp8

# ggsave("SlotaDownbarplot.jpeg", plot = ggp5, dpi = 300, width = 10, height = 10)
# ggsave("SlotaDowndotplot.jpeg", plot = ggp6, dpi = 300, width = 10, height = 10)
# ggsave("SlotaDowngoplot.jpeg", plot = ggp7, dpi = 300, width = 10, height = 10)
# ggsave("SlotaDowncnetplot.jpeg", plot = ggp8, dpi = 300, width = 10, height = 10)

```

##Plotting all astrocytes from this dataset for the presentation

```{r}

#subset, then cluster? Or add Slota first then do this analysis?
DefaultAssay(Slo_astrocytes)<-"RNA"

# Run the standard workflow for visualization and clustering
Slo_astrocytes_processed <- ScaleData(Slo_astrocytes, verbose = FALSE)
Slo_astrocytes_processed <- RunPCA(Slo_astrocytes_processed, npcs = 30, verbose = FALSE)
ElbowPlot(Slo_astrocytes_processed, ndims=30)+ggtitle("Elbowplot of subset Slota astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Slo_astrocytes_processed <- RunUMAP(Slo_astrocytes_processed, reduction = "pca", dims = 1:20)
Slo_astrocytes_processed <- FindNeighbors(Slo_astrocytes_processed, reduction = "pca", dims = 1:20)
Slo_astrocytes_processed <- FindClusters(Slo_astrocytes_processed, resolution = c(0.2,0.3,0.5,0.7,1))

FeaturePlot(Slo_astrocytes_processed,features="Gfap", raster=F)+ggtitle("Terminal data astrocytes", subtitle="Labelled with established astrocyte gene marker Slc1a2")

DimPlot(Slo_astrocytes_processed, 
        reduction = 'umap',
        raster=F, 
        group.by='treatment',
        cols=c('#00BFC4','#00BFC4','#f8766d','#f8766d'),
        order=c('RML_CX','RML_HP')
        )+ggtitle("Terminal data astrocytes", subtitle="Labelled by treatment")

#save for reloading
save(Slo_astrocytes_processed, file="Rdata/Slo_astrocytes_processed.Rdata")

#For presentation plotting
DimPlot(Slo_singlets_merged_clean,
        reduction = "umap",
        group.by = "treatment",
        raster=FALSE,
        cols=c('#00BFC4','#00BFC4','#f8766d','#f8766d'),
        order=c('PBS_CX','PBS_HP')
        )+ ggtitle("Terminal data", subtitle="Labelled by treatment. Clustering resolution 1, 30 PCs used")
```
