---
title: "Clean Verity workflow"
output: html_document
date: "2023-07-18"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

BiocManager::install("DropletUtils", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")
BiocManager::install("Seurat", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

remotes::install_github('chris-mcginnis-ucsf/DoubletFinder', lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

if (!require("SoupX", quietly = TRUE))
    install.packages("SoupX", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

if (!require("scCustomize", quietly = TRUE))
    install.packages("scCustomize", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

BiocManager::install("glmGamPoi", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

if (!require("DESeq2", quietly = TRUE))
    install.packages("DESeq2", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

install.packages("Polychrome", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")
BiocManager::install("org.Mm.eg.db", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")
BiocManager::install("clusterProfiler", lib="/exports/cmvm/eddie/eb/groups/mabbott_grp/Fraser/Project_23/Rpackages")

```

```{r}
library(DropletUtils)
library(Seurat)
library(tidyverse)
library(dplyr)
library(stringr)
library(knitr)
library(DoubletFinder)
library(SoupX)
library(scCustomize)
library(DESeq2)
library(Polychrome)
library(org.Mm.eg.db)
library(clusterProfiler)
```

#Verity workflow

This workflow loads samples for the control and prion-treated data from Mabbott group's hippocampal prion work. The samples (not including the ageing samples sample9 and 11) are QC assessed, filtered and plotted with clustering by gene expression profiles.

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).


```{r}
#load Nick Verity's data 

#Comment out the samples which aren't needed for the comparison of interest

# List of folder names
folder_names <- c(
  "sample1", "sample2",
  "sample5", "sample6"
  )

# Loop through each folder
for (folder_name in folder_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/Group_data/",folder_name, "/DGE_unfiltered")
  
  #Load the matrix file
  current_parse <- ReadParseBio(folder_path)
  
  #see if there are any empty gene names
  table(rownames(current_parse) == "") 
  
  #Load the metadata
  current_metadata<-read.csv(paste0(folder_path,"/cell_metadata.csv"))

  #Create a seurat object
  current_parse_interpreted <- CreateSeuratObject(current_parse, meta.data = current_metadata, names.delim = "")
  
  #Save the cell names as metadat and set identities to the sample
  current_parse_interpreted@meta.data$cell_name<-rownames(current_parse_interpreted@meta.data)
  current_parse_interpreted@meta.data$sample <- factor(rep(folder_name, nrow(current_parse_interpreted@meta.data)))
  
  Idents(current_parse_interpreted) <- current_parse_interpreted@meta.data$sample


  # Assign the object to a variable with the folder name as the variable name
  assign((paste0(folder_name, "_obj")),current_parse_interpreted)

}


#merge the samples into one object
Verity_total_clean <- merge(sample1_obj, y=c(sample2_obj, sample5_obj, sample6_obj), add.cell.ids=c("sample1", "sample2", "sample5", "sample6"), project="Verity")

#cleaning up memory
rm("sample1_obj","sample2_obj", "sample5_obj", "sample6_obj", current_metadata, current_parse_interpreted, current_parse)
```

# Tailor metadata

```{r}
metadata_to_remove<-c("bc_wells","species","gene_count","tscp_count", "mread_count","bc1_well","bc2_well","bc3_well","bc1_wind","bc2_wind", "bc3_wind")

for (var in metadata_to_remove){
  Verity_total_clean[[var]]<-NULL
}
```


##treatment annotation

```{r}
sampleLabs<-names(Verity_total_clean@active.ident)

#FIXME change these to be treatment levels of experimental setup

#Assigning names to the samples to keep identity
#From Chat GPT
treat_detect <- case_when(
  str_detect(sampleLabs, "sample1") ~ "NBH",
  str_detect(sampleLabs, "sample2") ~ "NBH",
  str_detect(sampleLabs, "sample5") ~ "ME7",
  str_detect(sampleLabs, "sample6") ~ "ME7",
  TRUE ~ NA_character_
)

#Save as metadata
Verity_total_clean@meta.data$treatment<-treat_detect
rm(sampleLabs, treat_detect)
```

## Identify mitochondrial and ribosomal populations

```{r}
#ID mitochondrial dominant cells 
Verity_total_clean[["percent.mt"]]<- PercentageFeatureSet(Verity_total_clean, pattern = "^mt")
Verity_total_clean[["percent.rp"]]<- PercentageFeatureSet(Verity_total_clean, pattern = "^Rp")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Verity_total_clean$log10GenesPerUMI <- log10(Verity_total_clean$nFeature_RNA) / log10(Verity_total_clean$nCount_RNA)

Verity_total_clean$mitoRatio <- Verity_total_clean@meta.data$percent.mt / 100


View(Verity_total_clean@meta.data)
```

##Initial data assessment

```{r}
VlnPlot(Verity_total_clean, features=c("nFeature_RNA","nCount_RNA","percent.mt", "percent.rp"), ncol=4, pt.size = 0.0)+ ggtitle("Preliminary Data assessment", subtitle= "Verity, unfiltered")


# Plot each of these respectively, showing trends in cell health

#show mitochondrial proportion in reads
FeatureScatter(Verity_total_clean, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0)+ggtitle("Verity_UMIs per cell vs.mitochondrial genes detected", subtitle="UMIs per cell vs. % mitochondrial genes detected")


#show ribosomal proportion in reads
FeatureScatter(Verity_total_clean, feature1 = "nCount_RNA", feature2 = "percent.rp")+ggplot2::ggtitle("Verity_UMIs_per_cell_vsribosomal_protein_genes_detected", subtitle = "UMIs per cell vs. % ribosomal protein  genes detected")


#Ribosomal vs mitochondrial 
FeatureScatter(Verity_total_clean, feature1 = "percent.rp", feature2 = "percent.mt")+ggplot2::ggtitle("Verity_ribosomal_protein_genes_detected_vsmitochondrial_genes_detected",subtitle="% ribosomal protein  genes detected vs. % mitochondrial genes detected")

#Show overall reads and gene counts for unfiltered data.
FeatureScatter(Verity_total_clean, feature1 = "nCount_RNA", feature2="nFeature_RNA",raster=F, pt.size = 0.5)+ geom_smooth(method="lm")+ ggtitle("Verity_UMIs_per_cell_vs_number_of_genes",subtitle="UMIs per cell vs. number of genes ")
```

### Save the whole data

```{r}
#Save the data object
save(Verity_total_clean, file="./Rdata/Verity_total_clean.RData")
```

#QC

##Exploration

### Create metadata variable to edit it without affecting core data

```{r}
# Rename columns from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md

#Create metadata variable to edit it without affecting core data
metadata_test<-Verity_total_clean@meta.data

#Saving cell barcodes as metainfo
metadata_test$cells<-rownames(metadata_test)

metadata_test <- metadata_test %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)

#Setting minimum feature count to filter cells with no genes UMIs vs. genes detected, as these will be discarded anyway
metadata_test <- subset(Verity_total_clean, subset= nFeature_RNA > 20)@meta.data

```


### Plots


```{r}
# Visualize the number of cell counts per sample with no filtering at all
metadata_test %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity_Cell_counts_per_sample", subtitle="Filtered to minimum 20 genes")

#Visualise the cutoff when setting nUMI to 35k, potential loss of biological information
metadata_test %>%
    filter(nUMI > 500, nGene >1100, nUMI <50000, percent.mt <20) %>%
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Verity/UMI_transcripts_per_cell", subtitle="Filtered to maximum of 50000 UMIs")

#This shows over 300k cells for some samples, indicating that minimum metrics must be set to remove junk.

# Visualize the number UMIs/transcripts per cell
metadata_test %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Verity/UMI_transcripts_per_cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via histogram
metadata_test %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via boxplot
metadata_test %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="Filtered to minimum 20 genes")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata_test %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Verity UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle = "Filtered to minimum 20 genes")

# Visualize the distribution of mitochondrial gene expression detected per cell
metadata_test %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Verity distribution of mitochondrial gene expression per cell", subtitle = "Filtered to minimum 20 genes")
```

## Filtering metadata

```{r}
metadata_filtered<- metadata_test %>%
    filter(nUMI > 500, nGene >1100, nUMI <35000, percent.mt <20) 
```

##Repeat QC plotting

```{r}
# Visualize the number of cell counts per sample with no filtering at all
metadata_filtered %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity_Cell_counts_per_sample", subtitle="nGene>1100, nUMI>500, % mito< 20")


# Visualize the number UMIs/transcripts per cell
metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Verity_UMI_transcripts_per_cell", subtitle="nGene>1100, nUMI>500, % mito< 20")


# Visualize the distribution of genes detected per cell via histogram
metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="nGene>1100, nUMI>500, % mito< 20")


# Visualize the distribution of genes detected per cell via boxplot
metadata_filtered %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="nGene>1100, nUMI>500, % mito< 20")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata_filtered %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Verity UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle="nGene>1100, nUMI>500, % mito< 20")

# Visualize the distribution of mitochondrial gene expression detected per cell
metadata_filtered %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Verity distribution of mitochondrial gene expression per cell", subtitle="nGene>1100, nUMI>500, % mito< 20")
```

## Assign filters to data

```{r}
#Assign the unfiltered metadata to Verity, updating the column names
Verity_total_clean@meta.data<-metadata_test 

Verity_filtered_clean<- subset(Verity_total_clean,subset= 
                              percent.mt <20 &
                              nUMI > 500 &
                              nUMI < 35000 &
                              nGene >1100
                            )


# Gene-level filtering
# Keeping genes which appear in more than 3 cells for statistical power.

# Output a logical vector for every gene on whether the more than zero counts per cell
# Extract counts
counts <- GetAssayData(object = Verity_filtered_clean, slot = "counts")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 3 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 3

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
Verity_filtered_clean <- CreateSeuratObject(filtered_counts, meta.data = Verity_filtered_clean@meta.data)

rm(metadata_test, metadata_filtered,filtered_counts, nonzero,counts, keep_genes,folder_name,folder_names,folder_path,metadata_to_remove,var)
gc()
```


### Saving object

```{r}
# Create .RData object to load at any time
save(Verity_filtered_clean, file="./Rdata/Verity_filtered_clean.RData")
```


# Doublet finder for all samples

```{r}

#load in the filtered Verity object
#load("./Rdata/Verity_filtered_clean.RData")

#adapted from https://youtu.be/p49seH2_i8Y , https://github.com/kpatel427/YouTubeTutorials/blob/5b3c795cecafb22f0e992da069eae2efd6b5cb95/singleCell_doublets.R , https://rpubs.com/kenneditodd/doublet_finder_example 

Verity_list_clean <- SplitObject(Verity_filtered_clean, split.by = "sample")

# Perform doublet detection on each sample using lapply
Verity_list_clean <- lapply(Verity_list_clean, function(sample) {
  print(paste0("Sample ",sample@meta.data$sample[1]))
  # Pre-process seurat object with standard seurat workflow
  sample <- SCTransform(sample)
  sample <- RunPCA(sample)
  
  # Find significant PCs
  stdv <- sample[["pca"]]@stdev
  sum.stdv <- sum(sample[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] -
                       percent.stdv[2:length(percent.stdv)]) > 0.1),
              decreasing = TRUE)[1] + 1
  min.pc <- min(co1, co2)
  
  # Finish pre-processing
  sample <- RunUMAP(sample, dims = 1:min.pc)
  sample <- FindNeighbors(object = sample, dims = 1:min.pc)              
  sample <- FindClusters(object = sample, resolution = 0.1)
  
  # pK identification (no ground-truth)
  sweep.list <- paramSweep_v3(sample, PCs = 1:min.pc, sct = TRUE)
  sweep.stats <- summarizeSweep(sweep.list)
  bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficient (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- as.numeric(bcmvn.max$pK)

  # Homotypic doublet proportion estimate
  annotations <- sample@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.poi <- round(0.03 * nrow(sample@meta.data)) #expected doublets based on values from PMC7424855
  nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # Run DoubletFinder
  sample <- doubletFinder_v3(seu = sample, 
                             PCs = 1:min.pc, 
                             pK = optimal.pk,
                             nExp = nExp.poi.adj, 
                             sct = TRUE, 
                             pN=0.25)
  
  return(sample)
})
```

##Make singlet list

```{r}
#Rename the columns
Verity_list_clean <- lapply(Verity_list_clean, function(sample) {
  doublet_col_index <- which(grepl("DF.classifications_", colnames(sample@meta.data)))
  sample$doublet_finder <- sample@meta.data[doublet_col_index]
  sample@meta.data[doublet_col_index]<-NULL
  return(sample)
})


```

###Merge and save the object

```{r}
Ver_merged_clean <- Merge_Seurat_List(list_seurat = Verity_list_clean)

DefaultAssay(Ver_merged_clean)<-'RNA'

#FIXME show how many doublets there were
VlnPlot(Ver_merged_clean, features = "nFeature_RNA", group.by = 'doublet_finder', pt.size = 0.1)

save(Ver_merged_clean, file="Rdata/Ver_merged_clean.RData")

#subset data by singlet status and save list of singlets to continue with
Ver_singlets_merged_clean <- subset(Ver_merged_clean, subset= doublet_finder=='Singlet')

gc()
```


```{r}
# #Set identities as treatment
# Idents(object=Ver_singlets_merged_clean)<-"treatment"
# 
# # split the dataset into a list of two seurat objects (ME7, NBH)
# Ver_list_clean<-SplitObject(Ver_singlets_merged_clean, split.by = "treatment")
# rm(Ver_singlets_merged_clean)
# gc()
# 
# #set up comparison from control and prion
# Ver_ME7_NBH_clean<-list(Ver_list_clean$NBH, Ver_list_clean$ME7)
# table(Ver_ME7_NBH_clean[[1]]$treatment)
# table(Ver_ME7_NBH_clean[[2]]$treatment)
```

# Dimensionality Reduction 

Is integration of our data justified on the basis of treatment? Showing where our data lies in the featurespace of genes to assess overlap of treatments

```{r}
#https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_SC_SCT_normalization.md
# Normalize the counts
Ver_singlets_merged_clean <- NormalizeData(Ver_singlets_merged_clean)

# Identify the most variable genes with defaults
Ver_singlets_merged_clean <- FindVariableFeatures(Ver_singlets_merged_clean, verbose = T)
		     
# Scale the counts
Ver_singlets_merged_clean <- ScaleData(Ver_singlets_merged_clean)

# Perform PCA
Ver_singlets_merged_clean <- RunPCA(Ver_singlets_merged_clean)

#Inspect the variability accounted for by the PCS
#It appears that 29 PCs are responsible, so using 30 from now on
ElbowPlot(Ver_singlets_merged_clean, ndims=30)+ggtitle("Elbowplot of Verity data", subtitle = "NBH and ME7")
DimHeatmap(Ver_singlets_merged_clean, dims=1:9, cells=500, balanced=T)


#Finish dimensinoality reduction steps
Ver_singlets_merged_clean <- RunUMAP(Ver_singlets_merged_clean, reduction = "pca", dims = 1:30)
Ver_singlets_merged_clean <- FindNeighbors(Ver_singlets_merged_clean, reduction = "pca", dims = 1:30)
Ver_singlets_merged_clean <- FindClusters(Ver_singlets_merged_clean, resolution = c(0.2,0.3,0.5,0.7,1))

```


###Visualise the clustering resolutions for cluster number

```{r}
DimPlot(Ver_singlets_merged_clean, reduction = "umap", group.by = "treatment", raster=FALSE)+ ggtitle("Verity data, labelled by treatment", subtitle="Clustering resolution 1, 30 PCs used")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Ver_singlets_merged_clean, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Verity data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#adapted from https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/08_SC_clustering_quality_control.md#segregation-of-clusters-by-various-sources-of-uninteresting-variation
# Determine metrics to plot present in Ver_singlets_merged_clean@meta.data
metrics <-  c("nUMI", "nGene", "mitoRatio")

#Explore any clustering mishaps by lookin for localisation of any of these traits in a given cluster
FeaturePlot(Ver_singlets_merged_clean, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            #min.cutoff = 'q10', #This was a default in the online script, I don't see the beenfit of colouring less of the range of values
            label = TRUE)
```

```{r}
#set the idents to 0.3
Ver_singlets_merged_clean@active.ident<-Ver_singlets_merged_clean@meta.data$RNA_snn_res.0.3
levels(Ver_singlets_merged_clean)

#Save the object
save(Ver_singlets_merged_clean,file="Rdata/Ver_singlets_merged_clean.Rdata")
```

#Can go from here to repeat astrocyte and DEG

```{r}
#load("Rdata/Ver_singlets_merged_clean.Rdata")

DefaultAssay(Ver_singlets_merged_clean)<-"RNA"

#Check for a gene of interest in the data's geneset
Gene_Names<-Ver_singlets_merged_clean@assays[["RNA"]]@counts@Dimnames[[1]]
filtered_gene_names<-Gene_Names[!grepl("^\\d", Gene_Names)]
filtered_gene_names[grep("^Slc1a2", filtered_gene_names)]

#try to identify astrocytes  with genes
astrocyte_genes<-c("Aldh1l1","Gfap","Slc1a2", "Slc1a3", "Glul", "Cd44")

FeaturePlot(Ver_singlets_merged_clean,features=astrocyte_genes, raster=F)#+ggtitle("Verity data, labelled with astrocyte gene markers from ABCAM", subtitle="Clustering resolution 0.3, 20 PCs used")

#As a reminder...
DimPlot(Ver_singlets_merged_clean, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Verity data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#No compare to the plots of astrocyte genes
VlnPlot(Ver_singlets_merged_clean, features = astrocyte_genes)

#Glul is maybe a good indicator?
```
Treatment overlaps alot, so intergation is not required, however nGene and nUMI are heavy skewed and lowest in clusters including our supposed astrocytes (as will be shown below). The astrocyte markers show that clusters 3 and 26 expressed the astrocyte markers the strongest in our clusters or most definitively anyway)

##(SingleR annotation)

Could go here, could be followed by bubbleplot https://rdrr.io/github/milescsmith/seuratBubblePlot/man/bubbleplot.html

Alternatively, load information from the spreadsheet at Analysis/ABCAM_markers.xlsx then bubble plot 

#Investigate astrocytes, ensure identification

```{r}
# Convert cluster names to character
Idents(Ver_singlets_merged_clean) <- Ver_singlets_merged_clean$RNA_snn_res.0.3

#Subset out astrocytes
Ver_astrocytes<-subset(Ver_singlets_merged_clean, idents=c('3', '26'))
```

#DEG work

```{r}
#subset, then cluster? Or add Slota first then do this analysis?
DefaultAssay(Ver_astrocytes)<-"RNA"

# Run the standard workflow for visualization and clustering
Ver_astrocytes <- ScaleData(Ver_astrocytes, verbose = FALSE)
Ver_astrocytes <- RunPCA(Ver_astrocytes, npcs = 30, verbose = FALSE)
ElbowPlot(Ver_astrocytes, ndims=30)+ggtitle("Elbowplot of subset Verity astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Ver_astrocytes <- RunUMAP(Ver_astrocytes, reduction = "pca", dims = 1:20)
Ver_astrocytes <- FindNeighbors(Ver_astrocytes, reduction = "pca", dims = 1:20)
Ver_astrocytes <- FindClusters(Ver_astrocytes, resolution = c(0.2,0.3,0.5,0.7,1))

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Ver_astrocytes, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Verity data's astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Ver_astrocytes, group.by = 'RNA_snn_res.0.5', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Ver_astrocytes, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Ver_astrocytes,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.5
Ver_astrocytes@active.ident<-(Ver_astrocytes@meta.data$RNA_snn_res.0.5)
levels(Ver_astrocytes)

DefaultAssay(Ver_astrocytes)<-'RNA'

#by treatment, show the genes DE in Prion condition. LogFC set to a *1 increase in fold change
Ver_astrocytes@active.ident<-as.factor(Ver_astrocytes$treatment)
Verity_astrocyte_prion_markers<-FindMarkers(Ver_astrocytes, ident.1 = 'ME7', ident.2 = 'NBH',min.pct = 0.10, verbose = TRUE, test.use = "DESeq2",logfc.threshold = 0.25)

#Sort the list in descending log2FC and ascending pvalues
Verity_astrocyte_prion_markers_clean<-Verity_astrocyte_prion_markers[order(-Verity_astrocyte_prion_markers$avg_log2FC), ]
#Save the file
write.csv(Verity_astrocyte_prion_markers_clean, file = "Verity_astrocyte_prion_markers_clean.csv")

save(Verity_astrocyte_prion_markers_clean, file="Verity_astrocyte_prion_markers_clean.RData")
```

#Functional annotation 

```{r}

# Upregulated genes
Verity_astrocyte_prion_markers_clean_UP<-subset(Verity_astrocyte_prion_markers_clean, subset=avg_log2FC>0.25)

#Create a list of just the genes: 
Verity_DEGs = rownames(Verity_astrocyte_prion_markers_clean_UP)
#Convert this gene SYMBOL list into Entrez ID:
Verity_DEGs_entrez = bitr(Verity_DEGs, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

pathway_results_Verity_DEGs = enrichGO(gene = Verity_DEGs_entrez$ENTREZID, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)

ggp = barplot(pathway_results_Verity_DEGs , showCategory=20)
ggp2 = dotplot(pathway_results_Verity_DEGs, showCategory=10)
ggp3 = goplot(pathway_results_Verity_DEGs, showCategory = 10)
ggp4 = cnetplot(pathway_results_Verity_DEGs, categorySize=p.adjust)

ggp
ggp2
ggp3
ggp4

# ggsave("VerityUpbarplot.jpeg", plot = ggp, dpi = 300, width = 10, height = 10)
# ggsave("VerityUpdotplot.jpeg", plot = ggp2, dpi = 300, width = 10, height = 10)
# ggsave("VerityUpgoplot.jpeg", plot = ggp3, dpi = 300, width = 10, height = 10)
# ggsave("VerityUpcnetplot.jpeg", plot = ggp4, dpi = 300, width = 10, height = 10)

```

# Downregulated DEGS doesn't seem to work yet, too few DEGs?

```{r}

# Donwregulated genes
Verity_astrocyte_prion_markers_clean_DOWN<-subset(Verity_astrocyte_prion_markers_clean, subset=avg_log2FC<(-0.25))

#Create a list of just the genes: 
Verity_DEGs_DOWN = rownames(Verity_astrocyte_prion_markers_clean_DOWN)
#Convert this gene SYMBOL list into Entrez ID:
Verity_DEGs_entrez_DOWN = bitr(Verity_DEGs_DOWN, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

pathway_results_Verity_DEGs_DOWN = enrichGO(gene = Verity_DEGs_entrez_DOWN$ENTREZID, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)

ggp5 = barplot(pathway_results_Verity_DEGs_DOWN , showCategory=20)
ggp6 = dotplot(pathway_results_Verity_DEGs_DOWN, showCategory=10)
ggp7 = goplot(pathway_results_Verity_DEGs_DOWN, showCategory = 10)
ggp8 = cnetplot(pathway_results_Verity_DEGs_DOWN, categorySize=p.adjust)

ggp5
ggp6
ggp7
ggp8

# ggsave("VerityDownbarplot.jpeg", plot = ggp5, dpi = 300, width = 10, height = 10)
# ggsave("VerityDowndotplot.jpeg", plot = ggp6, dpi = 300, width = 10, height = 10)
# ggsave("VerityDowngoplot.jpeg", plot = ggp7, dpi = 300, width = 10, height = 10)
# ggsave("VerityDowncnetplot.jpeg", plot = ggp8, dpi = 300, width = 10, height = 10)

```