---
title: "Clean Verity workflow"
output: html_document
date: "2023-07-18"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Package management

```{r}
#.libPaths(new = "/home/s2268606/University_Directories/Project_23/R_packages_project")
#/opt/R/4.2.1/lib/R/library

update.packages(ask = FALSE)


install.packages("BiocManager")

BiocManager::install(c(
  "org.Mm.eg.db",
  "AnnotationDbi",
  "clusterProfiler",
  "tidyverse",
  'knitr',
  "DropletUtils",
  "Seurat",
  "SoupX",
  "scCustomize",
  "glmGamPoi",
  "DESeq2",
  "Polychrome",
  "ReactomePA"
  ))


remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')

remotes::install_github("satijalab/seurat", "seurat5")

devtools::install_github("junjunlab/GseaVis")

```

```{r}
library(org.Mm.eg.db)
library(clusterProfiler)
library(DropletUtils)
library(Seurat)
library(tidyverse)
library(dplyr)
library(stringr)
library(knitr)
library(DoubletFinder)
library(SoupX)
library(scCustomize)
library(DESeq2)
library(Polychrome)
library(GseaVis)
library(ReactomePA)
```

Make folder for plots and set seed

```{r}
#Specify the plotting folder to be made in the current working directory EDIT THIS ON EACH RUN TO AVOID OVERWRITING PLOTS
dir_name <- "Verity_plots"

# create folder:
system(paste0("mkdir ", dir_name))

#save folder path for plot saving
plotting_path<-paste0("../Analysis/", dir_name,"/")


#set seed for reproducibility
set.seed(42)

rm(dir_name)
```

#Verity workflow

This workflow loads samples for the control and prion-treated data from
Mabbott group's hippocampal prion work. The samples (not including the
ageing samples sample9 and 11) are QC assessed, filtered and plotted
with clustering by gene expression profiles.

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

```{r}
#load Nick Verity's data 

#Comment out the samples which aren't needed for the comparison of interest

# List of folder names
folder_names <- c(
  "sample1", "sample2",
  "sample5", "sample6",
  "sample9", "sample11"
  )

# Loop through each folder
for (folder_name in folder_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/Group_data/",folder_name, "/DGE_unfiltered")
  
  #Load the matrix file
  current_parse <- ReadParseBio(folder_path)
  
  #see if there are any empty gene names
  table(rownames(current_parse) == "") 
  
  #Load the metadata
  current_metadata<-read.csv(paste0(folder_path,"/cell_metadata.csv"))

  #Create a seurat object
  current_parse_interpreted <- CreateSeuratObject(current_parse, meta.data = current_metadata, names.delim = "")
  
  #Save the cell names as metadat and set identities to the sample
  current_parse_interpreted@meta.data$cell_name<-rownames(current_parse_interpreted@meta.data)
  current_parse_interpreted@meta.data$sample <- factor(rep(folder_name, nrow(current_parse_interpreted@meta.data)))
  
  Idents(current_parse_interpreted) <- current_parse_interpreted@meta.data$sample


  # Assign the object to a variable with the folder name as the variable name
  assign((paste0(folder_name, "_obj")),current_parse_interpreted)

}


#merge the samples into one object
Verity_total_clean <- merge(sample1_obj, y=c(sample2_obj, sample5_obj, sample6_obj,sample9_obj, sample11_obj), add.cell.ids=c("sample1", "sample2", "sample5", "sample6", "sample9", "sample11"), project="Verity")

#cleaning up memory
rm("sample1_obj","sample2_obj", "sample5_obj", "sample6_obj", "sample9_obj", "sample11_obj",current_metadata, current_parse_interpreted, current_parse)
```

# Tailor metadata

```{r}
metadata_to_remove<-c("bc_wells","species","gene_count","tscp_count", "mread_count","bc1_well","bc2_well","bc3_well","bc1_wind","bc2_wind", "bc3_wind")

for (var in metadata_to_remove){
  Verity_total_clean[[var]]<-NULL
}

```

# Load the data and format files for processing

```{bash}
#FIXME for later addition
```

## treatment annotation

adapted from <https://youtu.be/p49seH2_i8Y?t=312>

```{r}
sampleLabs<-names(Verity_total_clean@active.ident)

#Assigning names to the samples to keep identity
#From Chat GPT
treat_detect <- case_when(
  str_detect(sampleLabs, "sample11") ~ "AGEING", #assign this treatment condition first to avoid mis-labelling with sample1's label via str_detect
  str_detect(sampleLabs, "sample1") ~ "NBH",
  str_detect(sampleLabs, "sample2") ~ "NBH",
  str_detect(sampleLabs, "sample5") ~ "ME7",
  str_detect(sampleLabs, "sample6") ~ "ME7",
  str_detect(sampleLabs, "sample9") ~ "AGEING",
  TRUE ~ NA_character_
)

#Save as metadata
Verity_total_clean@meta.data$treatment<-treat_detect
rm(sampleLabs, treat_detect)
```

## Identify mitochondrial and ribosomal populations

```{r}
#ID mitochondrial dominant cells 
Verity_total_clean[["percent.mt"]]<- PercentageFeatureSet(Verity_total_clean, pattern = "^mt")
Verity_total_clean[["percent.rp"]]<- PercentageFeatureSet(Verity_total_clean, pattern = "^Rp")
Verity_total_clean[["percent.hb"]]<-PercentageFeatureSet(Verity_total_clean, "^Hb[^(p)]")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Verity_total_clean$log10GenesPerUMI <- log10(Verity_total_clean$nFeature_RNA) / log10(Verity_total_clean$nCount_RNA)

Verity_total_clean$mitoRatio <- Verity_total_clean@meta.data$percent.mt / 100


View(Verity_total_clean@meta.data)
```

##Initial data assessment

```{r}
p1<-VlnPlot(Verity_total_clean, features=c("nFeature_RNA","nCount_RNA","percent.mt", "percent.rp","percent.hb"), ncol=5, pt.size = 0.0)#+ggtitle("Preliminary Data assessment", subtitle= "Verity, unfiltered")

# Plot each of these respectively, showing trends in cell health

#show mitochondrial proportion in reads
p2<-FeatureScatter(Verity_total_clean, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0)+ggtitle("Verity_UMIs per cell vs.mitochondrial genes detected", subtitle="UMIs per cell vs. % mitochondrial genes detected")


#show ribosomal proportion in reads
p3<-FeatureScatter(Verity_total_clean, feature1 = "nCount_RNA", feature2 = "percent.rp")+ggplot2::ggtitle("Verity_UMIs_per_cell_vsribosomal_protein_genes_detected", subtitle = "UMIs per cell vs. % ribosomal protein  genes detected")


#Ribosomal vs mitochondrial 
p4<-FeatureScatter(Verity_total_clean, feature1 = "percent.rp", feature2 = "percent.mt")+ggplot2::ggtitle("Verity_ribosomal_protein_genes_detected_vsmitochondrial_genes_detected",subtitle="% ribosomal protein  genes detected vs. % mitochondrial genes detected")

#Show overall reads and gene counts for unfiltered data.
p5<-FeatureScatter(Verity_total_clean, feature1 = "nCount_RNA", feature2="nFeature_RNA",raster=F, pt.size = 0.5)+ geom_smooth(method="lm")+ ggtitle("Verity_UMIs_per_cell_vs_number_of_genes",subtitle="UMIs per cell vs. number of genes ")

p1;p2;p3;p4;p5

#Save plots
png(paste0(plotting_path,"Verity_EDA.png"), width = 3000,height = 1000)
p1
dev.off()
png(paste0(plotting_path,"Verity_UMI_MT.png"), width = 1750,height = 1500)
p2
dev.off()
png(paste0(plotting_path,"Verity_UMI_RP.png"),width = 1750,height = 1500)
p3
dev.off()
png(paste0(plotting_path,"Verity_RP_MT.png"),  width = 1750,height = 1500)
p4
dev.off()
png(paste0(plotting_path,"Verity_UMI_nGene.png"), width = 1750,height = 1500)
p5
dev.off()
```

### Save the whole data

```{r}
#Save the data object
save(Verity_total_clean, file=paste0(plotting_path,"Verity_total_clean.RData"))
gc()
```

#QC

##Exploration

### Create metadata variable to edit it without affecting core data

```{r}
# Rename columns from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md

#Create metadata variable to edit it without affecting core data
metadata_test<-Verity_total_clean@meta.data

#Saving cell barcodes as metainfo
metadata_test$cells<-rownames(metadata_test)

#Setting minimum feature count to filter cells with no genes UMIs vs. genes detected, as these will be discarded anyway
metadata_test <- subset(Verity_total_clean, subset= nFeature_RNA > 20)@meta.data

metadata_test <- metadata_test %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)



```

### Plots

```{r}
# Visualize the number of cell counts per sample with no filtering at all
p6<-metadata_test %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity_Cell_counts_per_sample", subtitle="Filtered to minimum 20 genes")


# Visualize the number UMIs/transcripts per cell
p7<-metadata_test %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Verity/UMI_transcripts_per_cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via histogram
p8<-metadata_test %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via boxplot
p9<-metadata_test %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="Filtered to minimum 20 genes")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p10<-metadata_test %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Verity UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle = "Filtered to minimum 20 genes")

# Visualize the distribution of mitochondrial gene expression detected per cell
p11<-metadata_test %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Verity distribution of mitochondrial gene expression per cell", subtitle = "Filtered to minimum 20 genes")

p6;p7;p8;p9;p10;p11

#Save plots
png(paste0(plotting_path,"Verity_PreF_Counts_sample.png"), width = 2000,height = 1500)
p6
dev.off()
png(paste0(plotting_path,"Verity_PreF_UMI_cell.png"),width = 2000,height = 1500)
p7
dev.off()
png(paste0(plotting_path,"Verity_PreF_nGene_hist_sample.png"),width = 2000,height = 1500)
p8
dev.off()
png(paste0(plotting_path,"Verity_PreF_nGene_hist_sample.png"),width = 2000,height = 1500)
p9
dev.off()
png(paste0(plotting_path,"Verity_PreF_nGene_UMI_sample.png"),width = 2000,height = 1500)
p10
dev.off()
png(paste0(plotting_path,"Verity_PreF_MT_cell.png"),width = 2000,height = 1500)
p11
dev.off()
```

## Filtering metadata

```{r}
metadata_filtered<- metadata_test %>%
    filter(nUMI > 500, nGene >1100, nUMI <35000, percent.mt <20, percent.hb < 20) 
```

##Repeat QC plotting

```{r}
# Visualize the number of cell counts per sample with no filtering at all
p12<-metadata_filtered %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity_Cell_counts_per_sample", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")


# Visualize the number UMIs/transcripts per cell
p13<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Verity_UMI_transcripts_per_cell", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")


# Visualize the distribution of genes detected per cell via histogram
p14<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")


# Visualize the distribution of genes detected per cell via boxplot
p15<-metadata_filtered %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p16<-metadata_filtered %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Verity UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")

# Visualize the distribution of mitochondrial gene expression detected per cell
p17<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Verity distribution of mitochondrial gene expression per cell", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")

p12;p13;p14;p15;p16;p17

#Save plots
png(paste0(plotting_path,"Verity_PostF_Counts_sample.png"), width = 2000,height = 1500)
p12
dev.off()
png(paste0(plotting_path,"Verity_PostF_UMI_cell.png"),width = 2000,height = 1500)
p13
dev.off()
png(paste0(plotting_path,"Verity_PostF_nGene_hist_sample.png"),width = 2000,height = 1500)
p14
dev.off()
png(paste0(plotting_path,"Verity_PostF_nGene_hist_sample.png"),width = 2000,height = 1500)
p15
dev.off()
png(paste0(plotting_path,"Verity_PostF_nGene_UMI_sample.png"),width = 2000,height = 1500)
p16
dev.off()
png(paste0(plotting_path,"Verity_PostF_MT_cell.png"),width = 2000,height = 1500)
p17
dev.off()
```

## Assign filters to data

```{r}
#Assign the unfiltered metadata to Verity, updating the column names
Verity_total_clean@meta.data<-metadata_test 

Verity_filtered_clean<- subset(Verity_total_clean,subset= 
                              percent.mt <20 &
                              nUMI > 500 &
                              nUMI < 35000 &
                              nGene >1100 & 
                              percent.hb < 20
                            )


# Gene-level filtering
# Keeping genes which appear in more than 3 cells for statistical power.

# Output a logical vector for every gene on whether the more than zero counts per cell
# Extract counts
counts <- GetAssayData(object = Verity_filtered_clean, slot = "counts")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 3 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 3

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
Verity_filtered_clean <- CreateSeuratObject(filtered_counts, meta.data = Verity_filtered_clean@meta.data)

p17a<-Verity_filtered_clean@meta.data %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Verity_Cell_counts_per_sample", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")

p17b<-Verity_filtered_clean@meta.data %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Verity distribution of genes per cell", subtitle="nGene>1100, 35000>nUMI>500, % mito< 20, % hb< 20")

p17a;p17b

rm(metadata_test, metadata_filtered,filtered_counts, nonzero,counts, keep_genes,folder_name,folder_names,folder_path,metadata_to_remove,var)
gc()
```

### Saving object and tidying up

```{r}
# Create .RData object to load at any time
save(Verity_filtered_clean, file=paste0(plotting_path,"Verity_filtered_clean.RData"))

rm(p1, p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,p13, p14,p15,p16,p17,p17a,p17b)
```

# DoubletFinder

##Splitting the data by treatment

```{r}
Verity_Prion_comp<-subset(Verity_filtered_clean, subset=treatment==c('NBH','ME7'))
Verity_Ageing_comp<-subset(Verity_filtered_clean, subset=treatment==c('NBH','AGEING'))

Verity_comparisons<-list(Verity_Prion_comp,Verity_Ageing_comp)

rm(Verity_Prion_comp,Verity_Ageing_comp, Verity_filtered_clean, Verity_total_clean)
gc()
```

## Doublet finder for treatment comparisons separately

```{r}

#load in the filtered Verity object
#load(paste0(plotting_path,"Verity_filtered_clean.RData")

#adapted from https://youtu.be/p49seH2_i8Y , https://github.com/kpatel427/YouTubeTutorials/blob/5b3c795cecafb22f0e992da069eae2efd6b5cb95/singleCell_doublets.R , https://rpubs.com/kenneditodd/doublet_finder_example 

#DoubletFinder, keep singlets
Verity_comparisons <- lapply(Verity_comparisons, function(comp_pair) {
  
  #Normalise+Scale with SCTransform?
  comp_pair <- SCTransform(comp_pair)
  comp_pair <- RunPCA(comp_pair)
  
  # Find significant PCs
  stdv <- comp_pair[["pca"]]@stdev
  sum.stdv <- sum(comp_pair[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] -
                       percent.stdv[2:length(percent.stdv)]) > 0.1),
              decreasing = TRUE)[1] + 1
  min.pc <- min(co1, co2)
  
  # Finish pre-processing
  comp_pair <- RunUMAP(comp_pair, dims = 1:min.pc, seed.use=42)
  comp_pair <- FindNeighbors(object = comp_pair, dims = 1:min.pc)              
  comp_pair <- FindClusters(object = comp_pair, resolution = 0.1)
  
  # pK identification (no ground-truth)
  sweep.list <- paramSweep_v3(comp_pair, PCs = 1:min.pc, sct = TRUE)
  sweep.stats <- summarizeSweep(sweep.list)
  bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficient (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- as.numeric(bcmvn.max$pK)

  # Homotypic doublet proportion estimate
  annotations <- comp_pair@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.poi <- round(0.03 * nrow(comp_pair@meta.data)) #expected doublets based on values from PMC7424855
  nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # Run DoubletFinder
  comp_pair <- doubletFinder_v3(seu = comp_pair, 
                             PCs = 1:min.pc, 
                             pK = optimal.pk,
                             nExp = nExp.poi.adj, 
                             sct = TRUE, 
                             pN=0.25)
  
  return(comp_pair)
})
```

# Split the data, rename doubletfinder metadata and Show Doublets for each comparison group witihn Verity data. Retain singlets

```{r}

#Rename the columns
Verity_comparisons <- lapply(Verity_comparisons, function(comp_pair) {
  DefaultAssay(comp_pair)<-'RNA'
  doublet_col_index <- which(grepl("DF.classifications_", colnames(comp_pair@meta.data)))
  comp_pair$doublet_finder <- comp_pair@meta.data[doublet_col_index]
  comp_pair@meta.data[doublet_col_index]<-NULL

  #Show how many doublets there were
  table(comp_pair$doublet_finder)
  
  return(comp_pair)
})

Verity_Prion_comp<-subset(Verity_comparisons[[1]], subset=sample !="sample11")

#Split the data
Verity_Prion_comp<-Verity_comparisons[[1]]
Verity_Ageing_comp<-Verity_comparisons[[2]]

#Plot how many doublets there were
p18<-VlnPlot(Verity_Prion_comp, features = "nFeature_RNA", group.by = 'doublet_finder', pt.size = 0.0)#+ggtitle(paste0("Verity data", comp_pair))

p19<-VlnPlot(Verity_Ageing_comp, features = "nFeature_RNA", group.by = 'doublet_finder', pt.size = 0.0)#+ggtitle(paste0("Verity data", comp_pair))


png("Verity_Prion_comp_doublets.png", width = 2000,height = 1500)
p18
dev.off()

png("Verity_Ageing_comp_doublets.png", width = 2000,height = 1500)
p19
dev.off()


#Save the data with Doublets retained too
save(Verity_comparisons, file=paste0(plotting_path,"Verity_comparisons.RData"))

#subset data by singlet status and save list of singlets to continue with
#FIXME might need to be in loop or explicit if subsetting doesn't work
Verity_Prion_comp <- subset(Verity_Prion_comp, subset= doublet_finder=='Singlet')
Verity_Ageing_comp <- subset(Verity_Ageing_comp, subset= doublet_finder=='Singlet')


gc()
```

# Prion work

```{r}
#active assay RNA
DefaultAssay(Verity_Prion_comp)<-"RNA"

#DimReduc again
#https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_SC_SCT_normalization.md
# Normalize the counts
Verity_Prion_comp <- NormalizeData(Verity_Prion_comp)
# Identify the most variable genes with defaults
Verity_Prion_comp <- FindVariableFeatures(Verity_Prion_comp, verbose = T)
		     
# Scale the counts
Verity_Prion_comp <- ScaleData(Verity_Prion_comp)
# Perform PCA
Verity_Prion_comp <- RunPCA(Verity_Prion_comp)
#Inspect the variability accounted for by the PCS
#It appears that 23 PCs are responsible, so using 30 from now on
ElbowPlot(Verity_Prion_comp, ndims=30)+ggtitle("Elbowplot of Verity data", subtitle = "NBH and ME7")
DimHeatmap(Verity_Prion_comp, dims=1:9, cells=500, balanced=T)
#Finish dimensinoality reduction steps
Verity_Prion_comp <- RunUMAP(Verity_Prion_comp, reduction = "pca", dims = 1:30, seed.use=42)
Verity_Prion_comp <- FindNeighbors(Verity_Prion_comp, reduction = "pca", dims = 1:30)
Verity_Prion_comp <- FindClusters(Verity_Prion_comp, resolution = c(0.2,0.3,0.5,0.7,1))
#Save the object
save(Verity_Prion_comp,file=paste0(plotting_path,"Verity_Prion_comp.Rdata"))


# Visualise the clustering resolutions for cluster number

# Determine metrics to plot present in Ver_singlets_merged_clean@meta.data
metrics <-  c("nUMI", "nGene", "mitoRatio")

#FIXME titles
DimPlot(Verity_Prion_comp , reduction = "umap", group.by = "treatment", raster=FALSE)+ ggtitle("Verity Hippocampal cells labelled by treatment")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Verity_Prion_comp, group.by = 'RNA_snn_res.0.2', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Verity Hippocampal cells labelled by cluster numbers", subtitle = "Cluster resolution 0.2")

#adapted from https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/08_SC_clustering_quality_control.md#segregation-of-clusters-by-various-sources-of-uninteresting-variation

#Explore any clustering mishaps by looking for localisation of any of these traits in a given cluster
FeaturePlot(Verity_Prion_comp, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            #min.cutoff = 'q10', #This was a default in the online script, I don't see the benefit of colouring less of the range of values
            label = TRUE)


## Save the singlets with a cluster resolution?
#FIXME might not be required
 #set the idents to cluster resolution 0.2
Verity_Prion_comp@active.ident<-Verity_Prion_comp@meta.data$RNA_snn_res.0.2
levels(Verity_Prion_comp)

rm(metrics)
```

```{r}

DefaultAssay(Verity_Prion_comp)<-"RNA"

#Check for a gene of interest in the data's geneset
Gene_Names<-Verity_Prion_comp@assays[["RNA"]]@counts@Dimnames[[1]]
filtered_gene_names<-Gene_Names[!grepl("^\\d", Gene_Names)]
filtered_gene_names[grep("^Slc1a2", filtered_gene_names)]

#try to identify astrocytes  with genes
astrocyte_genes<-c("Aldh1l1","Gfap","Slc1a2", "Slc1a3", "Glul", "Cd44")

FeaturePlot(Verity_Prion_comp,features=astrocyte_genes, raster=F)#+ggtitle("Verity data, labelled with astrocyte gene markers from ABCAM", subtitle="Clustering resolution 0.3, 20 PCs used")

#As a reminder...
DimPlot(Verity_Prion_comp, group.by = 'RNA_snn_res.0.2', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Verity data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#No compare to the plots of astrocyte genes
VlnPlot(Verity_Prion_comp, features = astrocyte_genes, group.by = 'RNA_snn_res.0.2', pt.size = 0.0)

#Glul is maybe a good indicator?
```

## (SingleR annotation)

Could go here, could be followed by bubbleplot
<https://rdrr.io/github/milescsmith/seuratBubblePlot/man/bubbleplot.html>

Alternatively, load information from the spreadsheet at
Analysis/ABCAM_markers.xlsx then bubble plot

##Investigate astrocytes, ensure identification

```{r}
# Convert cluster names to character
Verity_Prion_comp@active.ident <- as.factor(Verity_Prion_comp$RNA_snn_res.0.2)

#Subset out astrocytes
Verity_Prion_comp_astrocytes<-subset(Verity_Prion_comp, idents=c('6','11'))

save(Verity_Prion_comp,file=paste0(plotting_path,"Verity_Prion_comp.Rdata"))
Verity_Prion_comp_astrocytes
```

## DEG work

Prepare the biomart info

```{r}
#Adapted from http://girke.bioinformatics.ucr.edu/CSHL_RNAseq/mydoc/mydoc_systemPipeRNAseq_07/
#Get annotation info
library("biomaRt")
m <- useMart("ensembl", dataset="mmusculus_gene_ensembl")
```

```{r}


# Run the standard workflow for visualization and clustering
Verity_Prion_comp_astrocytes <- ScaleData(Verity_Prion_comp_astrocytes, verbose = FALSE)
Verity_Prion_comp_astrocytes <- RunPCA(Verity_Prion_comp_astrocytes, npcs = 30, verbose = FALSE)
ElbowPlot(Verity_Prion_comp_astrocytes, ndims=30)+ggtitle("Elbowplot of subset Verity astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Verity_Prion_comp_astrocytes <- RunUMAP(Verity_Prion_comp_astrocytes, reduction = "pca", dims = 1:20, seed.use=42)
Verity_Prion_comp_astrocytes <- FindNeighbors(Verity_Prion_comp_astrocytes, reduction = "pca", dims = 1:20)
Verity_Prion_comp_astrocytes <- FindClusters(Verity_Prion_comp_astrocytes, resolution = c(0.2,0.3,0.5,0.7,1))

#save for reloading
save(Verity_Prion_comp_astrocytes, file=paste0(plotting_path,"Verity_Prion_comp_astrocytes.Rdata"))

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Verity_Prion_comp_astrocytes, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Verity data's HP astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Verity_Prion_comp_astrocytes, group.by = 'RNA_snn_res.0.3', reduction = 'umap', repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Verity_Prion_comp_astrocytes, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Verity_Prion_comp_astrocytes,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.3
Verity_Prion_comp_astrocytes@active.ident<-(Verity_Prion_comp_astrocytes@meta.data$RNA_snn_res.0.3)
levels(Verity_Prion_comp_astrocytes)

DefaultAssay(Verity_Prion_comp_astrocytes)<-'RNA'

#by treatment, show the genes DE in Prion condition. LogFC set to a *1 increase in fold change
Verity_Prion_comp_astrocytes@active.ident<-as.factor(Verity_Prion_comp_astrocytes$treatment)
Verity_astrocyte_prion_markers<-FindMarkers(Verity_Prion_comp_astrocytes, ident.1 = 'ME7', ident.2 = 'NBH',min.pct = 0.10, verbose = TRUE, test.use = "DESeq2")

#Sort the list in descending log2FC and ascending pvalues
Verity_astrocyte_prion_markers<-Verity_astrocyte_prion_markers[order(-Verity_astrocyte_prion_markers$avg_log2FC), ]


#Use biomart info
Verity_astrocyte_prion_markers$gene<-rownames(Verity_astrocyte_prion_markers)

#check for duplicates
Verity_astrocyte_prion_markers[duplicated(Verity_astrocyte_prion_markers$gene),]

Verity_astrocyte_prion_markers<-Verity_astrocyte_prion_markers[!duplicated(Verity_astrocyte_prion_markers$gene),]

#make list of genes
gene_ids <- unique(Verity_astrocyte_prion_markers$gene)
  
#get relevant info from the mart
gene_info <- getBM(attributes = c("external_gene_name", "description","chromosome_name","entrezgene_id"), filters = "external_gene_name", values = gene_ids, mart = m)
  
#apply the info to the gene list
Verity_astrocyte_prion_markers <- merge(Verity_astrocyte_prion_markers, gene_info, by.x = "gene", by.y = "external_gene_name", all.x = TRUE)
  

#Save the file
write.csv(Verity_astrocyte_prion_markers, file = paste0(plotting_path,"Verity_astrocyte_prion_markers.csv"))

save(Verity_astrocyte_prion_markers, file=paste0(plotting_path,"Verity_astrocyte_prion_markers.RData"))

table(Verity_Prion_comp_astrocytes$treatment)
```

## Functional annotation

### Address the potential non-mapping gene entrezIDs, make a gene list
with fewer NAs

```{r}

#make list of gene names not mapping
not_mapped<-(Verity_astrocyte_prion_markers[is.na(Verity_astrocyte_prion_markers$entrezgene_id),]$gene)
not_mapped<-not_mapped[!duplicated(not_mapped)]

# Create an empty data frame to store the mapped ENTREZID values
remapped <- data.frame(ALIAS = character(0), ENTREZID = character(0), stringsAsFactors = FALSE)

# Loop through each gene in 'not_mapped'
for (gene in not_mapped) {
  # Attempt to perform the mapping using tryCatch
  mapped_genes <- tryCatch(
    {
      bitr(
        gene,
        fromType = "ALIAS",
        toType = "ENTREZID",
        OrgDb = org.Mm.eg.db,
        drop = FALSE
      )
    },
    error = function(e) {
      # If an error occurs, handle it by returning a data frame with NA values
      message(paste("Error mapping gene:", gene))
      return(data.frame(ALIAS = gene, ENTREZID = NA, stringsAsFactors = FALSE))
    }
  )

  # Append the temporary data frame to 'remapped'
  remapped <- rbind(remapped, mapped_genes)
  Sys.sleep(0.5)
}

#Additionally, get biomart conversions for those still without entrez ids 
biomart_remapped<-getBM(attributes = c("external_gene_name","entrezgene_id"), filters = "external_gene_name", values = remapped[is.na(remapped$ENTREZID),]$ALIAS, mart = m)

#add any more which have been mapped
biomart_remapped<-biomart_remapped[!is.na(biomart_remapped$entrezgene_id),]

#put everything recovered into a single df
bloated_mapped_df <- merge(remapped, biomart_remapped, by.x = "ALIAS", by.y = "external_gene_name", all.x = TRUE)
bloated_mapped_df$entrezgene_id <- as.character(bloated_mapped_df$entrezgene_id)

#make a single column for everything recovered, remove old columns
condensed_mapped <- bloated_mapped_df %>%
  mutate(entrezgene_id = coalesce(ENTREZID, entrezgene_id))
condensed_mapped$entrezgene_id<-as.integer(condensed_mapped$entrezgene_id)
#remove old columns
condensed_mapped$ENTREZID<-ENTREZID<- NULL


#The final list of newly gnerated entrezIDs, removed NAs
c<-condensed_mapped[!is.na(condensed_mapped$entrezgene_id),] 

#Add them back to the original list of markers, remove eztraneous columns
Verity_astrocyte_prion_markers <- merge(Verity_astrocyte_prion_markers, c, by.x = "gene", by.y = "ALIAS", all.x = TRUE)
Verity_astrocyte_prion_markers$entrezgene_id<-coalesce(Verity_astrocyte_prion_markers$entrezgene_id.x, Verity_astrocyte_prion_markers$entrezgene_id.y)
Verity_astrocyte_prion_markers<-subset(Verity_astrocyte_prion_markers, select = -c(entrezgene_id.x, entrezgene_id.y))
```

### Discard any genes which don't have an ENTREZID, generate list of
background DEGs

```{r}
#Discard any genes which don't have an ENTREZID
Verity_astrocyte_prion_markers_entrez<-Verity_astrocyte_prion_markers[!is.na(Verity_astrocyte_prion_markers$entrezgene_id),]


#create the genelist background for Verity Up
## feature 1: numeric vector
geneList_VP <- Verity_astrocyte_prion_markers_entrez$avg_log2FC
## feature 2: named vector of enembl IDs
names(geneList_VP) <-Verity_astrocyte_prion_markers_entrez$entrezgene_id
## feature 3: decreasing order
geneList_VP <- sort(geneList_VP, decreasing = TRUE) #object of type double
```

### Subset upregulated genes and perform ORA

```{r}
#Create upregulated significant genes
# Subset by Upregulated genes for GO enrichment
Verity_astrocyte_prion_markers_up<-subset(Verity_astrocyte_prion_markers_entrez, subset=avg_log2FC>(0.25)&p_val_adj<0.05) #Added p value filter here

#Create a list of just the gene symbols 
Verity_Prion_DEGs_up = (Verity_astrocyte_prion_markers_up$entrezgene_id)


# Run GO over-representation on upregulated entrezIds, biological process selected from GO terms
VerityPrion_up_GO_ORA = enrichGO(gene = Verity_Prion_DEGs_up, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)
```

#### GO ORA visualisation and saving

```{r}
ggp <- barplot(VerityPrion_up_GO_ORA , showCategory=20)
ggp2 <- dotplot(VerityPrion_up_GO_ORA, showCategory=20)
ggp3 <- goplot(VerityPrion_up_GO_ORA, showCategory = 20)
ggp4<- cnetplot(VerityPrion_up_GO_ORA, foldChange=geneList_VP)
ggp5<-heatplot(VerityPrion_up_GO_ORA, showCategory=20)#,foldChange=geneList_VP)


ggp
ggp2
ggp3
ggp4
ggp5

png(paste0(plotting_path,"Verity_Prion_GOR_barplot.png"), width = 1000,height = 750)
ggp
dev.off()
png(paste0(plotting_path,"Verity_Prion_GOR_dotplot.png"), width = 1000,height = 750)
ggp2
dev.off()
png(paste0(plotting_path,"Verity_Prion_GOR_goplot.png"), width = 1000,height = 750)
ggp3
dev.off()
png(paste0(plotting_path,"Verity_Prion_GOR_cnetplot.png"), width = 1000,height = 750)
ggp4
dev.off()
png(paste0(plotting_path,"Verity_Prion_GOR_heatplot.png"), width = 3000,height = 1000)
ggp5
dev.off()
```

### GO GSEA

```{r}
#Perform GO GSEA
VP_GO_GSEA <- gseGO(geneList= geneList_VP, #This is the DEG list full list of DEGs
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE,
              seed=42)

#convert back to Symbols for plotting
VP_GO_GSEA <- setReadable(VP_GO_GSEA, 'org.Mm.eg.db', 'ENTREZID')

View(data.frame(VP_GO_GSEA))

write.csv(data.frame(VP_GO_GSEA), file = paste0(plotting_path,"VP_GO_GSEA.csv"))
```

#### GO GSEA visualisation

```{r}
ggp6<-dotplotGsea(data = VP_GO_GSEA,topn = 10,
            order.by = 'NES')

ggp7<-cnetplot(VP_GO_GSEA, colorEdge = TRUE,foldChange=geneList_VP, showCategory=5)

ggp6
ggp7

#calculate pairwise similarity of GSEA genesets
VP_GO_GSEA_interactions <- pairwise_termsim(VP_GO_GSEA)
#interaction between genesets
ggp7_1<-emapplot(VP_GO_GSEA_interactions, color ='NES',layout='nicely')

ggp7_1

png(paste0(plotting_path,"VP_GO_GSEA_dotplot.png"), width = 1000,height = 750)
ggp6
dev.off()
png(paste0(plotting_path,"VP_GO_GSEA_cnetplot.png"), width = 2000,height = 2000)
ggp7
dev.off()
png(paste0(plotting_path,"VP_GO_GSEA_interactions.png"), width = 1500,height = 1500)
ggp7_1
dev.off()

```

### Reactome ORA and GSEA

```{r}

#Reactome over-representation
VP_Reactome_ORA <- enrichPathway(gene=Verity_astrocyte_prion_markers_up$entrezgene_id, pvalueCutoff = 0.05, readable=TRUE, organism = "mouse")

#Reactome GSEA calculation
VP_Reactome_GSEA <- gsePathway(geneList_VP, 
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH",
                organism = "mouse",
                verbose = FALSE,
                seed=42,
                maxGSSize = 100) ##FIXME trying a lower limit to remove broad terms

#convert back to symbols
VP_Reactome_GSEA <- setReadable(VP_Reactome_GSEA, 'org.Mm.eg.db', 'ENTREZID')
View(data.frame(VP_Reactome_GSEA))
write.csv(data.frame(VP_Reactome_GSEA), file = paste0(plotting_path,"VP_Reactome_GSEA.csv"))
```

#### Reactome GSEA visualisation

```{r}
#calculate pairwise similarity of GSEA genesets
VP_Reactome_GSEA_interactions <- pairwise_termsim(VP_Reactome_GSEA)

ggp8<-heatplot(VP_Reactome_GSEA, showCategory = 10, foldChange=geneList_VP)
ggp9<-cnetplot(VP_Reactome_GSEA, circular = TRUE, colorEdge = TRUE, foldChange=geneList_VP)
ggp10<-cnetplot(VP_Reactome_GSEA, colorEdge = TRUE, foldChange=geneList_VP, showCategory=10)
ggp11<-ridgeplot(VP_Reactome_GSEA,fill = 'NES', showCategory = 10)

#interaction between genesets
ggp12<-emapplot(VP_Reactome_GSEA_interactions, color ='NES',layout='nicely')

ggp8
ggp9
ggp10
ggp11
ggp12

png(paste0(plotting_path,"Verity_Prion_GSEA_heatplot_Reactome.png"), width = 3000,height = 1000)
ggp8
dev.off()
png(paste0(plotting_path,"Verity_Prion_GSEA_cnetplot_circular_Reactome.png"), width =2000,height = 2000)
ggp9
dev.off()
png(paste0(plotting_path,"Verity_Prion_GSEA_cnetplot_Reactome.png"), width =2000,height = 2000)
ggp10
dev.off()
png(paste0(plotting_path,"Verity_Prion_GSEA_ridgeplot_Reactome.png"), width = 1000,height = 750)
ggp11
dev.off()
png(paste0(plotting_path,"Verity_Prion_GSEA_interactions_plot_Reactome.png"), width = 1500,height = 1500)
ggp12
dev.off()

```

# Ageing

```{r}
#active assay RNA
DefaultAssay(Verity_Ageing_comp)<-"RNA"

#DimReduc again
#https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_SC_SCT_normalization.md
# Normalize the counts
Verity_Ageing_comp <- NormalizeData(Verity_Ageing_comp)
# Identify the most variable genes with defaults
Verity_Ageing_comp <- FindVariableFeatures(Verity_Ageing_comp, verbose = T)
		     
# Scale the counts
Verity_Ageing_comp <- ScaleData(Verity_Ageing_comp)
# Perform PCA
Verity_Ageing_comp <- RunPCA(Verity_Ageing_comp)
#Inspect the variability accounted for by the PCS
#It appears that 23 PCs are responsible, so using 30 from now on
ElbowPlot(Verity_Ageing_comp, ndims=30)+ggtitle("Elbowplot of Verity data", subtitle = "NBH and ME7")
DimHeatmap(Verity_Ageing_comp, dims=1:9, cells=500, balanced=T)
#Finish dimensinoality reduction steps
Verity_Ageing_comp <- RunUMAP(Verity_Ageing_comp, reduction = "pca", dims = 1:30, seed.use=42)
Verity_Ageing_comp <- FindNeighbors(Verity_Ageing_comp, reduction = "pca", dims = 1:30)
Verity_Ageing_comp <- FindClusters(Verity_Ageing_comp, resolution = c(0.2,0.3,0.5,0.7,1))
#Save the object
save(Verity_Ageing_comp,file=paste0(plotting_path,"Verity_Ageing_comp.Rdata"))


# Visualise the clustering resolutions for cluster number

# Determine metrics to plot present in Ver_singlets_merged_clean@meta.data
metrics <-  c("nUMI", "nGene", "mitoRatio")

#FIXME titles
DimPlot(Verity_Ageing_comp , reduction = "umap", group.by = "treatment", raster=FALSE)+ ggtitle("Verity Hippocampal cells labelled by treatment")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Verity_Ageing_comp, group.by = 'RNA_snn_res.0.2', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Verity Hippocampal cells labelled by cluster numbers", subtitle = "Clustering resolution 0.2")

#adapted from https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/08_SC_clustering_quality_control.md#segregation-of-clusters-by-various-sources-of-uninteresting-variation

#Explore any clustering mishaps by lookin for localisation of any of these traits in a given cluster
FeaturePlot(Verity_Ageing_comp, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            #min.cutoff = 'q10', #This was a default in the online script, I don't see the benefit of colouring less of the range of values
            label = TRUE)


## Save the singlets with a cluster resolution?
#FIXME might not be required
 #set the idents to _____
Verity_Ageing_comp@active.ident<-Verity_Ageing_comp@meta.data$RNA_snn_res.0.2
levels(Verity_Ageing_comp)

rm(metrics)
```

```{r}
DefaultAssay(Verity_Ageing_comp)<-"RNA"

#Check for a gene of interest in the data's geneset
Gene_Names<-Verity_Ageing_comp@assays[["RNA"]]@counts@Dimnames[[1]]
filtered_gene_names<-Gene_Names[!grepl("^\\d", Gene_Names)]
filtered_gene_names[grep("^Slc1a2", filtered_gene_names)]

#try to identify astrocytes  with genes
astrocyte_genes<-c("Aldh1l1","Gfap","Slc1a2", "Slc1a3", "Glul", "Cd44")

FeaturePlot(Verity_Ageing_comp,features=astrocyte_genes, raster=F)#+ggtitle("Verity data, labelled with astrocyte gene markers from ABCAM", subtitle="Clustering resolution 0.3, 20 PCs used")

#As a reminder...
DimPlot(Verity_Ageing_comp, group.by = 'RNA_snn_res.0.2', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Verity data, labelled by cluster numbers", subtitle="Clustering resolution 0.2")

#No compare to the plots of astrocyte genes
VlnPlot(Verity_Ageing_comp, features = astrocyte_genes, group.by = 'RNA_snn_res.0.2', pt.size = 0.0)

#Glul is maybe a good indicator?
```

## (SingleR annotation)

Could go here, could be followed by bubbleplot
<https://rdrr.io/github/milescsmith/seuratBubblePlot/man/bubbleplot.html>

Alternatively, load information from the spreadsheet at
Analysis/ABCAM_markers.xlsx then bubble plot

## Investigate astrocytes

```{r}
# Convert cluster names to character
Idents(Verity_Ageing_comp) <- Verity_Ageing_comp$RNA_snn_res.0.2

#Subset out astrocytes
Verity_Ageing_comp_astrocytes<-subset(Verity_Ageing_comp, idents=c('4','18'))
```

## DEG work

```{r}


# Run the standard workflow for visualization and clustering
Verity_Ageing_comp_astrocytes <- ScaleData(Verity_Ageing_comp_astrocytes, verbose = FALSE)
Verity_Ageing_comp_astrocytes <- RunPCA(Verity_Ageing_comp_astrocytes, npcs = 30, verbose = FALSE)
ElbowPlot(Verity_Ageing_comp_astrocytes, ndims=30)+ggtitle("Elbowplot of subset Verity astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Verity_Ageing_comp_astrocytes <- RunUMAP(Verity_Ageing_comp_astrocytes, reduction = "pca", dims = 1:20, seed.use=42)
Verity_Ageing_comp_astrocytes <- FindNeighbors(Verity_Ageing_comp_astrocytes, reduction = "pca", dims = 1:20)
Verity_Ageing_comp_astrocytes <- FindClusters(Verity_Ageing_comp_astrocytes, resolution = c(0.2,0.3,0.5,0.7,1))

#save for reloading
save(Verity_Ageing_comp_astrocytes, file=paste0(plotting_path,"Verity_Ageing_comp_astrocytes.Rdata"))

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Verity_Ageing_comp_astrocytes, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Verity data's CX astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Verity_Ageing_comp_astrocytes, group.by = 'RNA_snn_res.0.2', reduction = 'umap', repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Verity_Ageing_comp_astrocytes, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Verity_Ageing_comp_astrocytes,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.2
Verity_Ageing_comp_astrocytes@active.ident<-(Verity_Ageing_comp_astrocytes@meta.data$RNA_snn_res.0.2)
levels(Verity_Ageing_comp_astrocytes)

DefaultAssay(Verity_Ageing_comp_astrocytes)<-'RNA'

#by treatment, show the genes DE in Prion condition. LogFC set to a *1 increase in fold change
Verity_Ageing_comp_astrocytes@active.ident<-as.factor(Verity_Ageing_comp_astrocytes$treatment)
Verity_astrocyte_ageing_markers<-FindMarkers(Verity_Ageing_comp_astrocytes, ident.1 = 'AGEING', ident.2 = 'NBH',min.pct = 0.10, verbose = TRUE, test.use = "DESeq2")

#Sort the list in descending log2FC and ascending pvalues
Verity_astrocyte_ageing_markers<-Verity_astrocyte_ageing_markers[order(-Verity_astrocyte_ageing_markers$avg_log2FC), ]


#Use biomart info
Verity_astrocyte_ageing_markers$gene<-rownames(Verity_astrocyte_ageing_markers)

gene_ids <- unique(Verity_astrocyte_ageing_markers$gene)


#make list of genes
gene_ids <- unique(Verity_astrocyte_ageing_markers$gene)
  
#get relevant info from the mart
gene_info <- getBM(attributes = c("external_gene_name", "description","chromosome_name","entrezgene_id"), filters = "external_gene_name", values = gene_ids, mart = m)

  
#apply the info to the gene list
  Verity_astrocyte_ageing_markers <- merge(Verity_astrocyte_ageing_markers, gene_info, by.x = "gene", by.y = "external_gene_name", all.x = TRUE)
  

#Save the file
write.csv(Verity_astrocyte_ageing_markers, file = paste0(plotting_path,"Verity_astrocyte_ageing_markers.csv"))

save(Verity_astrocyte_ageing_markers, file=paste0(plotting_path,"Verity_astrocyte_ageing_markers.RData"))

table(Verity_Ageing_comp_astrocytes$treatment)
```

## Functional annotation

Ageing is not giving many genes with logfc2 \> 0.25, so functional
annotation not working for upregulated genes. Could do down?


### Address the potential non-mapping gene entrezIDs, make a gene list
with fewer NAs

```{r}

#make list of gene names not mapping
not_mapped<-(Verity_astrocyte_ageing_markers[is.na(Verity_astrocyte_ageing_markers$entrezgene_id),]$gene)
not_mapped<-not_mapped[!duplicated(not_mapped)]

# Create an empty data frame to store the mapped ENTREZID values
remapped <- data.frame(ALIAS = character(0), ENTREZID = character(0), stringsAsFactors = FALSE)

# Loop through each gene in 'not_mapped'
for (gene in not_mapped) {
  # Attempt to perform the mapping using tryCatch
  mapped_genes <- tryCatch(
    {
      bitr(
        gene,
        fromType = "ALIAS",
        toType = "ENTREZID",
        OrgDb = org.Mm.eg.db,
        drop = FALSE
      )
    },
    error = function(e) {
      # If an error occurs, handle it by returning a data frame with NA values
      message(paste("Error mapping gene:", gene))
      return(data.frame(ALIAS = gene, ENTREZID = NA, stringsAsFactors = FALSE))
    }
  )

  # Append the temporary data frame to 'remapped'
  remapped <- rbind(remapped, mapped_genes)
  Sys.sleep(0.5)
}

#Additionally, get biomart conversions for those still without entrez ids 
biomart_remapped<-getBM(attributes = c("external_gene_name","entrezgene_id"), filters = "external_gene_name", values = remapped[is.na(remapped$ENTREZID),]$ALIAS, mart = m)

#add any more which have been mapped
biomart_remapped<-biomart_remapped[!is.na(biomart_remapped$entrezgene_id),]

#put everything recovered into a single df
bloated_mapped_df <- merge(remapped, biomart_remapped, by.x = "ALIAS", by.y = "external_gene_name", all.x = TRUE)
bloated_mapped_df$entrezgene_id <- as.character(bloated_mapped_df$entrezgene_id)

#make a single column for everything recovered, remove old columns
condensed_mapped <- bloated_mapped_df %>%
  mutate(entrezgene_id = coalesce(ENTREZID, entrezgene_id))
condensed_mapped$entrezgene_id<-as.integer(condensed_mapped$entrezgene_id)
#remove old columns
condensed_mapped$ENTREZID<-ENTREZID<- NULL


#The final list of newly gnerated entrezIDs, removed NAs
c<-condensed_mapped[!is.na(condensed_mapped$entrezgene_id),] 

#Add them back to the original list of markers, remove eztraneous columns
Verity_astrocyte_ageing_markers <- merge(Verity_astrocyte_ageing_markers, c, by.x = "gene", by.y = "ALIAS", all.x = TRUE)
Verity_astrocyte_ageing_markers$entrezgene_id<-coalesce(Verity_astrocyte_ageing_markers$entrezgene_id.x, Verity_astrocyte_ageing_markers$entrezgene_id.y)
Verity_astrocyte_ageing_markers<-subset(Verity_astrocyte_ageing_markers, select = -c(entrezgene_id.x, entrezgene_id.y))
```

### Discard any genes which don't have an ENTREZID, generate list of
background DEGs

```{r}
#Discard any genes which don't have an ENTREZID
Verity_astrocyte_ageing_markers_entrez<-Verity_astrocyte_ageing_markers[!is.na(Verity_astrocyte_ageing_markers$entrezgene_id),]


#create the genelist background for Verity Up
## feature 1: numeric vector
geneList_VA <- Verity_astrocyte_ageing_markers_entrez$avg_log2FC
## feature 2: named vector of enembl IDs
names(geneList_VA) <-Verity_astrocyte_ageing_markers_entrez$entrezgene_id
## feature 3: decreasing order
geneList_VA <- sort(geneList_VA, decreasing = TRUE) #object of type double
```

### Subset upregulated genes and perform ORA

```{r}
#Create upregulated significant genes
# Subset by Upregulated genes for GO enrichment
Verity_astrocyte_ageing_markers_up<-subset(Verity_astrocyte_ageing_markers_entrez, subset=avg_log2FC>(0.25)&p_val_adj<0.05) #Added p value filter here

#Create a list of just the gene symbols 
Verity_Ageing_DEGs_up = (Verity_astrocyte_ageing_markers_up$entrezgene_id)


# Run GO over-representation on upregulated entrezIds, biological process selected from GO terms
VerityAgeing_up_GO_ORA = enrichGO(gene = Verity_Ageing_DEGs_up, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)
```

#### GO ORA visualisation and saving

```{r}
ggp <- barplot(VerityAgeing_up_GO_ORA , showCategory=20)
ggp2 <- dotplot(VerityAgeing_up_GO_ORA, showCategory=20)
ggp3 <- goplot(VerityAgeing_up_GO_ORA, showCategory = 20)
ggp4<- cnetplot(VerityAgeing_up_GO_ORA, foldChange=geneList_VA)#, foldChange=geneList_VA)
ggp5<-heatplot(VerityAgeing_up_GO_ORA, showCategory=20)#,foldChange=geneList_VA)


ggp
ggp2
ggp3
ggp4
ggp5

png(paste0(plotting_path,"Verity_Ageing_GOR_barplot.png"), width = 1000,height = 750)
ggp
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GOR_dotplot.png"), width = 1000,height = 750)
ggp2
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GOR_goplot.png"), width = 1000,height = 750)
ggp3
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GOR_cnetplot.png"), width = 1000,height = 750)
ggp4
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GOR_heatplot.png"), width = 1000,height = 750)
ggp5
dev.off()
```

### GO GSEA

```{r}
#Perform GO GSEA
VA_GO_GSEA <- gseGO(geneList= geneList_VA, #This is the DEG list full list of DEGs
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE,
              seed=42)

#convert back to Symbols for plotting
VA_GO_GSEA <- setReadable(VA_GO_GSEA, 'org.Mm.eg.db', 'ENTREZID')

View(data.frame(VA_GO_GSEA))

write.csv(data.frame(VA_GO_GSEA), file = paste0(plotting_path,"VA_GO_GSEA.csv"))
```

#### GO GSEA visualisation

```{r}
ggp6<-dotplotGsea(data = VA_GO_GSEA,topn = 10,
            order.by = 'NES')

ggp7<-cnetplot(VA_GO_GSEA, colorEdge = TRUE, circular = TRUE,foldChange=geneList_VA, showCategory=5)


#calculate pairwise similarity of GSEA genesets
VA_GO_GSEA_interactions <- pairwise_termsim(VA_GO_GSEA)
#interaction between genesets
ggp7_1<-emapplot(VP_Reactome_GSEA_interactions, color ='NES',layout='nicely')

ggp6
ggp7
ggp7_1

png(paste0(plotting_path,"VA_GO_GSEA_dotplot_Reactome.png"), width = 1000,height = 750)
ggp6
dev.off()
png(paste0(plotting_path,"VA_GO_GSEA_cnetplot_Reactome.png"), width = 2000,height = 2000)
ggp7
dev.off()
png(paste0(plotting_path,"VA_GO_GSEA_interactions_Reactome.png"), width = 1500,height = 1500)
ggp7_1
dev.off()

```

### Reactome ORA and GSEA

```{r}

#Reactome over-representation
VA_Reactome_ORA <- enrichPathway(gene=Verity_astrocyte_ageing_markers_up$entrezgene_id, pvalueCutoff = 0.05, readable=TRUE, organism = "mouse")

#Reactome GSEA calculation
VA_Reactome_GSEA <- gsePathway(geneList_VA, 
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH",
                organism = "mouse",
                verbose = FALSE,
                seed=42,
                maxGSSize = 100) ##FIXME trying a lower limit to remove broad terms

#convert back to symbols
VA_Reactome_GSEA <- setReadable(VA_Reactome_GSEA, 'org.Mm.eg.db', 'ENTREZID')
View(data.frame(VA_Reactome_GSEA))
write.csv(data.frame(VA_Reactome_GSEA), file = paste0(plotting_path,"VA_Reactome_GSEA.csv"))
```

#### Reactome GSEA visualisation

```{r}
#calculate pairwise similarity of GSEA genesets
VA_Reactome_GSEA_interactions <- pairwise_termsim(VA_Reactome_GSEA)

ggp8<-heatplot(VA_Reactome_GSEA, showCategory = 10, foldChange=geneList_VA)
ggp9<-cnetplot(VA_Reactome_GSEA, circular = TRUE, colorEdge = TRUE, foldChange=geneList_VA)
ggp10<-cnetplot(VA_Reactome_GSEA, colorEdge = TRUE, foldChange=geneList_VA, showCategory=10)
ggp11<-ridgeplot(VA_Reactome_GSEA,fill = 'NES', showCategory = 10)

#interaction between genesets
ggp12<-emapplot(VA_Reactome_GSEA_interactions, color ='NES',layout='nicely')

ggp8
ggp9
ggp10
ggp11
ggp12

png(paste0(plotting_path,"Verity_Ageing_GSEA_heatplot.png"), width = 1000,height = 750)
ggp8
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GSEA_cnetplot_circular.png"), width =2000,height = 2000)
ggp9
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GSEA_cnetplot.png"), width =2000,height = 2000)
ggp10
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GSEA_ridgeplot.png"), width = 1000,height = 750)
ggp11
dev.off()
png(paste0(plotting_path,"Verity_Ageing_GSEA_interactions_plot.png"), width = 1500,height = 1500)
ggp12
dev.off()

```