---
title: "Clean Slota workflow"
output: html_document
date: "2023-07-18"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


#Package management

```{r}
#.libPaths(new = "/home/s2268606/University_Directories/Project_23/R_packages_project")
#/opt/R/4.2.1/lib/R/library

update.packages(ask = FALSE)


install.packages("BiocManager")

BiocManager::install(c(
  "org.Mm.eg.db",
  "AnnotationDbi",
  "clusterProfiler",
  "tidyverse",
  'knitr',
  "DropletUtils",
  "Seurat",
  "SoupX",
  "scCustomize",
  "glmGamPoi",
  "DESeq2",
  "Polychrome",
  "ReactomePA"
  ))


remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')

remotes::install_github("satijalab/seurat", "seurat5")

devtools::install_github("junjunlab/GseaVis")

```

```{r}
library(org.Mm.eg.db)
library(clusterProfiler)
library(DropletUtils)
library(Seurat)
library(tidyverse)
library(dplyr)
library(stringr)
library(knitr)
library(DoubletFinder)
library(SoupX)
library(scCustomize)
library(DESeq2)
library(Polychrome)
library(GseaVis)
library(ReactomePA)
library(patchwork)
```

Make folder for plots

```{r}
#Specify the plotting folder to be made in the current working directory EDIT THIS ON EACH RUN TO AVOID OVERWRITING PLOTS
dir_name <- "Terminal_sample_investigation"

# create folder:
system(paste0("mkdir ", dir_name))

#save folder path for plot saving
plotting_path<-paste0("../Analysis/", dir_name,"/")

#set seed for reproducibility
set.seed(42)

rm(dir_name)
```

#Slota workflow

This workflow loads samples for the control and prion-treated data from
Slota group's hippocampal and cortex prion work. The samples are QC
assessed, filtered and plotted with clustering by gene expression
profiles.

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

# Load the data and format files for processing

```{bash}
# #CHATGPT assisted
# 
# #!/bin/bash
# 
# # Define the root directory
# root_dir="/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/SCP1962/other"
# 
# # Assuming the sample names are in a file called "Slota_samples.txt"
# sample_names_file="/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/Slota_samples.txt"
# 
# # Loop through each sample name in the file
# while IFS= read -r sample
# do
#     # Create a directory for the sample if it doesn't exist
#     mkdir -p "${root_dir}/${sample}"
# 
#     # Move the relevant files to the directory
#     mv "${root_dir}/${sample}"* "${root_dir}/${sample}"
# 
#     # Rename the files
#     mv "${root_dir}/${sample}/${sample}_barcodes.tsv" "${root_dir}/${sample}/barcodes.tsv"
#     mv "${root_dir}/${sample}/${sample}_features.tsv" "${root_dir}/${sample}/features.tsv"
#     mv "${root_dir}/${sample}/${sample}_matrix.mtx" "${root_dir}/${sample}/matrix.mtx"
# done < "$sample_names_file"
# 
# cd $root_dir
# gzip */{barcodes,features,matrix}*
```

```{r}
# Create list of folder names
sample_names<- readLines("/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/Slota_samples.txt")
#rename them for Slota folders


# Loop through each folder to process it
for (folder in sample_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper/SCP1962/other/",folder)
  
  
  current_data<-Read10X(data.dir =folder_path)
  current_obj<-CreateSeuratObject(counts=current_data, project=folder)
  assign(paste0(folder), current_obj)
  
}

Slota_raw_obj_clean<-merge(PBS25HP, y=c(PBS48CX,PBS48HP,PBS60CX,PBS61CX,PBS73CX,RML122CX,RML122HP,RML132CX,RML132HP,RML133CX,RML133HP,RML134CX,RML134HP,RML138CX,RML138HP,RML140CX,RML140HP,RML142CX,RML142HP,RML145CX,RML145HP))

rm(current_data,current_obj, folder, sample_names)
```

# Tailor metadata

## Treatment annotation

adapted from <https://youtu.be/p49seH2_i8Y?t=312>

```{r}
Slota_raw_obj_clean$sample<-Slota_raw_obj_clean$orig.ident
#Dealing with new merged object 
Slota_raw_obj_clean@active.ident<-as.factor(Slota_raw_obj_clean$sample)

#Add metadata on treatment, adapted from Chat GPT

#Slota
sampleLabs_Slota<- (Slota_raw_obj_clean@active.ident)

treat_detect_Slota <- case_when(
  str_detect(sampleLabs_Slota, "RML.*HP*") ~ "RML_HP",
  str_detect(sampleLabs_Slota, "PBS.*HP*") ~ "PBS_HP",
  str_detect(sampleLabs_Slota, "RML.*CX*") ~ "RML_CX",
  str_detect(sampleLabs_Slota, "PBS.*CX*") ~ "PBS_CX",
  TRUE ~ NA_character_
)

Slota_raw_obj_clean@meta.data$treatment<-treat_detect_Slota
rm(treat_detect_Slota, sampleLabs_Slota)

```

## Identify mitochondrial and ribosomal populations

```{r}
#identify mitochondrial and ribosomal protein gene portions, annotations syntax tailored to mouse
Slota_raw_obj_clean[["percent.mt"]]<- PercentageFeatureSet(Slota_raw_obj_clean, pattern = "^Mt")
Slota_raw_obj_clean[["percent.rp"]]<- PercentageFeatureSet(Slota_raw_obj_clean, pattern = "^Rp")
Slota_raw_obj_clean[["percent.hb"]]<-PercentageFeatureSet(Slota_raw_obj_clean, "^Hb[^(p)]")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Slota_raw_obj_clean[["log10GenesPerUMI"]] <- log10(Slota_raw_obj_clean$nFeature_RNA) / log10(Slota_raw_obj_clean$nCount_RNA)

Slota_raw_obj_clean[["mitoRatio"]] <- Slota_raw_obj_clean@meta.data$percent.mt / 100
```

##Initial data assessment

```{r}
p1<-VlnPlot(Slota_raw_obj_clean, features=c("nFeature_RNA","nCount_RNA","percent.mt", "percent.rp","percent.hb"), ncol=5, pt.size = 0.0)#+ggtitle("Preliminary Data assessment", subtitle= "Slota, unfiltered")

# Plot each of these respectively, showing trends in cell health

#show mitochondrial proportion in reads
p2<-FeatureScatter(Slota_raw_obj_clean, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0)+ggtitle("Slota_UMIs per cell vs.mitochondrial genes detected", subtitle="UMIs per cell vs. % mitochondrial genes detected")


#show ribosomal proportion in reads
p3<-FeatureScatter(Slota_raw_obj_clean, feature1 = "nCount_RNA", feature2 = "percent.rp")+ggplot2::ggtitle("Slota_UMIs_per_cell_vsribosomal_protein_genes_detected", subtitle = "UMIs per cell vs. % ribosomal protein  genes detected")


#Ribosomal vs mitochondrial 
p4<-FeatureScatter(Slota_raw_obj_clean, feature1 = "percent.rp", feature2 = "percent.mt")+ggplot2::ggtitle("Slota_ribosomal_protein_genes_detected_vsmitochondrial_genes_detected",subtitle="% ribosomal protein  genes detected vs. % mitochondrial genes detected")

#Show overall reads and gene counts for unfiltered data.
p5<-FeatureScatter(Slota_raw_obj_clean, feature1 = "nCount_RNA", feature2="nFeature_RNA",raster=F, pt.size = 0.5)+ geom_smooth(method="lm")+ ggtitle("Slota_UMIs_per_cell_vs_number_of_genes",subtitle="UMIs per cell vs. number of genes ")

p1;p2;p3;p4;p5

#Save plots
png(paste0(plotting_path,"Terminal_EDA.png"), width = 3000,height = 1000)
p1
dev.off()
png(paste0(plotting_path,"Terminal_UMI_MT.png"), width = 1750,height = 1500)
p2
dev.off()
png(paste0(plotting_path,"Terminal_UMI_RP.png"),width = 1750,height = 1500)
p3
dev.off()
png(paste0(plotting_path,"Terminal_RP_MT.png"),  width = 1750,height = 1500)
p4
dev.off()
png(paste0(plotting_path,"Terminal_UMI_nGene.png"), width = 1750,height = 1500)
p5
dev.off()

```

Low mitochondrial counts, good overall quality indicating prefilters?

### Save the whole data and tidy up

```{r}
#Save the data object
save(Slota_raw_obj_clean, file="./Rdata/Slota_raw_obj_clean.RData")

rm(PBS25HP,PBS48CX,PBS48HP,PBS60CX,PBS61CX,PBS73CX,RML122CX,RML122HP,RML132CX,RML132HP,RML133CX,RML133HP,RML134CX,RML134HP,RML138CX,RML138HP,RML140CX,RML140HP,RML142CX,RML142HP,RML145CX,RML145HP)

gc()
```

#QC

##Exploration

### Create metadata variable to edit it without affecting core data

```{r}
# Rename columns from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md

#Create metadata variable to edit it without affecting core data
metadata_test<-Slota_raw_obj_clean@meta.data

#Saving cell barcodes as metainfo
metadata_test$cells<-rownames(metadata_test)

#Setting minimum feature count to filter cells with no genes UMIs vs. genes detected, as these will be discarded anyway
metadata_test <- subset(Slota_raw_obj_clean, subset= nFeature_RNA > 20)@meta.data

metadata_test <- metadata_test %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)
```

### Plots

```{r}
# Visualize the number of cell counts per sample with no filtering at all
p6<-metadata_test %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota_Cell_counts_per_sample", subtitle="Filtered to minimum 20 genes")


# Visualize the number UMIs/transcripts per cell
p7<-metadata_test %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Slota/UMI_transcripts_per_cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via histogram
p8<-metadata_test %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="Filtered to minimum 20 genes")


# Visualize the distribution of genes detected per cell via boxplot
p9<-metadata_test %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="Filtered to minimum 20 genes")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p10<-metadata_test %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Slota UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle = "Filtered to minimum 20 genes")

# Visualize the distribution of mitochondrial gene expression detected per cell
p11<-metadata_test %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Slota distribution of mitochondrial gene expression per cell", subtitle = "Filtered to minimum 20 genes")

p6;p7;p8;p9;p10;p11

#Save plots
png(paste0(plotting_path,"Terminal_PreF_Counts_sample.png"), width = 2000,height = 1500)
p6
dev.off()
png(paste0(plotting_path,"Terminal_PreF_UMI_cell.png"),width = 2000,height = 1500)
p7
dev.off()
png(paste0(plotting_path,"Terminal_PreF_nGene_hist_sample.png"),width = 2000,height = 1500)
p8
dev.off()
png(paste0(plotting_path,"Terminal_PreF_nGene_hist_sample.png"),width = 2000,height = 1500)
p9
dev.off()
png(paste0(plotting_path,"Terminal_PreF_nGene_UMI_sample.png"),width = 2000,height = 1500)
p10
dev.off()
png(paste0(plotting_path,"Terminal_PreF_MT_cell.png"),width = 2000,height = 1500)
p11
dev.off()
```

## Filtering metadata

```{r}

#running author's filters

metadata_filtered<- metadata_test %>%
    filter(nGene > 1000 & percent.mt < 20 & percent.rp > 1 & percent.hb < 20 & nUMI < 35000) #FIXME check upper limit on nUMI
```

##Repeat QC plotting

```{r}
# Visualize the number of cell counts per sample with no filtering at all
p12<-metadata_filtered %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota_Cell_counts_per_sample", subtitle="nGene>150, nUMI>0, % mito< 20")


# Visualize the number UMIs/transcripts per cell
p13<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("Slota_UMI_transcripts_per_cell", subtitle="nGene>150, nUMI>0, % mito< 20")


# Visualize the distribution of genes detected per cell via histogram
p14<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="nGene>150, nUMI>0, % mito< 20")


# Visualize the distribution of genes detected per cell via boxplot
p15<-metadata_filtered %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="nGene>150, nUMI>0, % mito< 20")

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p16<-metadata_filtered %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("Slota UMIs vs. genes detected. Highlighting mitochondrial ratio in genes", subtitle="nGene>150, nUMI>0, % mito< 20")

# Visualize the distribution of mitochondrial gene expression detected per cell
p17<-metadata_filtered %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Slota distribution of mitochondrial gene expression per cell", subtitle="nGene>150, nUMI>0, % mito< 20")

p12;p13;p14;p15;p16;p17

#Save plots
png(paste0(plotting_path,"Terminal_PostF_Counts_sample.png"), width = 2000,height = 1500)
p12
dev.off()
png(paste0(plotting_path,"Terminal_PostF_UMI_cell.png"),width = 2000,height = 1500)
p13
dev.off()
png(paste0(plotting_path,"Terminal_PostF_nGene_hist_sample.png"),width = 2000,height = 1500)
p14
dev.off()
png(paste0(plotting_path,"Terminal_PostF_nGene_hist_sample.png"),width = 2000,height = 1500)
p15
dev.off()
png(paste0(plotting_path,"Terminal_PostF_nGene_UMI_sample.png"),width = 2000,height = 1500)
p16
dev.off()
png(paste0(plotting_path,"Terminal_PostF_MT_cell.png"),width = 2000,height = 1500)
p17
dev.off()
```

## Assign filters to data, removing three poor quality samples

```{r}
#Assign the unfiltered metadata to Slota, updating the column names
Slota_raw_obj_clean@meta.data<-metadata_test 

Slota_filtered_clean<- subset(Slota_raw_obj_clean,subset= 
                              nGene > 1000 & 
                              percent.mt < 20 &
                              percent.rp > 1 &
                              percent.hb < 20 &
                              nUMI < 35000 &
                              sample !="PBS73CX" &
                              sample !="RML133HP" &
                              sample !="RML134HP" 
                            )




# Gene-level filtering
# Keeping genes which appear in more than 3 cells for statistical power.

# Output a logical vector for every gene on whether the more than zero counts per cell
# Extract counts
counts <- GetAssayData(object = Slota_filtered_clean, slot = "counts")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 3 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 3

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
Slota_filtered_clean <- CreateSeuratObject(filtered_counts, meta.data = Slota_filtered_clean@meta.data)


p17a<-Slota_filtered_clean@meta.data %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Slota_Cell_counts_per_sample", subtitle="nGene>1000, nUMI>0<35,000, % mito< 20, % rp>1, % hb <20")

p17b<-Slota_filtered_clean@meta.data %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Slota distribution of genes per cell", subtitle="nGene>1000, nUMI>0<35,000, % mito< 20, % rp>1, % hb <20")

p17a;p17b

#rm(metadata_test, metadata_filtered,filtered_counts, nonzero,counts, keep_genes,Slota_raw_obj_clean)
gc()
```

### Saving object

```{r}
# Create .RData object to load at any time
save(Slota_filtered_clean, file="./Rdata/Slota_filtered_clean.RData")
```


# Splitting the data by tissue

```{r}
Slota_HP<-subset(Slota_filtered_clean, subset=treatment=='PBS_HP' | treatment=='RML_HP')
Slota_CX<-subset(Slota_filtered_clean, subset=treatment=='PBS_CX' | treatment=='RML_CX')

Slota_tissues<-list(Slota_HP,Slota_CX)

rm(Slota_HP,Slota_CX)
```



# Doublet finder for tissue groups separately

```{r}

#load in the filtered Slota object
#load("./Rdata/Slota_filtered_clean.RData")

#adapted from https://youtu.be/p49seH2_i8Y , https://github.com/kpatel427/YouTubeTutorials/blob/5b3c795cecafb22f0e992da069eae2efd6b5cb95/singleCell_doublets.R , https://rpubs.com/kenneditodd/doublet_finder_example 

#DoubletFinder, keep singlets
Slota_tissues <- lapply(Slota_tissues, function(tissue_group) {
  #Normalise+Scale with SCTransform?
  tissue_group <- SCTransform(tissue_group)
  tissue_group <- RunPCA(tissue_group)
  
  # Find significant PCs
  stdv <- tissue_group[["pca"]]@stdev
  sum.stdv <- sum(tissue_group[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] -
                       percent.stdv[2:length(percent.stdv)]) > 0.1),
              decreasing = TRUE)[1] + 1
  min.pc <- min(co1, co2)
  
  # Finish pre-processing
  tissue_group <- RunUMAP(tissue_group, dims = 1:min.pc, seed.use=42 )
  tissue_group <- FindNeighbors(object = tissue_group, dims = 1:min.pc)              
  tissue_group <- FindClusters(object = tissue_group, resolution = 0.1)
  
  # pK identification (no ground-truth)
  sweep.list <- paramSweep_v3(tissue_group, PCs = 1:min.pc, sct = TRUE)
  sweep.stats <- summarizeSweep(sweep.list)
  bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficient (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- as.numeric(bcmvn.max$pK)

  # Homotypic doublet proportion estimate
  annotations <- tissue_group@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.poi <- round(0.087 * nrow(tissue_group@meta.data)) #expected doublets based on values from 10X resources
  nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # Run DoubletFinder
  tissue_group <- doubletFinder_v3(seu = tissue_group, 
                             PCs = 1:min.pc, 
                             pK = optimal.pk,
                             nExp = nExp.poi.adj, 
                             sct = TRUE, 
                             pN=0.25)
  return(tissue_group)
  })
```

## Split the data, rename doubletfinder metadata and Show Doublets for each comparison group witihn Slota data. Retain singlets

```{r}

#Rename the columns
Slota_tissues <- lapply(Slota_tissues, function(comp_pair) {
  DefaultAssay(comp_pair)<-'RNA'
  doublet_col_index <- which(grepl("DF.classifications_", colnames(comp_pair@meta.data)))
  comp_pair$doublet_finder <- comp_pair@meta.data[doublet_col_index]
  comp_pair@meta.data[doublet_col_index]<-NULL

  #Show how many doublets there were
  table(comp_pair$doublet_finder)
  
  return(comp_pair)
})

#Split the data
Slota_HP_comp<-Slota_tissues[[1]]
Slota_CX_comp<-Slota_tissues[[2]]

#Plot how many doublets there were
p18<-VlnPlot(Slota_HP_comp, features = "nFeature_RNA", group.by = 'doublet_finder', pt.size = 0.0)#+ggtitle(paste0("Slota data", comp_pair))

p19<-VlnPlot(Slota_CX_comp, features = "nFeature_RNA", group.by = 'doublet_finder', pt.size = 0.0)#+ggtitle(paste0("Slota data", comp_pair))


png("Slota_HP_doublets.png", width = 2000,height = 1500)
p18
dev.off()

png("Slota_CX_doublets.png", width = 2000,height = 1500)
p19
dev.off()


#Save the data with Doublets retained too
save(Slota_tissues, file=paste0(plotting_path,"Slota_tissues.RData"))

#subset data by singlet status and save list of singlets to continue with
#FIXME might need to be in loop or explicit if subsetting doesn't work
Slota_CX_comp <- subset(Slota_CX_comp, subset= doublet_finder=='Singlet')
Slota_HP_comp <- subset(Slota_HP_comp, subset= doublet_finder=='Singlet')


gc()
``` 

# HP work
```{r}
#active assay RNA
DefaultAssay(Slota_HP_comp)<-"RNA"


#DimReduc again
#https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_SC_SCT_normalization.md
# Normalize the counts
Slota_HP_comp <- NormalizeData(Slota_HP_comp)
# Identify the most variable genes with defaults
Slota_HP_comp <- FindVariableFeatures(Slota_HP_comp, verbose = T)
		     
# Scale the counts
Slota_HP_comp <- ScaleData(Slota_HP_comp)
# Perform PCA
Slota_HP_comp <- RunPCA(Slota_HP_comp)
#Inspect the variability accounted for by the PCS
#It appears that 23 PCs are responsible, so using 30 from now on
ElbowPlot(Slota_HP_comp, ndims=30)+ggtitle("Elbowplot of Slota data", subtitle = "NBH and ME7")
DimHeatmap(Slota_HP_comp, dims=1:9, cells=500, balanced=T)
#Finish dimensinoality reduction steps
Slota_HP_comp <- RunUMAP(Slota_HP_comp, reduction = "pca", dims = 1:30, seed.use=42)
Slota_HP_comp <- FindNeighbors(Slota_HP_comp, reduction = "pca", dims = 1:30)
Slota_HP_comp <- FindClusters(Slota_HP_comp, resolution = c(0.2,0.3,0.5,0.7,1))
#Save the object
save(Slota_HP_comp,file=paste0(plotting_path,"Slota_HP_comp.Rdata"))


# Visualise the clustering resolutions for cluster number

# Determine metrics to plot present in Ver_singlets_merged_clean@meta.data
metrics <-  c("nUMI", "nGene", "mitoRatio")

#FIXME titles
DimPlot(Slota_HP_comp , reduction = "umap", group.by = "treatment", raster=FALSE)+ ggtitle("Slota Hippocampal cells labelled by treatment")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slota_HP_comp, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Slota Hippocampal cells labelled by cluster numbers")

#adapted from https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/08_SC_clustering_quality_control.md#segregation-of-clusters-by-various-sources-of-uninteresting-variation

#Explore any clustering mishaps by lookin for localisation of any of these traits in a given cluster
FeaturePlot(Slota_HP_comp, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            #min.cutoff = 'q10', #This was a default in the online script, I don't see the benefit of colouring less of the range of values
            label = TRUE)


## Save the singlets with a cluster resolution?
#FIXME might not be required
 #set the idents to _____
Slota_HP_comp@active.ident<-comp_pair@meta.data$RNA_snn_res.0.3
levels(Slota_HP_comp)

rm(metrics)
```

```{r}
DefaultAssay(Slota_HP_comp)<-"RNA"

#Check for a gene of interest in the data's geneset
Gene_Names<-Slota_HP_comp@assays[["RNA"]]@counts@Dimnames[[1]]
filtered_gene_names<-Gene_Names[!grepl("^\\d", Gene_Names)]
filtered_gene_names[grep("^Slc1a2", filtered_gene_names)]

#try to identify astrocytes  with genes
astrocyte_genes<-c("Aldh1l1","Gfap","Slc1a2", "Slc1a3", "Glul", "Cd44")

FeaturePlot(Slota_HP_comp,features=astrocyte_genes, raster=F)#+ggtitle("Slota data, labelled with astrocyte gene markers from ABCAM", subtitle="Clustering resolution 0.3, 20 PCs used")

#As a reminder...
DimPlot(Slota_HP_comp, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Slota data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#No compare to the plots of astrocyte genes
VlnPlot(Slota_HP_comp, features = astrocyte_genes, group.by = 'RNA_snn_res.0.3', pt.size = 0.0)

#Glul is maybe a good indicator?
```



## (SingleR annotation)

Could go here, could be followed by bubbleplot
<https://rdrr.io/github/milescsmith/seuratBubblePlot/man/bubbleplot.html>

Alternatively, load information from the spreadsheet at
Analysis/ABCAM_markers.xlsx then bubble plot

##Investigate astrocytes, ensure identification

```{r}
# Convert cluster names to character
Idents(Slota_HP_comp) <- Slota_HP_comp$RNA_snn_res.0.3

#Subset out astrocytes
Slota_HP_comp_astrocytes<-subset(Slota_HP_comp, idents='4')
```

##DEG work

Prepare the biomart info

```{r}
#Adapted from http://girke.bioinformatics.ucr.edu/CSHL_RNAseq/mydoc/mydoc_systemPipeRNAseq_07/
#Get annotation info
library("biomaRt")
m <- useMart("ensembl", dataset="mmusculus_gene_ensembl")
```



```{r}


# Run the standard workflow for visualization and clustering
Slota_HP_comp_astrocytes <- ScaleData(Slota_HP_comp_astrocytes, verbose = FALSE)
Slota_HP_comp_astrocytes <- RunPCA(Slota_HP_comp_astrocytes, npcs = 30, verbose = FALSE)
ElbowPlot(Slota_HP_comp_astrocytes, ndims=30)+ggtitle("Elbowplot of subset Slota astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Slota_HP_comp_astrocytes <- RunUMAP(Slota_HP_comp_astrocytes, reduction = "pca", dims = 1:20, seed.use=42)
Slota_HP_comp_astrocytes <- FindNeighbors(Slota_HP_comp_astrocytes, reduction = "pca", dims = 1:20)
Slota_HP_comp_astrocytes <- FindClusters(Slota_HP_comp_astrocytes, resolution = c(0.2,0.3,0.5,0.7,1))

#save for reloading
save(Slota_HP_comp_astrocytes, file=paste0(plotting_path,"Slota_HP_comp_astrocytes.Rdata"))

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slota_HP_comp_astrocytes, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Slota data's HP astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Slota_HP_comp_astrocytes, group.by = 'RNA_snn_res.0.2', reduction = 'umap', repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Slota_HP_comp_astrocytes, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Slota_HP_comp_astrocytes,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.2
Slota_HP_comp_astrocytes@active.ident<-(Slota_HP_comp_astrocytes@meta.data$RNA_snn_res.0.2)
levels(Slota_HP_comp_astrocytes)

DefaultAssay(Slota_HP_comp_astrocytes)<-'RNA'

#by treatment, show the genes DE in Prion condition.
Slota_HP_comp_astrocytes@active.ident<-as.factor(Slota_HP_comp_astrocytes$treatment)
Slota_astrocyte_prion_markers_HP<-FindMarkers(Slota_HP_comp_astrocytes, ident.1 = 'RML_HP', ident.2 = 'PBS_HP',min.pct = 0.10, verbose = TRUE, test.use = "DESeq2")

#Sort the list in descending log2FC and ascending pvalues
Slota_astrocyte_prion_markers_clean_HP<-Slota_astrocyte_prion_markers_HP[order(-Slota_astrocyte_prion_markers_HP$avg_log2FC), ]


#Use biomart info
Slota_astrocyte_prion_markers_clean_HP$gene<-rownames(Slota_astrocyte_prion_markers_clean_HP)

gene_ids <- unique(Slota_astrocyte_prion_markers_clean_HP$gene)
  
#get relevant info from the mart
gene_info <- getBM(attributes = c("external_gene_name", "description","chromosome_name","entrezgene_id"), filters = "external_gene_name", values = gene_ids, mart = m)
  
#apply the info to the gene list
Slota_astrocyte_prion_markers_clean_HP <- merge(Slota_astrocyte_prion_markers_clean_HP, gene_info, by.x = "gene", by.y = "external_gene_name", all.x = TRUE)
  

#Save the file
write.csv(Slota_astrocyte_prion_markers_clean_HP, file = paste0(plotting_path,"Slota_astrocyte_prion_markers_clean_HP.csv"))

save(Slota_astrocyte_prion_markers_clean_HP, file=paste0(plotting_path,"Slota_astrocyte_prion_markers_clean_HP.RData"))

table(Slota_HP_comp_astrocytes$treatment)
```


## Functional annotation

### Address the potential non-mapping gene entrezIDs, make a gene list
with fewer NAs

```{r}

#make list of gene names not mapping
not_mapped<-(Slota_astrocyte_prion_markers_clean_HP[is.na(Slota_astrocyte_prion_markers_clean_HP$entrezgene_id),]$gene)
not_mapped<-not_mapped[!duplicated(not_mapped)]

# Create an empty data frame to store the mapped ENTREZID values
remapped <- data.frame(ALIAS = character(0), ENTREZID = character(0), stringsAsFactors = FALSE)

# Loop through each gene in 'not_mapped'
for (gene in not_mapped) {
  # Attempt to perform the mapping using tryCatch
  mapped_genes <- tryCatch(
    {
      bitr(
        gene,
        fromType = "ALIAS",
        toType = "ENTREZID",
        OrgDb = org.Mm.eg.db,
        drop = FALSE
      )
    },
    error = function(e) {
      # If an error occurs, handle it by returning a data frame with NA values
      message(paste("Error mapping gene:", gene))
      return(data.frame(ALIAS = gene, ENTREZID = NA, stringsAsFactors = FALSE))
    }
  )

  # Append the temporary data frame to 'remapped'
  remapped <- rbind(remapped, mapped_genes)
  Sys.sleep(0.5)
}

#Additionally, get biomart conversions for those still without entrez ids 
biomart_remapped<-getBM(attributes = c("external_gene_name","entrezgene_id"), filters = "external_gene_name", values = remapped[is.na(remapped$ENTREZID),]$ALIAS, mart = m)

#add any more which have been mapped
biomart_remapped<-biomart_remapped[!is.na(biomart_remapped$entrezgene_id),]

#put everything recovered into a single df
bloated_mapped_df <- merge(remapped, biomart_remapped, by.x = "ALIAS", by.y = "external_gene_name", all.x = TRUE)
bloated_mapped_df$entrezgene_id <- as.character(bloated_mapped_df$entrezgene_id)

#make a single column for everything recovered, remove old columns
condensed_mapped <- bloated_mapped_df %>%
  mutate(entrezgene_id = coalesce(ENTREZID, entrezgene_id))
condensed_mapped$entrezgene_id<-as.integer(condensed_mapped$entrezgene_id)
#remove old columns
condensed_mapped$ENTREZID<-ENTREZID<- NULL


#The final list of newly gnerated entrezIDs, removed NAs
c<-condensed_mapped[!is.na(condensed_mapped$entrezgene_id),] 

#Add them back to the original list of markers, remove eztraneous columns
Slota_astrocyte_prion_markers_clean_HP <- merge(Slota_astrocyte_prion_markers_clean_HP, c, by.x = "gene", by.y = "ALIAS", all.x = TRUE)
Slota_astrocyte_prion_markers_clean_HP$entrezgene_id<-coalesce(Slota_astrocyte_prion_markers_clean_HP$entrezgene_id.x, Slota_astrocyte_prion_markers_clean_HP$entrezgene_id.y)
Slota_astrocyte_prion_markers_clean_HP<-subset(Slota_astrocyte_prion_markers_clean_HP, select = -c(entrezgene_id.x, entrezgene_id.y))
```

### Discard any genes which don't have an ENTREZID, generate list of
background DEGs

```{r}
#Discard any genes which don't have an ENTREZID
Slota_astrocyte_HP_markers_entrez<-Slota_astrocyte_prion_markers_clean_HP[!is.na(Slota_astrocyte_prion_markers_clean_HP$entrezgene_id),]


#create the genelist background for Slota Up
## feature 1: numeric vector
geneList_SHp <- Slota_astrocyte_HP_markers_entrez$avg_log2FC
## feature 2: named vector of enembl IDs
names(geneList_SHp) <-Slota_astrocyte_HP_markers_entrez$entrezgene_id
## feature 3: decreasing order
geneList_SHp <- sort(geneList_SHp, decreasing = TRUE) #object of type double
```

### Subset upregulated genes and perform ORA

```{r}
#Create upregulated significant genes
# Subset by Upregulated genes for GO enrichment
Slota_astrocyte_prion_markers_up<-subset(Slota_astrocyte_HP_markers_entrez, subset=avg_log2FC>(0.25)&p_val_adj<0.05) #Added p value filter here

#Create a list of just the gene symbols 
Slota_Prion_DEGs_up = (Slota_astrocyte_prion_markers_up$entrezgene_id)


# Run GO over-representation on upregulated entrezIds, biological process selected from GO terms
Slota_Hp_up_GO_ORA = enrichGO(gene = Slota_Prion_DEGs_up, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)
```

#### GO ORA visualisation and saving

```{r}
ggp <- barplot(Slota_Hp_up_GO_ORA , showCategory=20)
ggp2 <- dotplot(Slota_Hp_up_GO_ORA, showCategory=20)
ggp3 <- goplot(Slota_Hp_up_GO_ORA, showCategory = 20)
ggp4<- cnetplot(Slota_Hp_up_GO_ORA, foldChange=geneList_SHp)
ggp5<-heatplot(Slota_Hp_up_GO_ORA, showCategory=20)#,foldChange=geneList_SHp)


ggp
ggp2
ggp3
ggp4
ggp5

png(paste0(plotting_path,"Slota_Prion_GOR_barplot.png"), width = 1000,height = 750)
ggp
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_dotplot.png"), width = 1000,height = 750)
ggp2
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_goplot.png"), width = 1000,height = 750)
ggp3
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_cnetplot.png"), width = 1000,height = 750)
ggp4
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_heatplot.png"), width = 3000,height = 1000)
ggp5
dev.off()
```

####Same for downregulated

```{r}

#Create upregulated significant genes
# Subset by Upregulated genes for GO enrichment
Slota_astrocyte_prion_markers_down<-subset(Slota_astrocyte_HP_markers_entrez, subset=avg_log2FC<(-0.25)&p_val_adj<0.05) #Added p value filter here

#Create a list of just the gene symbols 
Slota_Prion_DEGs_down = (Slota_astrocyte_prion_markers_down$entrezgene_id)


# Run GO over-representation on upregulated entrezIds, biological process selected from GO terms
Slota_Hp_down_GO_ORA = enrichGO(gene = Slota_Prion_DEGs_down, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)

ggp_b <- barplot(Slota_Hp_down_GO_ORA , showCategory=20)
ggp2_b <- dotplot(Slota_Hp_down_GO_ORA, showCategory=20)
ggp3_b <- goplot(Slota_Hp_down_GO_ORA, showCategory = 20)
ggp4_b<- cnetplot(Slota_Hp_down_GO_ORA, foldChange=geneList_SHp)
ggp5_b<-heatplot(Slota_Hp_down_GO_ORA, showCategory=20,foldChange=geneList_SHp)


ggp_b
ggp2_b
ggp3_b
ggp4_b
ggp5_b

png(paste0(plotting_path,"Slota_Prion_GOR_down_barplot.png"), width = 1000,height = 750)
ggp_b
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_down_dotplot.png"), width = 1000,height = 750)
ggp2_b
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_down_goplot.png"), width = 1000,height = 750)
ggp3_b
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_down_cnetplot.png"), width = 1500,height = 1000)
ggp4_b
dev.off()
png(paste0(plotting_path,"Slota_Prion_GOR_down_heatplot.png"), width = 3600,height = 750)
ggp5_b
dev.off()
```



### GO GSEA

```{r}
#Perform GO GSEA
SHp_GO_GSEA <- gseGO(geneList= geneList_SHp, #This is the DEG list full list of DEGs
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE,
              seed=42)

#convert back to Symbols for plotting
SHp_GO_GSEA <- setReadable(SHp_GO_GSEA, 'org.Mm.eg.db', 'ENTREZID')

View(data.frame(SHp_GO_GSEA))

write.csv(data.frame(SHp_GO_GSEA), file = paste0(plotting_path,"SHp_GO_GSEA.csv"))
```

#### GO GSEA visualisation

```{r}
ggp6<-dotplotGsea(data = SHp_GO_GSEA,topn = 10,
            order.by = 'NES')

ggp7<-cnetplot(SHp_GO_GSEA,foldChange=geneList_SHp, showCategory=5)

ggp6
ggp7




#calculate pairwise similarity of GSEA genesets
SHp_GO_GSEA_interactions <- pairwise_termsim(SHp_GO_GSEA)
#interaction between genesets
ggp7_1<-emapplot(SHp_GO_GSEA_interactions, color ='NES',layout='nicely')
ggp7_1

#CHOOSING PATHWAYS TO SHOW
#retrieve synapse related pathways
View(data.frame(subset(SHp_GO_GSEA@result, grepl("synap", Description, ignore.case = TRUE))))
subset_result <- subset(SHp_GO_GSEA@result, grepl("synap", Description, ignore.case = TRUE))$Description
ggp7_2_Hp<-cnetplot(SHp_GO_GSEA,foldChange=geneList_SHp, showCategory=subset_result)

png(paste0(plotting_path_term,"SHp_GO_GSEA_dotplot.png"), width = 1000,height = 750)
ggp6
dev.off()
png(paste0(plotting_path_term,"SHp_GO_GSEA_cnetplot.png"), width = 2000,height = 2000)
ggp7
dev.off()
png(paste0(plotting_path_term,"SHp_GO_GSEA_interactions.png"), width = 1500,height = 1500)
ggp7_1
dev.off()
png(paste0(plotting_path_term,"SHp_GO_GSEA_synapse_snet.png"), width = 3000,height = 1000)
ggp7_2_Hp
dev.off()

```

### Reactome ORA and GSEA

```{r}

#Reactome over-representation
SHp_Reactome_ORA <- enrichPathway(gene=Slota_astrocyte_prion_markers_up$entrezgene_id, pvalueCutoff = 0.05, readable=TRUE, organism = "mouse")

#Reactome GSEA calculation
SHp_Reactome_GSEA <- gsePathway(geneList_SHp, 
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH",
                organism = "mouse",
                verbose = FALSE,
                seed=42,
                maxGSSize = 100) ##FIXME trying a lower limit to remove broad terms

#convert back to symbols
SHp_Reactome_GSEA <- setReadable(SHp_Reactome_GSEA, 'org.Mm.eg.db', 'ENTREZID')
View(data.frame(SHp_Reactome_GSEA))
write.csv(data.frame(SHp_Reactome_GSEA), file = paste0(plotting_path,"SHp_Reactome_GSEA.csv"))
```

#### Reactome GSEA visualisation

```{r}
#calculate pairwise similarity of GSEA genesets
SHp_Reactome_GSEA_interactions <- pairwise_termsim(SHp_Reactome_GSEA)

ggp8<-heatplot(SHp_Reactome_GSEA, showCategory = 10, foldChange=geneList_SHp)
ggp9<-cnetplot(SHp_Reactome_GSEA, circular = TRUE, colorEdge = TRUE, foldChange=geneList_SHp)
ggp10<-cnetplot(SHp_Reactome_GSEA, colorEdge = TRUE, foldChange=geneList_SHp, showCategory=10)
ggp11<-ridgeplot(SHp_Reactome_GSEA,fill = 'NES', showCategory = 10)

#interaction between genesets
ggp12<-emapplot(SHp_Reactome_GSEA_interactions, color ='NES',layout='nicely')

ggp8
ggp9
ggp10
ggp11
ggp12

png(paste0(plotting_path,"Slota_Prion_GSEA_heatplot_Reactome.png"), width = 3000,height = 1000)
ggp8
dev.off()
png(paste0(plotting_path,"Slota_Prion_GSEA_cnetplot_circular_Reactome.png"), width =2000,height = 2000)
ggp9
dev.off()
png(paste0(plotting_path,"Slota_Prion_GSEA_cnetplot_Reactome.png"), width =2000,height = 2000)
ggp10
dev.off()
png(paste0(plotting_path,"Slota_Prion_GSEA_ridgeplot_Reactome.png"), width = 1000,height = 750)
ggp11
dev.off()
png(paste0(plotting_path,"Slota_Prion_GSEA_interactions_plot_Reactome.png"), width = 1500,height = 1500)
ggp12
dev.off()
```



# CX
```{r}
#active assay RNA
DefaultAssay(Slota_CX_comp)<-"RNA"

#DimReduc again
#https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/06_SC_SCT_normalization.md
# Normalize the counts
Slota_CX_comp <- NormalizeData(Slota_CX_comp)
# Identify the most variable genes with defaults
Slota_CX_comp <- FindVariableFeatures(Slota_CX_comp, verbose = T)
		     
# Scale the counts
Slota_CX_comp <- ScaleData(Slota_CX_comp)
# Perform PCA
Slota_CX_comp <- RunPCA(Slota_CX_comp)
#Inspect the variability accounted for by the PCS
#It appears that 23 PCs are responsible, so using 30 from now on
ElbowPlot(Slota_CX_comp, ndims=30)+ggtitle("Elbowplot of Slota data", subtitle = "NBH and ME7")
DimHeatmap(Slota_CX_comp, dims=1:9, cells=500, balanced=T)
#Finish dimensinoality reduction steps
Slota_CX_comp <- RunUMAP(Slota_CX_comp, reduction = "pca", dims = 1:30, seed.use=42)
Slota_CX_comp <- FindNeighbors(Slota_CX_comp, reduction = "pca", dims = 1:30)
Slota_CX_comp <- FindClusters(Slota_CX_comp, resolution = c(0.2,0.3,0.5,0.7,1))
#Save the object
save(Slota_CX_comp,file=paste0(plotting_path,"Slota_CX_comp.Rdata"))


# Visualise the clustering resolutions for cluster number

# Determine metrics to plot present in Ver_singlets_merged_clean@meta.data
metrics <-  c("nUMI", "nGene", "mitoRatio")

#FIXME titles
DimPlot(Slota_CX_comp , reduction = "umap", group.by = "treatment", raster=FALSE)+ ggtitle("Slota Hippocampal cells labelled by treatment")

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slota_CX_comp, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Slota Hippocampal cells labelled by cluster numbers")

#adapted from https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/08_SC_clustering_quality_control.md#segregation-of-clusters-by-various-sources-of-uninteresting-variation

#Explore any clustering mishaps by lookin for localisation of any of these traits in a given cluster
FeaturePlot(Slota_CX_comp, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            #min.cutoff = 'q10', #This was a default in the online script, I don't see the benefit of colouring less of the range of values
            label = TRUE)


## Save the singlets with a cluster resolution?
#FIXME might not be required
 #set the idents to _____
Slota_CX_comp@active.ident<-Slota_CX_comp@meta.data$RNA_snn_res.0.3
levels(Slota_CX_comp)

rm(metrics)
```
```{r}
#load(paste0(plotting_path,"Slo_singlets_merged_clean.Rdata")

DefaultAssay(Slota_CX_comp)<-"RNA"

#Check for a gene of interest in the data's geneset
Gene_Names<-Slota_CX_comp@assays[["RNA"]]@counts@Dimnames[[1]]
filtered_gene_names<-Gene_Names[!grepl("^\\d", Gene_Names)]
filtered_gene_names[grep("^Slc1a2", filtered_gene_names)]

#try to identify astrocytes  with genes
astrocyte_genes<-c("Aldh1l1","Gfap","Slc1a2", "Slc1a3", "Glul", "Cd44")

FeaturePlot(Slota_CX_comp,features=astrocyte_genes, raster=F)#+ggtitle("Slota data, labelled with astrocyte gene markers from ABCAM", subtitle="Clustering resolution 0.3, 20 PCs used")

#As a reminder...
DimPlot(Slota_CX_comp, group.by = 'RNA_snn_res.0.3', reduction = 'umap', label = TRUE, repel = TRUE, raster=F)+ggtitle("Integrated Slota data, labelled by cluster numbers", subtitle="RPCA,Clustering resolution 0.3, 20 PCs used")

#No compare to the plots of astrocyte genes
VlnPlot(Slota_CX_comp, features = astrocyte_genes, group.by = 'RNA_snn_res.0.3', pt.size = 0.0)

#Glul is maybe a good indicator?
```



## (SingleR annotation)

Could go here, could be followed by bubbleplot
<https://rdrr.io/github/milescsmith/seuratBubblePlot/man/bubbleplot.html>

Alternatively, load information from the spreadsheet at
Analysis/ABCAM_markers.xlsx then bubble plot

##Investigate astrocytes, ensure identification

```{r}
# Convert cluster names to character
Idents(Slota_CX_comp) <- Slota_CX_comp$RNA_snn_res.0.3

#Subset out astrocytes
Slota_CX_comp_astrocytes<-subset(Slota_CX_comp, idents=c('5','10','22'))
```

##DEG work

```{r}


# Run the standard workflow for visualization and clustering
Slota_CX_comp_astrocytes <- ScaleData(Slota_CX_comp_astrocytes, verbose = FALSE)
Slota_CX_comp_astrocytes <- RunPCA(Slota_CX_comp_astrocytes, npcs = 30, verbose = FALSE)
ElbowPlot(Slota_CX_comp_astrocytes, ndims=30)+ggtitle("Elbowplot of subset Slota astrocytes", subtitle = "RunPCA npcs=30")

#15 PCs sems ok, so use 20 from now

#20 PCs seem to cover the variability
Slota_CX_comp_astrocytes <- RunUMAP(Slota_CX_comp_astrocytes, reduction = "pca", dims = 1:20, seed.use=42)
Slota_CX_comp_astrocytes <- FindNeighbors(Slota_CX_comp_astrocytes, reduction = "pca", dims = 1:20)
Slota_CX_comp_astrocytes <- FindClusters(Slota_CX_comp_astrocytes, resolution = c(0.2,0.3,0.5,0.7,1))

#save for reloading
save(Slota_CX_comp_astrocytes, file=paste0(plotting_path,"Slota_CX_comp_astrocytes.Rdata"))

#cyclechange resolution number in group.by to see difference in clusters
DimPlot(Slota_CX_comp_astrocytes, reduction = 'umap',  raster=F, group.by='treatment')+ggtitle("Slota data's CX astrocytes", subtitle="Labelled by treatment")

#Show the clustering of subset astrocytes
DimPlot(Slota_CX_comp_astrocytes, group.by = 'RNA_snn_res.0.2', reduction = 'umap', repel = TRUE, raster=F)

#Show sample disctribution in clusters of astrocytes
DimPlot(Slota_CX_comp_astrocytes, group.by = 'sample', reduction = 'umap', repel = TRUE, raster=F)

#Show astrocyte markers in astrocyte subset
FeaturePlot(Slota_CX_comp_astrocytes,features=c('Gfap','Cd44'), raster=F)

#set the idents to 0.2
Slota_CX_comp_astrocytes@active.ident<-(Slota_CX_comp_astrocytes@meta.data$RNA_snn_res.0.2)
levels(Slota_CX_comp_astrocytes)

DefaultAssay(Slota_CX_comp_astrocytes)<-'RNA'

#by treatment, show the genes DE in Prion condition. LogFC set to a *1 increase in fold change
Slota_CX_comp_astrocytes@active.ident<-as.factor(Slota_CX_comp_astrocytes$treatment)
Slota_astrocyte_prion_markers_CX<-FindMarkers(Slota_CX_comp_astrocytes, ident.1 = 'RML_CX', ident.2 = 'PBS_CX',min.pct = 0.10, verbose = TRUE, test.use = "DESeq2")

#Sort the list in descending log2FC and ascending pvalues
Slota_astrocyte_prion_markers_clean_CX<-Slota_astrocyte_prion_markers_CX[order(-Slota_astrocyte_prion_markers_CX$avg_log2FC), ]


#Use biomart info
Slota_astrocyte_prion_markers_clean_CX$gene<-rownames(Slota_astrocyte_prion_markers_clean_CX)

gene_ids <- unique(Slota_astrocyte_prion_markers_clean_CX$gene)
  
#get relevant info from the mart
gene_info <- getBM(attributes = c("external_gene_name", "description","chromosome_name","entrezgene_id"), filters = "external_gene_name", values = gene_ids, mart = m)
  
#apply the info to the gene list
  Slota_astrocyte_prion_markers_clean_CX <- merge(Slota_astrocyte_prion_markers_clean_CX, gene_info, by.x = "gene", by.y = "external_gene_name", all.x = TRUE)
  

#Save the file
write.csv(Slota_astrocyte_prion_markers_clean_CX, file = paste0(plotting_path,"Slota_astrocyte_prion_markers_clean_CX.csv"))

save(Slota_astrocyte_prion_markers_clean_CX, file=paste0(plotting_path,"Slota_astrocyte_prion_markers_clean_CX.RData"))

table(Slota_CX_comp_astrocytes$treatment)
```

## Functional annotation

### Address the potential non-mapping gene entrezIDs, make a gene list
with fewer NAs

```{r}

#make list of gene names not mapping
not_mapped<-(Slota_astrocyte_prion_markers_clean_CX[is.na(Slota_astrocyte_prion_markers_clean_CX$entrezgene_id),]$gene)
not_mapped<-not_mapped[!duplicated(not_mapped)]

# Create an empty data frame to store the mapped ENTREZID values
remapped <- data.frame(ALIAS = character(0), ENTREZID = character(0), stringsAsFactors = FALSE)

# Loop through each gene in 'not_mapped'
for (gene in not_mapped) {
  # Attempt to perform the mapping using tryCatch
  mapped_genes <- tryCatch(
    {
      bitr(
        gene,
        fromType = "ALIAS",
        toType = "ENTREZID",
        OrgDb = org.Mm.eg.db,
        drop = FALSE
      )
    },
    error = function(e) {
      # If an error occurs, handle it by returning a data frame with NA values
      message(paste("Error mapping gene:", gene))
      return(data.frame(ALIAS = gene, ENTREZID = NA, stringsAsFactors = FALSE))
    }
  )

  # Append the temporary data frame to 'remapped'
  remapped <- rbind(remapped, mapped_genes)
  Sys.sleep(0.5)
}

#Additionally, get biomart conversions for those still without entrez ids 
biomart_remapped<-getBM(attributes = c("external_gene_name","entrezgene_id"), filters = "external_gene_name", values = remapped[is.na(remapped$ENTREZID),]$ALIAS, mart = m)

#add any more which have been mapped
biomart_remapped<-biomart_remapped[!is.na(biomart_remapped$entrezgene_id),]

#put everything recovered into a single df
bloated_mapped_df <- merge(remapped, biomart_remapped, by.x = "ALIAS", by.y = "external_gene_name", all.x = TRUE)
bloated_mapped_df$entrezgene_id <- as.character(bloated_mapped_df$entrezgene_id)

#make a single column for everything recovered, remove old columns
condensed_mapped <- bloated_mapped_df %>%
  mutate(entrezgene_id = coalesce(ENTREZID, entrezgene_id))
condensed_mapped$entrezgene_id<-as.integer(condensed_mapped$entrezgene_id)
#remove old columns
condensed_mapped$ENTREZID<-ENTREZID<- NULL


#The final list of newly gnerated entrezIDs, removed NAs
c<-condensed_mapped[!is.na(condensed_mapped$entrezgene_id),] 

#Add them back to the original list of markers, remove eztraneous columns
Slota_astrocyte_prion_markers_clean_CX <- merge(Slota_astrocyte_prion_markers_clean_CX, c, by.x = "gene", by.y = "ALIAS", all.x = TRUE)
Slota_astrocyte_prion_markers_clean_CX$entrezgene_id<-coalesce(Slota_astrocyte_prion_markers_clean_CX$entrezgene_id.x, Slota_astrocyte_prion_markers_clean_CX$entrezgene_id.y)
Slota_astrocyte_prion_markers_clean_CX<-subset(Slota_astrocyte_prion_markers_clean_CX, select = -c(entrezgene_id.x, entrezgene_id.y))
```

### Discard any genes which don't have an ENTREZID, generate list of
background DEGs

```{r}
#Discard any genes which don't have an ENTREZID
Slota_astrocyte_prion_markers_clean_CX_entrez<-Slota_astrocyte_prion_markers_clean_CX[!is.na(Slota_astrocyte_prion_markers_clean_CX$entrezgene_id),]


#create the genelist background for Verity Up
## feature 1: numeric vector
geneList_SCx <- Slota_astrocyte_prion_markers_clean_CX_entrez$avg_log2FC
## feature 2: named vector of enembl IDs
names(geneList_SCx) <-Slota_astrocyte_prion_markers_clean_CX_entrez$entrezgene_id
## feature 3: decreasing order
geneList_SCx <- sort(geneList_SCx, decreasing = TRUE) #object of type double
```

### Subset upregulated genes and perform ORA

```{r}
#Create upregulated significant genes
# Subset by Upregulated genes for GO enrichment
Slota_astrocyte_prion_markers_clean_CX_up<-subset(Slota_astrocyte_prion_markers_clean_CX_entrez, subset=avg_log2FC>(0.25)&p_val_adj<0.05) #Added p value filter here

#Create a list of just the gene symbols 
Slota_Cx_DEGs_up = (Slota_astrocyte_prion_markers_clean_CX_up$entrezgene_id)


# Run GO over-representation on upregulated entrezIds, biological process selected from GO terms
Slota_Cx_up_GO_ORA = enrichGO(gene = Slota_Cx_DEGs_up, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)
```

#### GO ORA visualisation and saving

```{r}
ggp <- barplot(Slota_Cx_up_GO_ORA , showCategory=20)
ggp2 <- dotplot(Slota_Cx_up_GO_ORA, showCategory=20)
ggp3 <- goplot(Slota_Cx_up_GO_ORA, showCategory = 20)
ggp4<- cnetplot(Slota_Cx_up_GO_ORA, foldChange=geneList_SCx)
ggp5<-heatplot(Slota_Cx_up_GO_ORA, showCategory=20)#,foldChange=geneList_SCx)


ggp
ggp2
ggp3
ggp4
ggp5

png(paste0(plotting_path,"Slota_Cx_GOR_barplot.png"), width = 1000,height = 750)
ggp
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_dotplot.png"), width = 1000,height = 750)
ggp2
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_goplot.png"), width = 1000,height = 750)
ggp3
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_cnetplot.png"), width = 1000,height = 750)
ggp4
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_heatplot.png"), width = 3000,height = 1000)
ggp5
dev.off()
```


####Same for downregulated

```{r}

#Create down significant genes
# Subset by down genes for GO enrichment
Slota_astrocyte_prion_markers_clean_CX_down<-subset(Slota_astrocyte_prion_markers_clean_CX_entrez, subset=avg_log2FC<(-0.25)&p_val_adj<0.05) #Added p value filter here

#Create a list of just the gene symbols 
Slota_Cx_DEGs_down = (Slota_astrocyte_prion_markers_clean_CX_down$entrezgene_id)


# Run GO over-representation on upregulated entrezIds, biological process selected from GO terms
Slota_Cx_down_GO_ORA = enrichGO(gene = Slota_Cx_DEGs_down, OrgDb = org.Mm.eg.db, readable = T, ont= "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.10)

ggp_b <- barplot(Slota_Cx_down_GO_ORA , showCategory=20)
ggp2_b <- dotplot(Slota_Cx_down_GO_ORA, showCategory=20)
ggp3_b <- goplot(Slota_Cx_down_GO_ORA, showCategory = 20)
ggp4_b<- cnetplot(Slota_Cx_down_GO_ORA, foldChange=geneList_SHp)
ggp5_b<-heatplot(Slota_Cx_down_GO_ORA, showCategory=20,foldChange=geneList_SCx)


ggp_b
ggp2_b
ggp3_b
ggp4_b
ggp5_b

png(paste0(plotting_path,"Slota_Cx_GOR_down_barplot.png"), width = 1000,height = 750)
ggp_b
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_down_dotplot.png"), width = 1000,height = 750)
ggp2_b
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_down_goplot.png"), width = 1000,height = 750)
ggp3_b
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_down_cnetplot.png"), width = 1500,height = 1000)
ggp4_b
dev.off()
png(paste0(plotting_path,"Slota_Cx_GOR_down_heatplot.png"), width = 3600,height = 750)
ggp5_b
dev.off()
```


### GO GSEA

```{r}
#Perform GO GSEA
SCx_GO_GSEA <- gseGO(geneList= geneList_SCx, #This is the DEG list full list of DEGs
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE,
              seed=42)

#convert back to Symbols for plotting
SCx_GO_GSEA <- setReadable(SCx_GO_GSEA, 'org.Mm.eg.db', 'ENTREZID')

View(data.frame(SCx_GO_GSEA))

write.csv(data.frame(SCx_GO_GSEA), file = paste0(plotting_path,"SCx_GO_GSEA.csv"))
```

#### GO GSEA visualisation

```{r}
ggp6<-dotplotGsea(data = SCx_GO_GSEA,topn = 10,
            order.by = 'NES')

ggp7<-cnetplot(SCx_GO_GSEA, colorEdge = TRUE,foldChange=geneList_SCx, showCategory=5)

ggp6
ggp7

#calculate pairwise similarity of GSEA genesets
SCx_GO_GSEA_interactions <- pairwise_termsim(SCx_GO_GSEA)
#interaction between genesets
ggp7_1<-emapplot(SCx_GO_GSEA_interactions, color ='NES',layout='nicely')
ggp7_1

#CHOOSING PATHWAYS TO SHOW
#retrieve synapse related pathways
View(data.frame(subset(SCx_GO_GSEA@result, grepl("synap", Description, ignore.case = TRUE))))
subset_result_Cx <- subset(SCx_GO_GSEA@result, grepl("synap", Description, ignore.case = TRUE))$Description
ggp7_2<-cnetplot(SCx_GO_GSEA,foldChange=geneList_SCx, showCategory=subset_result_Cx)


png(paste0(plotting_path,"SCx_GO_GSEA_dotplot.png"), width = 1000,height = 750)
ggp6
dev.off()
png(paste0(plotting_path,"SCx_GO_GSEA_cnetplot.png"), width = 2000,height = 2000)
ggp7
dev.off()
png(paste0(plotting_path,"SCx_GO_GSEA_interactions.png"), width = 1500,height = 1500)
ggp7_1
dev.off()
png(paste0(plotting_path_term,"SCx_GO_GSEA_synapse_cnet.png"), width = 3000,height = 1000)
ggp7_2
dev.off()

```

### Reactome ORA and GSEA

```{r}

#Reactome over-representation
SCx_Reactome_ORA <- enrichPathway(gene=Slota_astrocyte_prion_markers_clean_CX_up$entrezgene_id, pvalueCutoff = 0.05, readable=TRUE, organism = "mouse")

#Reactome GSEA calculation
SCx_Reactome_GSEA <- gsePathway(geneList_SCx, 
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH",
                organism = "mouse",
                verbose = FALSE,
                seed=42,
                maxGSSize = 100) ##FIXME trying a lower limit to remove broad terms

#convert back to symbols
SCx_Reactome_GSEA <- setReadable(SCx_Reactome_GSEA, 'org.Mm.eg.db', 'ENTREZID')
View(data.frame(SCx_Reactome_GSEA))
write.csv(data.frame(SCx_Reactome_GSEA), file = paste0(plotting_path,"SCx_Reactome_GSEA.csv"))
```

#### Reactome GSEA visualisation

```{r}
#calculate pairwise similarity of GSEA genesets
SCx_Reactome_GSEA_interactions <- pairwise_termsim(SCx_Reactome_GSEA)

ggp8<-heatplot(SCx_Reactome_GSEA, showCategory = 10, foldChange=geneList_SCx)
ggp9<-cnetplot(SCx_Reactome_GSEA, circular = TRUE, colorEdge = TRUE, foldChange=geneList_SCx)
ggp10<-cnetplot(SCx_Reactome_GSEA, colorEdge = TRUE, foldChange=geneList_SCx, showCategory=10)
ggp11<-ridgeplot(SCx_Reactome_GSEA,fill = 'NES', showCategory = 10)

#interaction between genesets
ggp12<-emapplot(SCx_Reactome_GSEA_interactions, color ='NES',layout='nicely')

ggp8
ggp9
ggp10
ggp11
ggp12

png(paste0(plotting_path,"Slota_Cx_GSEA_heatplot_Reactome.png"), width = 3000,height = 1000)
ggp8
dev.off()
png(paste0(plotting_path,"Slota_Cx_GSEA_cnetplot_circular_Reactome.png"), width =2000,height = 2000)
ggp9
dev.off()
png(paste0(plotting_path,"Slota_Cx_GSEA_cnetplot_Reactome.png"), width =2000,height = 2000)
ggp10
dev.off()
png(paste0(plotting_path,"Slota_Cx_GSEA_ridgeplot_Reactome.png"), width = 1000,height = 750)
ggp11
dev.off()
png(paste0(plotting_path,"Slota_Cx_GSEA_interactions_plot_Reactome.png"), width = 1500,height = 1500)
ggp12
dev.off()

```
