---
title: "Verity workflow"
output: html_document
date: "2023-05-23"
editor_options: 
  markdown: 
    wrap: 72
---

# Verity dataset

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

# Load the required packages

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DropletUtils")
BiocManager::install("Seurat")

remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')

if (!require("SoupX", quietly = TRUE))
    install.packages("SoupX")

if (!require("scCustomize", quietly = TRUE))
    install.packages("scCustomize")

BiocManager::install("glmGamPoi")

library(DropletUtils)
library(Seurat)
library(tidyverse)
library(dplyr)
library(stringr)
library(knitr)
library(DoubletFinder)
library(SoupX)
library(scCustomize)
```

# Load the data from the group

```{r}

#Write samplethe sample names which you want to clean, corresponds to the directory names on datastore

# List of folder names
folder_names <- c(
  "sample1", "sample2",
  "sample5", "sample6",
  "sample9","sample11"
  )

# Loop through each folder as specified in the variable set at the start to retrieve the unfiltered matrices

#edit the path below to access your data, eg: "/exports/cmvm/eddie/eb/groups/mabbott_grp/NVerity/snRNAseq/expdata/A231222/combined...."
for (folder_name in folder_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/Group_data/",folder_name, "/DGE_unfiltered")
  
  #Load the matrix file
  current_parse <- ReadParseBio(folder_path)
  
  #see if there are any empty gene names
  table(rownames(current_parse) == "") 
  
  #Load the metadata for the sample
  current_metadata<-read.csv(paste0(folder_path,"/cell_metadata.csv"))

  #Create a seurat object
  current_parse_interpreted <- CreateSeuratObject(current_parse, meta.data = current_metadata, names.delim = "")
  
  #set identities to the sample, may be overwritten later
  current_parse_interpreted@meta.data$orig.ident <- factor(rep(folder_name, nrow(current_parse_interpreted@meta.data)))
  Idents(current_parse_interpreted) <- current_parse_interpreted@meta.data$orig.ident

  # Assign the object to a variable with the folder name as the variable name
  assign((paste0(folder_name, "_obj")),current_parse_interpreted)

}


#merge the samples into one object
Verity_total <- merge(sample1_obj, y=c(sample2_obj, sample5_obj, sample6_obj, sample9_obj,sample11_obj), add.cell.ids=c("sample1", "sample2", "sample5", "sample6", "sample9", "sample11"), project="Verity")

#cleaning up memory
rm("sample1_obj","sample2_obj", "sample5_obj", "sample6_obj", "sample9_obj", "sample11_obj", current_metadata, current_parse_interpreted, current_parse)
```

# Identify mitochondrial and ribosomal populations

```{r}
#ID mitochondrial dominant cells 
Verity_total[["percent.mt"]]<- PercentageFeatureSet(Verity_total, pattern = "^mt")
Verity_total[["percent.rp"]]<- PercentageFeatureSet(Verity_total, pattern = "^Rp")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Verity_total$log10GenesPerUMI <- log10(Verity_total$nFeature_RNA) / log10(Verity_total$nCount_RNA)

Verity_total$mitoRatio <- Verity_total@meta.data$percent.mt / 100


View(Verity_total@meta.data)
```

### Add metadata, adding sample identity

adapted from <https://youtu.be/p49seH2_i8Y?t=312>

```{r}
sampleLabs<-names(Verity_total@active.ident)

#FIXME change these to be treatment levels of experimental setup

#Assigning names to the samples to keep identity
#From Chat GPT
sample_detect <- case_when(
  str_detect(sampleLabs, "sample11") ~ "sample11",
  str_detect(sampleLabs, "sample1") ~ "sample1",
  str_detect(sampleLabs, "sample2") ~ "sample2",
  str_detect(sampleLabs, "sample5") ~ "sample5",
  str_detect(sampleLabs, "sample6") ~ "sample6",
  str_detect(sampleLabs, "sample9") ~ "sample9",
  TRUE ~ NA_character_
)
#
Verity_total@meta.data$sample<-sample_detect
Idents(object=Verity_total)<-"sample"
rm(sampleLabs, sample_detect)
```

# Emptydrops and SoupX

```{r}
#SoupX pre-processing and setup

# #csv taken from DGE_unfiltered in combined folder (all samples)
raw_objects_metadata<- read.csv("/home/s2268606/University_Directories/Project_23/Group_data/combined/cell_metadata.csv")

#Load the previously creates variables generated from Verity_emptydrop.Rmd which contain either cells or droplets

#niters set above the 10,000 default to 
#BPPARAM allows multithreading to speed up process, might need to be adjusted when the script is executed outside of the Bioinformatics servers.
out<-emptyDrops(Verity_total@assays$RNA@counts,niters= 17500,BPPARAM=MulticoreParam())
is.cell <- out$FDR <= 0.001

#show how many barcodes are truly cells
sum(is.cell, na.rm=TRUE)

#If any Limited==TRUE and Sig==FALSE are shown here, increase niters
table(Sig=is.cell, Limited=out$Limited)

#Make new matrix files
Verity_total_cells <- Verity_total@assays$RNA@counts[,which(is.cell),drop=FALSE]
Verity_total_drops<- Verity_total@assays$RNA@counts[,which(is.cell),drop=TRUE]


#Create a seurat object of each as specified by SoupX
filtered_obj <- CreateSeuratObject(Verity_total_cells, meta.data = raw_objects_metadata, names.delim = "")
raw_obj<- CreateSeuratObject(Verity_total_drops, meta.data = raw_objects_metadata, names.delim = "")

#extract the counts which have been annotated with the metadata
toc<-filtered_obj@assays$RNA@counts
tod<-raw_obj@assays$RNA@counts

#convert to SoupX object
sc = SoupChannel(tod, toc)

#start SoupX calculations
sc = estimateSoup(sc)

#tidy up
gc()


####WORKS UP TO HERE, SoupX need clustering info from now on####


## trying to cluster the Verity data to provide SoupX some info

#This variable was generated from loading the whole dataset, labelling samples and splitting by sample name to have a list of 6 objects
Verity.list <- SplitObject(Verity_filtered, split.by = "sample")

#Preprocessing to scale, normalise and find clusters, plot it too if you uncomment the dimplot line

Verity.list <- lapply(Verity.list, function(sample) {
  #state the sample being processed
  print(paste0("Sample ",sample@meta.data$sample[1]))
  
  #FIXME SCTransform the sample. The computational demands of this are currently too great. When converse memory is set to TRUE the residual matrix for all genes is never created in full; useful for large data sets, but will take longer to run; this will also set return.only.var.genes to TRUE (https://rdrr.io/github/satijalab/seurat/man/SCTransform.html)
  sample <- SCTransform(sample, method = "glmGamPoi", verbose = TRUE,conserve.memory = TRUE)
  
  sample <- RunPCA(sample, verbose = TRUE)
  sample <- RunUMAP(sample, dims = 1:30, verbose = TRUE)
  sample <- FindNeighbors(sample, dims = 1:30, verbose = TRUE)
  sample <- FindClusters(sample, verbose = TRUE)
  #DimPlot(sample, label = TRUE) + NoLegend()
})

#tidy up
gc()


#Setup and apply clusters
cluster_file<-read.csv("/localdisk/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper_original_format/cluster_file.csv")

#FIXME change this to use cluster_file
clusters<-raw_objects_metadata[2:nrow(raw_objects_metadata),c("NAME",'seurat_clusters')]

#set name of cell to rowname
rownames(clusters)<-clusters$NAME

#Get the seurat clustering info available in the metadata, it appeared to be in the wrong orientation
clusters<-t(clusters)


#set clustering info for the Soupx object
sc=setClusters(sc,clusters)
# alternative path, more manual and might not be required
# sc<-setClusters(scNoDrops,clusters)

#calculate soup
sc<- autoEstCont(sc, verbose = T)

#save the soup-free counts matrix, rounding the coutns to integers
souped_matrix <-adjustCounts(sc, roundToInt = T)

#Hopefully use the reduced matrix
current_obj <- CreateSeuratObject(counts = souped_matrix, project = "Verity")
```
