---
title: "Slota workflow"
output: html_document
date: "2023-05-23"
---

## Seurat for Slota

Adapted from [here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data) and [here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

```{r include=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DropletUtils")
library(DropletUtils)

if (!require("irlba", quietly = TRUE))
    install.packages("irlba")

BiocManager::install("Seurat")
library(Seurat)
library(tidyverse)

#library(ggplot2)

library(dplyr)
library(stringr)
library(knitr)

remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
library(DoubletFinder)

if (!require("SoupX", quietly = TRUE))
    install.packages("SoupX")
library(SoupX)
library(Matrix)
```

```{r}
folder_path_slota<-c("/localdisk/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper")

#load Slota paper 

raw_objects_metadata<- read.csv(paste0(folder_path_slota,"/metadata.csv"), header = T)
#remove the extra row about categories
raw_objects_metadata<-raw_objects_metadata[2:nrow(raw_objects_metadata),]

rawtable<-Read10X(paste0(folder_path_slota,"/raw_feature_bc_matrix"))

Slota_raw_obj<-CreateSeuratObject(counts=rawtable, project="Slota")
```

## Identify mitochondrial and ribosomal populations

```{r}
#identify mitochondrial and ribosomal protein gene portions, annotations syntax tailored to mouse
Slota_raw_obj[["percent.mt"]]<- PercentageFeatureSet(Slota_raw_obj, pattern = "^mt")
Slota_raw_obj[["percent.rp"]]<- PercentageFeatureSet(Slota_raw_obj, pattern = "^Rp")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Slota_raw_obj$log10GenesPerUMI <- log10(Slota_raw_obj$nFeature_RNA) / log10(Slota_raw_obj$nCount_RNA)

Slota_raw_obj$mitoRatio <- Slota_raw_obj@meta.data$percent.mt / 100
```
