---
title: "Slota workflow"
output: html_document
date: "2023-05-23"
---

## SCE starting

Adapted from [here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data) and [here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DropletUtils")
library(DropletUtils)

#mouse annotation https://bioconductor.org/packages/release/data/annotation/html/org.Mm.eg.db.html
BiocManager::install("org.Mm.eg.db")
library(org.Mm.eg.db)

#and https://bioconductor.org/packages/release/data/annotation/html/EnsDb.Mmusculus.v79.html
BiocManager::install("EnsDb.Mmusculus.v79")
library(EnsDb.Mmusculus.v79)
```

```{r}
#load Slota paper ('original gene expression matrices generated by cellranger' according to https://singlecell.broadinstitute.org/single_cell/study/SCP1962/dysregulation-of-neuroprotective-astrocytes-a-spectrum-of-microglial-activation-states-and-altered-hippocampal-neurogenesis-are-revealed-by-single-cell-rna-sequencing-in-prion-disease#study-download)

Slota_sce<-read10xCounts('/localdisk/home/s2268606/University_Directories/Project_23/External_datasets/Slota_paper', col.names=TRUE)
```

```{r}
Slota_sce
```

There are 2868 genes (rows) and 147536 cell (columns) (corresponds with origin of data).

```{r}
#get log2 counts
assay(Slota_sce, "logcounts")<-log2(counts(Slota_sce)+1)
```

```{r}
logcounts(Slota_sce)[1:10, 1:4]
```

```{r}
#get mean counts per cell
colData(Slota_sce)$mean_counts <- colMeans(counts(Slota_sce))
colData(Slota_sce)
```

```{r}
#Sum of counts in each cell
colData(Slota_sce)$total_counts <- colSums(counts(Slota_sce))
```

```{r}
#Counts per million
# function sweep takes following arguments:
# matrix to normalize
# dimention to normalize along (1 - by rows, 2 - by columns)
# statistics to normalize by
# function to use for normalization
assay(Slota_sce, "cpm") <- sweep(counts(Slota_sce),2,Slota_sce$total_counts/1e6,'/')
# check that column sums are 1e6 now
colSums(cpm(Slota_sce))[1:10]
```

## Filtering (might need to be repeated after some EDA)

```{r}
#we can retain genes with mean count > 0.01
gene_means <- rowMeans(counts(Slota_sce))
Slota_sce[gene_means > 0.01, ]

#We can filter by a threshold of genes
#how many genes per cell
total_detected_per_cell <- colSums(counts(Slota_sce) > 0)

#Use vector to apply final condition (at least 1k genes as suggested by Nick on 23/05/2023)
Slota_sce_filtered <- Slota_sce[, total_detected_per_cell >= 1000]


#Filter by read number. At least 3.5k reads(counts) is too harsh so commented out (see note below)
#cell_filter <- colSums(counts(Slota_sce)) >= 3500
# check how many TRUE/FALSE have
#table(cell_filter)
#Slota_sce_filtered <- Slota_sce[ , cell_filter]

Slota_sce_filtered
```

The Slota_sce_filtered object has only 38 cells left after all of the filtering above was applied, so filters will need to be relaxed. Without the 3.5 k reads filter there were 90318 cells left.

```{r}
#trying to annotate with ensembl, but think that the matrix file might already have this.
#check that the rownames , keytypes
#gene_names <- mapIds(org.Mm.eg.db, keys=rownames(counts(Slota_sce_filtered)), keytype='ENSEMBL',columns='SYMBOL', column='SYMBOL')

#The following gave 0 mitochondrial genes, has the dataset been filtered already?

ensdb_genes<- genes(EnsDb.Mmusculus.v79)
MT_names<- ensdb_genes[seqnames(ensdb_genes)=="MT"]$gene_id
is_mito<-rownames(counts(Slota_sce_filtered) %in% MT_names
)
is_mito<-rownames(counts(Slota_sce_filtered)) %in% MT_names
table(is_mito)
```

```{r}
#could do without filtering
#cell_info <- as.data.frame(colData(Slota_sce))

cell_info_filtered <- as.data.frame(colData(Slota_sce))
cell_info_filtered$rownames <- metadata$biosample_id[match(cell_info_filtered$Barcode, metadata$NAME)]
```

```{r}
library(ggplot2)

ggplot(data = cell_info_filtered, aes(x = rownames, y = total_counts)) +
  geom_violin(fill = 'brown') +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  labs(title = "Total counts per cell", x = "Sample Origin (by Brain Region)")

```
