---
title: "Verity workflow"
output: html_document
date: "2023-05-23"
editor_options: 
  markdown: 
    wrap: 72
---

# Verity dataset

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

```{r include=FALSE}

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DropletUtils")
library(DropletUtils)

BiocManager::install("Seurat")
library(Seurat)
library(tidyverse)

#library(ggplot2)

library(dplyr)
library(stringr)
library(knitr)

remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
library(DoubletFinder)

if (!require("SoupX", quietly = TRUE))
    install.packages("SoupX")
library(SoupX)

if (!require("scCustomize", quietly = TRUE))
    install.packages("scCustomize")
library(scCustomize)
```

```{r}
#load Verity paper FIXME this will need to be from group folder eventually, so amend paths at the start

#Comment out the samples which aren't needed for the comparison of interest

# List of folder names
folder_names <- c(
  "sample1", "sample2",
  "sample5", "sample6",
  "sample9","sample11"
  )

# Loop through each folder
for (folder_name in folder_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/Group_data/",folder_name, "/DGE_unfiltered")
  
  
  #Load the matrix file
  current_parse <- ReadParseBio(folder_path)
  
  #see if there are any empty gene names
  table(rownames(current_parse) == "") 
  
  #Load the metadata
  current_metadata<-read.csv(paste0(folder_path,"/cell_metadata.csv"))

  #Create a seurat object
  current_parse_interpreted <- CreateSeuratObject(current_parse, meta.data = current_metadata, names.delim = "")
  
  #set identities to the sample, may be overwritten later
  current_parse_interpreted@meta.data$orig.ident <- factor(rep(folder_name, nrow(current_parse_interpreted@meta.data)))
  
  Idents(current_parse_interpreted) <- current_parse_interpreted@meta.data$orig.ident

  
  # #Alternatively
  # current_metadata<-read.csv(paste0(folder_path,"/cell_metadata.csv"))
  # ##replace commas or underscores (view other mtx files to get an idea of correct syntax and the head of our data for our current syntax) FIXME
  # current_metadata[] <- lapply(current_metadata, function(x) gsub("_", ",", x))
  # 
  # current_genes<-read.csv(paste0(folder_path, "/all_genes.csv"))
  # 
  # #Add cell identifiers
  # rownames(current_current_parserix)<-current_metadata[,1]
  # 
  # current_Seurat<-CreateSeuratObject(
  #   counts= current_current_parserix,
  #   project = "Verity",
  #   assay = "RNA",
  #   # names.field = 3, #FIXME CHANGE
  #   # names.delim = "_", #FIXME CHANGE
  #   meta.data = current_metadata,
  #   row.names = current_genes$gene_name
  #     )
  # 
  

  # Assign the object to a variable with the folder name as the variable name
  assign((paste0(folder_name, "_obj")),current_parse_interpreted)

}


#merge the samples into one object
Verity_total <- merge(sample1_obj, y=c(sample2_obj, sample5_obj, sample6_obj, sample9_obj,sample11_obj), add.cell.ids=c("sample1", "sample2", "sample5", "sample6", "sample9", "sample11"), project="Verity")

#cleaning up memory
rm("sample1_obj","sample2_obj", "sample5_obj", "sample6_obj", "sample9_obj", "sample11_obj", current_metadata, current_parse_interpreted, current_parse)
```

## Identify mitochondrial and ribosomal populations

```{r}
#ID mitochondrial dominant cells 
Verity_total[["percent.mt"]]<- PercentageFeatureSet(Verity_total, pattern = "^mt")
Verity_total[["percent.rp"]]<- PercentageFeatureSet(Verity_total, pattern = "^Rp")

View(Verity_total@meta.data)
```

## Plot violins of overall characteristics in raw data

```{r}
#plot(density(Verity_total@meta.data$nFeature_RNA))
VlnPlot(Verity_total, features=c("nFeature_RNA","nCount_RNA","percent.mt", "percent.rp"), ncol=4, pt.size = 0.05)
```

## Plot each of these respectively, showing trends in cell health

```{r}
#show mitochondrial proportion in reads
FeatureScatter(Verity_total, feature1 = "nCount_RNA", feature2 = "percent.mt")

#show ribosomal proportion in reads
FeatureScatter(Verity_total, feature1 = "nCount_RNA", feature2 = "percent.rp")

#show 
FeatureScatter(Verity_total, feature1 = "percent.rp", feature2 = "percent.mt")

```

```{r}
sampleLabs<-names(Verity_total@active.ident)

#FIXME change these to be treatment levels of experimental setup

#Assigning names to the samples to keep identity
#From Chat GPT
sample_detect <- case_when(
  str_detect(sampleLabs, "sample1") ~ "sample1",
  str_detect(sampleLabs, "sample2") ~ "sample2",
  str_detect(sampleLabs, "sample5") ~ "sample5",
  str_detect(sampleLabs, "sample6") ~ "sample6",
  str_detect(sampleLabs, "sample9") ~ "sample9",
  str_detect(sampleLabs, "sample11") ~ "sample11",
  TRUE ~ NA_character_
)
#
Verity_total@meta.data$sample<-sample_detect
Idents(object=Verity_total)<-"sample"
rm(sampleLabs, sample_detect)
```

```{r}
#Show quailty of dataset 
FeatureScatter(Verity_total, feature1="nCount_RNA",feature2="nFeature_RNA",raster=FALSE)+geom_smooth(method="lm")+ggtitle("Transcripts vs. genes of unfiltered Verity dataset")
```

```{r}
#Filter dataset by mitochondrial proportion and gene counts
Verity_total<- subset(Verity_total, subset= percent.mt <5 &nFeature_RNA > .........) #FIXME filter appropriately
```

## Doublet finder for all samples

```{r}
#adapted from https://youtu.be/p49seH2_i8Y , https://github.com/kpatel427/YouTubeTutorials/blob/5b3c795cecafb22f0e992da069eae2efd6b5cb95/singleCell_doublets.R , https://rpubs.com/kenneditodd/doublet_finder_example 

Verity.list <- SplitObject(Verity_total, split.by = "sample")

# Perform doublet detection on each sample using lapply
Verity.list <- lapply(Verity.list, function(sample) {
  print(paste0("Sample ",sample@meta.data$sample[1]))
  # Pre-process seurat object with standard seurat workflow
  sample <- NormalizeData(sample)
  sample <- SCTransform(sample)
  sample <- RunPCA(sample)
  
  # Find significant PCs
  stdv <- sample[["pca"]]@stdev
  sum.stdv <- sum(sample[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] -
                       percent.stdv[2:length(percent.stdv)]) > 0.1),
              decreasing = TRUE)[1] + 1
  min.pc <- min(co1, co2)
  
  # Finish pre-processing
  sample <- RunUMAP(sample, dims = 1:min.pc)
  sample <- FindNeighbors(object = sample, dims = 1:min.pc)              
  sample <- FindClusters(object = sample, resolution = 0.1)
  
  # pK identification (no ground-truth)
  sweep.list <- paramSweep_v3(sample, PCs = 1:min.pc, sct = TRUE)
  sweep.stats <- summarizeSweep(sweep.list)
  bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficient (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- as.numeric(bcmvn.max$pK)

  # Homotypic doublet proportion estimate
  annotations <- sample@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.poi <- round(0.05 * nrow(sample@meta.data))
  nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # Run DoubletFinder
  sample <- doubletFinder_v3(seu = sample, 
                             PCs = 1:min.pc, 
                             pK = optimal.pk,
                             nExp = nExp.poi.adj, 
                             sct = TRUE, 
                             pN=0.25)
  
  return(sample)
  
})

#Rename the columns
Verity.list <- lapply(Verity.list, function(sample) {
  doublet_col_index <- which(grepl("DF.classifications_", colnames(sample@meta.data)))
  sample$doublet_finder <- sample@meta.data[doublet_col_index]
  sample@meta.data[doublet_col_index]<-NULL
  return(sample)
})

#subset data by singlet status
Verity.list <- lapply(Verity.list, function(sample) {
  sample<- subset(sample, subset= doublet_finder == 'Singlet')
  return(sample)
  })
```
