---
title: "Verity workflow"
output: html_document
date: "2023-05-23"
editor_options: 
  markdown: 
    wrap: 72
---

# Verity dataset

Adapted from
[here](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#reading-in-10x-genomics-data)
and
[here](https://www.singlecellcourse.org/scrna-seq-analysis-with-bioconductor.html).

```{r include=FALSE}

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DropletUtils")
library(DropletUtils)

BiocManager::install("Seurat")
library(Seurat)
library(tidyverse)

#library(ggplot2)

library(dplyr)
library(stringr)
library(knitr)

remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
library(DoubletFinder)

if (!require("SoupX", quietly = TRUE))
    install.packages("SoupX")
library(SoupX)

if (!require("scCustomize", quietly = TRUE))
    install.packages("scCustomize")
library(scCustomize)
```

```{r}
#load Verity paper FIXME this will need to be from group folder eventually, so amend paths at the start

#Comment out the samples which aren't needed for the comparison of interest

# List of folder names
folder_names <- c(
  "sample1", "sample2",
  "sample5", "sample6",
  "sample9","sample11"
  )

# Loop through each folder
for (folder_name in folder_names) {
  folder_path=paste0("/home/s2268606/University_Directories/Project_23/Group_data/",folder_name, "/DGE_unfiltered")
  
  
  #Load the matrix file
  current_parse <- ReadParseBio(folder_path)
  
  #see if there are any empty gene names
  table(rownames(current_parse) == "") 
  
  #Load the metadata
  current_metadata<-read.csv(paste0(folder_path,"/cell_metadata.csv"))

  #Create a seurat object
  current_parse_interpreted <- CreateSeuratObject(current_parse, meta.data = current_metadata, names.delim = "")
  
  #set identities to the sample, may be overwritten later
  current_parse_interpreted@meta.data$orig.ident <- factor(rep(folder_name, nrow(current_parse_interpreted@meta.data)))
  
  Idents(current_parse_interpreted) <- current_parse_interpreted@meta.data$orig.ident


  # Assign the object to a variable with the folder name as the variable name
  assign((paste0(folder_name, "_obj")),current_parse_interpreted)

}


#merge the samples into one object
Verity_total <- merge(sample1_obj, y=c(sample2_obj, sample5_obj, sample6_obj, sample9_obj,sample11_obj), add.cell.ids=c("sample1", "sample2", "sample5", "sample6", "sample9", "sample11"), project="Verity")

#cleaning up memory
rm("sample1_obj","sample2_obj", "sample5_obj", "sample6_obj", "sample9_obj", "sample11_obj", current_metadata, current_parse_interpreted, current_parse)
```

## Identify mitochondrial and ribosomal populations

```{r}
#ID mitochondrial dominant cells 
Verity_total[["percent.mt"]]<- PercentageFeatureSet(Verity_total, pattern = "^mt")
Verity_total[["percent.rp"]]<- PercentageFeatureSet(Verity_total, pattern = "^Rp")

#Generate genes per umi counts, from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md
Verity_total$log10GenesPerUMI <- log10(Verity_total$nFeature_RNA) / log10(Verity_total$nCount_RNA)

Verity_total$mitoRatio <- Verity_total@meta.data$percent.mt / 100


View(Verity_total@meta.data)
```

## Initial data assessment

```{r}
png("Verity_Preliminary_Data_assessment.png", width=2000)
VlnPlot(Verity_total, features=c("nFeature_RNA","nCount_RNA","percent.mt", "percent.rp"), ncol=4, pt.size = 0.01)+ ggplot2::ggtitle("Preliminary Data assessment", subtitle= "Verity, unfiltered")
dev.off()
```

## Plot each of these respectively, showing trends in cell health

```{r}
#show mitochondrial proportion in reads
png("Verity_UMIs per cell vs.mitochondrial genes detected.png", width=1000)
FeatureScatter(Verity_total, feature1 = "nCount_RNA", feature2 = "percent.mt")+ggplot2::ggtitle("UMIs per cell vs. % mitochondrial genes detected")
dev.off()

png("Verity_UMIs_per_cell_vsribosomal_protein_genes_detected.png", width=1000)
#show ribosomal proportion in reads
FeatureScatter(Verity_total, feature1 = "nCount_RNA", feature2 = "percent.rp")+ggplot2::ggtitle("UMIs per cell vs. % ribosomal protein  genes detected")
dev.off()

png("Verity_ribosomal_protein_genes_detected_vsmitochondiral_genes_detected.png", width=1000)
#Ribosomal vs mitochondrial 
FeatureScatter(Verity_total, feature1 = "percent.rp", feature2 = "percent.mt")+ggplot2::ggtitle("% ribosomal protein  genes detected vs. % mitochondiral genes detected")
dev.off()
```

```{r}
#Show overall reads and gene counts for unfiltered data.
png("Verity_UMIs_per_cell_vs_number_of_genes.png", width=1000)
FeatureScatter(Verity_total, feature1 = "nCount_RNA", feature2="nFeature_RNA",raster=FALSE, pt.size = 0.05)+ ggplot2::geom_smooth(method="lm")+ ggplot2::ggtitle("UMIs per cell vs. number of genes ")
dev.off()
```


### Add metadata, adding sample identity

adapted from <https://youtu.be/p49seH2_i8Y?t=312>

```{r}
sampleLabs<-names(Verity_total@active.ident)

#FIXME change these to be treatment levels of experimental setup

#Assigning names to the samples to keep identity
#From Chat GPT
sample_detect <- case_when(
  str_detect(sampleLabs, "sample1") ~ "sample1",
  str_detect(sampleLabs, "sample2") ~ "sample2",
  str_detect(sampleLabs, "sample5") ~ "sample5",
  str_detect(sampleLabs, "sample6") ~ "sample6",
  str_detect(sampleLabs, "sample9") ~ "sample9",
  str_detect(sampleLabs, "sample11") ~ "sample11",
  TRUE ~ NA_character_
)
#
Verity_total@meta.data$sample<-sample_detect
Idents(object=Verity_total)<-"sample"
rm(sampleLabs, sample_detect)
```

### Save the whole data

```{r}
#Save the data object
save(Verity_total, file="./Rdata/Verity_total_labelled.RData")
```

### Create metadata variable to edit it without affecting core data

```{r}
# Rename columns from https://github.com/hbctraining/scRNA-seq/blob/master/lessons/04_SC_quality_control.md

#Create metadata variable to edit it without affecting core data
metadata_test<-Verity_total@meta.data

#Saving cell barcodes as metainfo
metadata_test$cells<-rownames(metadata_test)

metadata_test <- metadata_test %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)
```


### Counts per sample

This shows over 300k cells for some samples, indicating that minimum metrics must be set to remove junk.

```{r}
# Visualize the number of cell counts per sample with no filtering at all
png("Verity_Cell_counts_per_sample1.png", width=1500)
metadata_test %>% 
  	ggplot2::ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Cell counts per sample", subtitle="Verity, unfiltered")
dev.off()
```

## Setting a minimum gene count for quality control



```{r}
#Setting minimum feature count to filter cells with no genes UMIs vs. genes detected

metadata_test <- subset(Verity_total, subset= nFeature_RNA > 20)@meta.data
metadata_test <- metadata_test %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)
```

```{r}
# Visualize the number of cell counts per sample with a minimum gene count set
png("Verity_Cell_counts_per_sample2.png", width=1000)
metadata_test %>% 
  	ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Cell counts per sample", subtitle="Verity, filtered to minimum 20 genes")
dev.off()
```

### UMI/ transcripts per cell

With line at 500 for ideal minimum

```{r}
# Visualize the number UMIs/transcripts per cell
png("Verity_UMI_transcripts_per_cell.png", width=1000)
metadata_test %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("UMI/transcripts per cell", subtitle="Verity, filtered to minimum 20 genes")
dev.off()
```

### Genes per cell

```{r}
# Visualize the distribution of genes detected per cell via histogram
png("Verity_Distribution_of_genes_per_cell1.png", width=1000)
metadata_test %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Distribution of genes per cell", subtitle="Verity, filtered to minimum 20 genes")
dev.off()

# Visualize the distribution of genes detected per cell via boxplot
png("Verity_Distribution_of_genes_per_cell2.png", width=1000)
metadata_test %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Distribution of genes per cell", subtitle="Verity, filtered to minimum 20 genes")
dev.off()
```

### UMI vs. genes detected

We want cells in top right

```{r}
#Some code commented out because it introduced infitine values with a log transform
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
png("Verity_UMIs_vs_genes_detected.png", width=2000)
metadata_test %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	#stat_smooth(method=lm) +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("UMIs vs. genes detected", subtitle = "Verity, minimum 20 genes")
dev.off()
```

### MitoRatio

We want data lower than 0.2

```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
png("Verity_Distribution_of_mitochondrial_gene_expression_per_cell.png", width=1000)
metadata_test %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Distribution of mitochondrial gene expression per cell", subtitle="Verity, filtered to minimum 20 genes")
dev.off()
```

```{r}
png("Verity_Repeat_mitoratio.png", width=1000)
FeatureScatter((subset(Verity_total, subset= nFeature_RNA > 20)), feature1 = "nFeature_RNA", feature2 = "percent.mt")
dev.off()
```

### Complexity

Should be above 0.8.

```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
png("Verity_Complexity_of_gene_expression_genes_per_UMI2.png", width=1000)
metadata_test %>%
  	ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)+
    ggplot2::ggtitle("Complexity of gene expression (genes per UMI)", subtitle="Verity, filtered to minimum 20 genes")
dev.off()
```



# STOP HERE






### Filtering metadata

```{r}
metadata_filtered<- metadata_test %>%
    filter(nUMI > 0, nGene >150, percent.mt <20)
```

## Repeat the QC analysis

```{r}
# Visualize the number of cell counts per sample with a minimum gene count set
png("Verity_Cell_counts_per_sample2.png", width=1000)
metadata_filtered %>% 
  	ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Cell counts per sample", subtitle="Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()

table(metadata_filtered$sample)
```

### UMI/ transcripts per cell

With line at 500 for ideal minimum

```{r}
# Visualize the number UMIs/transcripts per cell
png("Verity_UMI_transcripts_per_cell2.png", width=1000)
metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)+
    ggplot2::ggtitle("UMI/transcripts per cell", subtitle="Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()
```

### Genes per cell

```{r}
# Visualize the distribution of genes detected per cell via histogram
png("Verity_Distribution_of_genes_per_cell3.png", width=1000)
metadata_filtered %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)+
    ggplot2::ggtitle("Distribution of genes per cell", subtitle="Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()

# Visualize the distribution of genes detected per cell via boxplot
png("Verity_Distribution_of_genes_per_cell4.png", width=1000)
metadata_filtered %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggplot2::ggtitle("Distribution of genes per cell", subtitle="Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()
```

### UMI vs. genes detected

We want cells in top right

```{r,fig.dim = c(8, 6)}
#Some code commented out because it introduced infitine values with a log transform
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
png("Verity_UMIs_vs_genes_detected2.png", width=1000)
metadata_filtered %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	#stat_smooth(method=lm) +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample) +
    ggplot2::ggtitle("UMIs vs. genes detected", subtitle = "Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()
```

### MitoRatio

We want data lower than 0.2

```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
png("Verity_Distribution_of_mitochondrial_gene_expression_per_cell2.png", width=1000)
metadata_filtered %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)+
    ggplot2::ggtitle("Distribution of mitochondrial gene expression per cell", subtitle="Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()
```

### Complexity

Should be above 0.8.

```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
png("Verity_Complexity_of_gene_expression_genes_per_UMI_2.png", width=1000)
metadata_filtered %>%
  	ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)+
    ggplot2::ggtitle("Complexity of gene expression (genes per UMI)", subtitle="Verity, filtered to minimum 0 transcripts, 300 genes and maVerum 20% mitochondrial gene content")
dev.off()
```

### Assign filters to data

```{r}
#filter out the cells with mitochondrial genes more than 5% and genes with more than .... genes as instructed by Nick
temp_meta<-Verity_total@meta.data
temp_meta <- temp_meta %>%
    dplyr::rename(nUMI = nCount_RNA,
                  nGene = nFeature_RNA)

Verity_total@meta.data<-temp_meta


Verity_filtered<- subset(Verity_total,subset= 
                              percent.mt <20 &
                              nUMI > 0 &
                              nGene >150
                            )
rm(temp_meta, metadata_test, metadata_filtered)
```

### Gene-level filtering

Keeping genes which appear in more than 3 cells for statistical power.

```{r}

# Output a logical vector for every gene on whether the more than zero counts per cell
# Extract counts
counts <- GetAssayData(object = Verity_filtered, slot = "counts")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 3 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 3

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
Verity_filtered <- CreateSeuratObject(filtered_counts, meta.data = Verity_filtered@meta.data)
rm(filtered_counts, nonzero,counts, keep_genes)
```

### Saving object

```{r}
# Create .RData object to load at any time
save(Verity_filtered, file="./Rdata/Verity_predoublet.RData")
```

```{r}
# clearing some space by removing the original seurat objects
rm(list=paste0(folder_names,"obj"))
```










## Doublet finder for all samples

```{r}

#load in the filtered Verity object
#load("./Rdata/Verity_filtered.RData")

#adapted from https://youtu.be/p49seH2_i8Y , https://github.com/kpatel427/YouTubeTutorials/blob/5b3c795cecafb22f0e992da069eae2efd6b5cb95/singleCell_doublets.R , https://rpubs.com/kenneditodd/doublet_finder_example 

Verity.list <- SplitObject(Verity_filtered, split.by = "sample")

# Perform doublet detection on each sample using lapply
Verity.list <- lapply(Verity.list, function(sample) {
  print(paste0("Sample ",sample@meta.data$sample[1]))
  # Pre-process seurat object with standard seurat workflow
  sample <- NormalizeData(sample)
  sample <- SCTransform(sample)
  sample <- RunPCA(sample)
  
  # Find significant PCs
  stdv <- sample[["pca"]]@stdev
  sum.stdv <- sum(sample[["pca"]]@stdev)
  percent.stdv <- (stdv / sum.stdv) * 100
  cumulative <- cumsum(percent.stdv)
  co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
  co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] -
                       percent.stdv[2:length(percent.stdv)]) > 0.1),
              decreasing = TRUE)[1] + 1
  min.pc <- min(co1, co2)
  
  # Finish pre-processing
  sample <- RunUMAP(sample, dims = 1:min.pc)
  sample <- FindNeighbors(object = sample, dims = 1:min.pc)              
  sample <- FindClusters(object = sample, resolution = 0.1)
  
  # pK identification (no ground-truth)
  sweep.list <- paramSweep_v3(sample, PCs = 1:min.pc, sct = TRUE)
  sweep.stats <- summarizeSweep(sweep.list)
  bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficient (BCmvn) distribution
  bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
  optimal.pk <- as.numeric(bcmvn.max$pK)

  # Homotypic doublet proportion estimate
  annotations <- sample@meta.data$seurat_clusters
  homotypic.prop <- modelHomotypic(annotations) 
  nExp.poi <- round(0.087 * nrow(sample@meta.data))
  nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # Run DoubletFinder
  sample <- doubletFinder_v3(seu = sample, 
                             PCs = 1:min.pc, 
                             pK = optimal.pk,
                             nExp = nExp.poi.adj, 
                             sct = TRUE, 
                             pN=0.25)
  
  return(sample)
  
})

#Rename the columns
Verity.list <- lapply(Verity.list, function(sample) {
  doublet_col_index <- which(grepl("DF.classifications_", colnames(sample@meta.data)))
  sample$doublet_finder <- sample@meta.data[doublet_col_index]
  sample@meta.data[doublet_col_index]<-NULL
  return(sample)
})

#subset data by singlet status
Verity.list.singlets <- lapply(Verity.list, function(sample) {
  sample<- subset(sample, subset= doublet_finder == 'Singlet')
  return(sample)
  })

```

```{r}
#merge the data back together
Ver_singlets_merged <- Merge_Seurat_List(list_seurat = Verity.list.singlets)
rm(Verity.list.singlets)
```

### Saving object

```{r}
# Create .RData object to load at any time
save(Ver_singlets_merged, file="Rdata/Ver_singlets_merged.RData")

```