```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages

```{r}
BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
library(harmony)
library(DropletUtils)
library(Seurat)
library(tidyverse)
library(dplyr)
library(stringr)
library(knitr)
library(DoubletFinder)
library(SoupX)
library(scCustomize)
library(DESeq2)
# library(devtools)
# install_github("immunogenomics/harmony")
```

# Set env variables

```{r}

#For loading
dir_name_ver <- "Verity_plots" #change depending on run to pull data from
plotting_path_ver<-paste0("../Analysis/", dir_name_ver,"/")


dir_name_term <- "Terminal_sample_investigation"
plotting_path_term<-paste0("../Analysis/", dir_name_term,"/")

#For saving #Change this for each run!
dir_name_term <- "../Analysis/Venn_020823_1"
# create folder:
system(paste0("mkdir ", dir_name_term))
plotting_path_DEG_comp<-paste0(dir_name_term,"/")

```


# Add annotation from ensembl for gene description, ensembl ID and chromosome location

```{r}
# load the markers for Verity
load(paste0(plotting_path_ver,"Verity_astrocyte_prion_markers.RData"))
load(paste0(plotting_path_ver,"Verity_astrocyte_ageing_markers.RData"))

#load the markers for Slota Hippocampus
load(paste0(plotting_path_term,"Slota_astrocyte_prion_markers_clean_HP.RData"))
load(paste0(plotting_path_term,"Slota_astrocyte_prion_markers_clean_CX.RData"))


#Make list from DEG lists
DEG_data_Verity_Slota<-list(Slota_astrocyte_prion_markers_clean_CX,Slota_astrocyte_prion_markers_clean_HP,Verity_astrocyte_prion_markers,Verity_astrocyte_ageing_markers)

```

#Remove duplicates based on emsembl IDs and make volcano plots

```{r}

lapply(DEG_data_Verity_Slota, function(DEG_part) {
  #remove duplicates
  DEG_part<-(DEG_part[!duplicated(DEG_part$gene),])
  rownames(DEG_part)<-DEG_part$gene
  #plot volcano
  EnhancedVolcano(DEG_part,
lab = rownames(DEG_part),
x = 'avg_log2FC',
y = 'p_val_adj',
pCutoff = 0.05,
FCcutoff = 0.25)
})

```


# VennDiagram 
##Prep

```{r}

#Plot raw list

DEG_names_Terminal_CX_RAW<-(DEG_data_Verity_Slota[[1]]$gene)
DEG_names_Terminal_HP_RAW<-(DEG_data_Verity_Slota[[2]]$gene)
DEG_names_Verity_PRION_RAW<-(DEG_data_Verity_Slota[[3]]$gene)
DEG_names_Verity_AGEING_RAW<-(DEG_data_Verity_Slota[[4]]$gene)

library(gplots)
# Create a Venn-diagram given just the list of gene-names for both sets
png(paste0(plotting_path_DEG_comp,"RAW_DEGs_Venndiagram.png"), width = 1000,height = 750)
venn_diagram_RAW <- venn(list("Verity PRION" = DEG_names_Verity_PRION_RAW,
                              "Verity AGEING" = DEG_names_Verity_AGEING_RAW,
                              "Terminal HP" = DEG_names_Terminal_HP_RAW,
                              "Terminal CX" = DEG_names_Terminal_CX_RAW))
dev.off()


# Inspect the list of classifications
interactions_RAW<-attr(venn_diagram_RAW,"intersections")
interactions_RAW
interactions_RAW<-(t(data.frame(t(sapply(interactions_RAW,c)))))
write.csv(interactions_RAW, file=paste0(plotting_path_DEG_comp,"VennGenes_RAW.csv"))


#Filter data
DEG_data_Verity_Slota_UP <- lapply(DEG_data_Verity_Slota, function(DEG_part) {
  DEG_part<-subset(DEG_part,subset=avg_log2FC > (0.25) & p_val_adj<(0.05))
  return(DEG_part)
})

DEG_data_Verity_Slota_DOWN <- lapply(DEG_data_Verity_Slota, function(DEG_part) {
  DEG_part<-subset(DEG_part,subset=avg_log2FC < (-0.25) & p_val_adj<(0.05))
  return(DEG_part)
})

#Plot Up

#make character vectors of names
DEG_names_Terminal_CX_UP<-(DEG_data_Verity_Slota_UP[[1]]$gene)
DEG_names_Terminal_HP_UP<-(DEG_data_Verity_Slota_UP[[2]]$gene)
DEG_names_Verity_PRION_UP<-(DEG_data_Verity_Slota_UP[[3]]$gene)
DEG_names_Verity_AGEING_UP<-(DEG_data_Verity_Slota_UP[[4]]$gene)


# Create a Venn-diagram given just the list of gene-names for both sets
png(paste0(plotting_path_DEG_comp,"UP_DEGs_Venndiagram.png"), width = 1000,height = 750)
venn_diagram_UP <- venn(list("Verity PRION UP" = DEG_names_Verity_PRION_UP,
                              # "Verity AGEING UP" = DEG_names_Verity_AGEING_UP,
                              "Terminal HP UP" = DEG_names_Terminal_HP_UP,
                              "Terminal CX UP" = DEG_names_Terminal_CX_UP))
dev.off()

# Inspect the list of classifications
interactions_UP<-attr(venn_diagram_UP,"intersections")
interactions_UP
interactions_UP<-(t(data.frame(t(sapply(interactions_UP,c)))))
write.csv(interactions_UP, file=paste0(plotting_path_DEG_comp,"VennGenes_UP.csv"))

#Plot Down

#make character vectors of names
DEG_names_Terminal_CX_DOWN<-(DEG_data_Verity_Slota_DOWN[[1]]$gene)
DEG_names_Terminal_HP_DOWN<-(DEG_data_Verity_Slota_DOWN[[2]]$gene)
DEG_names_Verity_PRION_DOWN<-(DEG_data_Verity_Slota_DOWN[[3]]$gene)
DEG_names_Verity_AGEING_DOWN<-(DEG_data_Verity_Slota_DOWN[[4]]$gene)

library(gplots)
# Create a Venn-diagram given just the list of gene-names for both sets
png(paste0(plotting_path_DEG_comp,"DOWN_DEGs_Venndiagram.png"), width = 1000,height = 750)
venn_diagram_DOWN <- venn(list("Verity PRION DOWN" = DEG_names_Verity_PRION_DOWN,
                              # "Verity AGEING DOWN" = DEG_names_Verity_AGEING_DOWN,
                              "Terminal HP DOWN" = DEG_names_Terminal_HP_DOWN,
                              "Terminal CX DOWN" = DEG_names_Terminal_CX_DOWN))
dev.off()

# Inspect the list of classifications
interactions_DOWN<-attr(venn_diagram_DOWN,"intersections")
interactions_DOWN
interactions_DOWN<-(t(data.frame(t(sapply(interactions_DOWN,c)))))
write.csv(interactions_DOWN, file=paste0(plotting_path_DEG_comp,"VennGenes_DOWN.csv"))

```

##Show number of DEGs in the comparison tables

```{r}
interactions_UP
interactions_DOWN
#interactions_RAW
```


```{r}
####MAKE SURE THAT NUMBER OF GENES BEING SELECTED FROM INTERACTIONS OBJECTS ARE THE APPROPRIATE COPMARISON POINTS


##THIS IS A HARDCODED BLOCK, inspect comparisons above to ensure the correct data is being saved

#filter Verity by the genes which crossover with Slota to show the crossoverlist

#Verity and both Slota all three categories' shared DGEs
Verity_Slota_HP_shared_DEGs_RAW <- subset(DEG_data_Verity_Slota[[3]], gene %in% interactions_RAW[[9]]) #interactions 9 is the crossover between Verity_HP.Slota_HP.Slota_CX
Verity_Slota_HP_shared_DEGs_RAW<-subset(Verity_Slota_HP_shared_DEGs_RAW, select = c('gene','description.x')) #HAd to add .x in description due to erroenous column duplication

write.csv(Verity_Slota_HP_shared_DEGs_RAW, file=paste0(plotting_path_DEG_comp,"Shared_DGEs_Vprion_S_RAW.csv"))

#Verity and Slota HP upregulated DGEs
Verity_Slota_HP_shared_DEGs_UP <- subset(DEG_data_Verity_Slota[[3]], gene %in% interactions_UP[[6]]) #interactions 6 is the crossover between Verity_HP.SlotA_HP
Verity_Slota_HP_shared_DEGs_UP <- subset(Verity_Slota_HP_shared_DEGs_UP, select = c('gene','description.x')) 

write.csv(Verity_Slota_HP_shared_DEGs_UP, file=paste0(plotting_path_DEG_comp,"Shared_DGEs_Vprion_Shp_UP.csv"))

#Verity HP and Slota CX upregulated DGEs
Verity_HP_Slota_CX_shared_DEGs_UP <- subset(DEG_data_Verity_Slota[[3]], gene %in% interactions_UP[[5]]) #interactions 5 is the crossover between Verity_HP.Slota_CX
Verity_HP_Slota_CX_shared_DEGs_UP <- subset(Verity_HP_Slota_CX_shared_DEGs_UP, select = c('gene','description.x'))
write.csv(Verity_HP_Slota_CX_shared_DEGs_UP, file=paste0(plotting_path_DEG_comp,"Shared_DGEs_VHP_SCX_UP.csv"))

#Verity and both Slota downregulated DGEs
Verity_Slota_shared_DEGs_DOWN <- subset(DEG_data_Verity_Slota[[3]], gene %in% interactions_DOWN[[6]]) #interactions 5 is the crossover between Verity_HP.SlotA_HP
Verity_Slota_shared_DEGs_DOWN <- subset(Verity_Slota_shared_DEGs_DOWN, select = c('gene','description.x')) 
write.csv(Verity_Slota_shared_DEGs_DOWN, file=paste0(plotting_path_DEG_comp,"Shared_DGEs_V_S_DOWN.csv"))

#Verity and Slota Hippocampus downregulated DGEs
Verity_Slota_HP_shared_DEGs_DOWN <- subset(DEG_data_Verity_Slota[[3]], gene %in% interactions_DOWN[[5]]) #interactions 5 is the crossover between Verity_HP.SlotA_HP
Verity_Slota_HP_shared_DEGs_DOWN <- subset(Verity_Slota_HP_shared_DEGs_DOWN, select = c('gene','description.x')) 
write.csv(Verity_Slota_HP_shared_DEGs_DOWN, file=paste0(plotting_path_DEG_comp,"Shared_DGEs_V_Shp_DOWN.csv"))
```

#Make joint QC plot

```{r}
EDA_comp<-Xim_violin_EDA/p1_Terminal/p1+plot_annotation(tag_levels = 'A', title = 'EDA of each dataset')
ggsave("joint_EDA.png",plot=EDA_comp,device='png', path=plotting_path_Terminal,width = 40, height = 45, units = 'cm' )
```


#interesting list

```{r}
interesting_list<-c('Nrg1','Adgrb3', 'Nrxn3', 'Rimbp2', 'Tanc1', 'Tnr','Grm5', 'Plxna2', 'Prkag2', 'Prkca','Cacnb2', 'Dlgap2', 'Gria1', 'Grin2a', 'Nrp1', 'Plxna4','Ptn', 'Gpc5','C1qc', 'Ccl3', 'Nfkbia', 'Tnf', 'Tyrobp','Atp1a2', 'Chchd2', 'Aldh1l1', 'Aldoc', 'Cox7c', 'Mt3', 'Slc4a4', 'Ugp2', 'Erbb4', 'Tubb2b', 'Dgkb', 'Grid2', 'Glul', 'Grm3', 'Serpine2', 'C1qa', 'C1qb', 'Ptprc', 'Trem2', 'Icam1','Npm1','Atp2b2', 'Gpm6a', 'Nrxn1', 'Malat1', 'Msmo1', 'Vegfa','Magi2', 'Ctnnd2', 'Syn2', 'Unc13c','Gpc5', 'Slco1c1', 'Tspan7','Abca1', 'C4b', 'Gfap', 'Ldlrad4', 'Ppm1h', 'Ptprj', 'Ski', 'Slc39a14', 'Stat3', 'Trf', 'Zbtb16')

lit_pan_genes<-c("Lcn2","Serpina3n","Steap4","Cxcl10","Hspb1","Vim","Serping1","C3","GLAST","SLC1A3","GLT1","SLC1A2","S1pr3","asprg","tip","Steap4","Hspb1","Osmr","Cp","LR1P","Aldh1l1","Gfap","Aspg","Ggta1","H2-D1","Hsbp1","Iigp1","Stat3")
lit_A1_genes<-c("Il6","Il12b","Il33","Ccl4","Thbs1","Thbs4","Sparcl1","Gyp4","Gyp6","Serping1","H2t23","Ggta1","H2-D1","H2-T23","ligp1","Gbp2","Fbln5","Fkbp5","Srgn","Amigo2")
lit_A2_genes<-c("S100a10","Vegfa","Bdnf","Ntf3","Cd109","Emp1","Cx","SPARCL1","Clcf1","Tgm1","Ptx3","S100a10","Sphk1","Cd109","Ptgs2","Emp1","Slc10a6","Tm4sf1","B3gnt5","Cd14","Hmox1")

interesting_list[interesting_list %in% lit_A2_genes]

```



