#!/usr/bin/sh
#$ -./Ximerakis/Data    #run in Ximerakis directory
#$ -N CRcountsFull    #name of job
#$ -l h_rt=48:00:00 #approximate time taken, (specify more than required for safety)
#$ -l h_vmem=16G     #How much RAM is required
#$ -pe sharedmem 8  #how many cores?
#$ -e CRcountsFull.e #where errors go
#$ -hold_jid Cellranger_genome #wait until the preivous job is done

# Make the job resubmit itself if it runs out of time
#$ -r yes
#$ -notify
trap 'exit 99' sigusr1 sigusr2 sigterm

#This script generates gene counts for the Cimerakis dataset using the reference genome created in the previous script. Last script before the Rmd document

#Grant access to cellranger
#uncomment if the previously updated path has not updated to allow cellranger to be used!
#export PATH=/exports/eddie/scratch/s2268606/yard/cellranger-7.1.0:$PATH #fixme

#Run cellranger count on each fastq generated by bamtofastq
for FASTQ in $(cat Ximerakis_BAM_names.txt) ; do
	cellranger count --id=counts${FASTQ} --fastqs=./${FASTQ} --transcriptome=Cellranger_Mouse_genome/
done